// @m3e/web/dist/bottom-sheet.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { ActionElementBase, EventAttribute, VelocityTracker, ScrollLockController, ResizeController, focusWhenReady, DesignToken, hasAssignedNodes, computeCssSize, prefersReducedMotion, spaceSeparatedStringConverter, HtmlFor } from '@m3e/web/core';
import { LitElement, nothing, html, css, unsafeCSS } from 'lit';
import { M3eInteractivityChecker, isModifierAllowed, addAriaReferencedId, removeAriaReferencedId } from '@m3e/web/core/a11y';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = globalThis,
  e$1 = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s = Symbol(),
  o$2 = new WeakMap();
let n$2 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$1 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$2.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$2.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$2 = t => new n$2("string" == typeof t ? t : t + "", void 0, s),
  S = (s, o) => {
    if (e$1) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$1 = e$1 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$2(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i,
    defineProperty: e,
    getOwnPropertyDescriptor: h,
    getOwnPropertyNames: r$1,
    getOwnPropertySymbols: o$1,
    getPrototypeOf: n$1
  } = Object,
  a = globalThis,
  c = a.trustedTypes,
  l = c ? c.emptyScript : "",
  p = a.reactiveElementPolyfillSupport,
  d = (t, s) => t,
  u = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f = (t, s) => !i(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    useDefault: false,
    hasChanged: f
  };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties"))) return;
    const t = n$1(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t = this.properties,
        s = [...r$1(t), ...o$1(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$1(s));
    } else void 0 !== s && i.push(c$1(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d("elementProperties")] = new Map(), y[d("finalized")] = new Map(), p?.({
  ReactiveElement: y
}), (a.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    hasChanged: f
  },
  r = (t = o, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n(t) {
  return (e, o) => "object" == typeof o ? r(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * An element, nested within a clickable element, used to close a parenting bottom sheet.
 * @tag m3e-bottom-sheet-action
 *
 * @slot - Renders the content of the action.
 */
let M3eBottomSheetActionElement = class M3eBottomSheetActionElement extends ActionElementBase {
  /** @inheritdoc */
  _onClick() {
    this.closest("m3e-bottom-sheet")?.hide();
  }
};
M3eBottomSheetActionElement = __decorate([t$1("m3e-bottom-sheet-action")], M3eBottomSheetActionElement);

var _M3eBottomSheetElement_instances, _M3eBottomSheetElement_documentClickHandler, _M3eBottomSheetElement_documentKeyDownHandler, _M3eBottomSheetElement_windowResizeHandler, _M3eBottomSheetElement_inerts, _M3eBottomSheetElement_velocityTracker, _M3eBottomSheetElement_scrollLockController, _M3eBottomSheetElement_resizeController, _M3eBottomSheetElement_trigger, _M3eBottomSheetElement_dragState, _M3eBottomSheetElement_dragged, _M3eBottomSheetElement_activeDetent, _M3eBottomSheetElement_requestDetent, _M3eBottomSheetElement_cachedContentHeight, _M3eBottomSheetElement_cachedHeaderHeight, _M3eBottomSheetElement_snapAnimation, _M3eBottomSheetElement_handleHeaderSlotChange, _M3eBottomSheetElement_handleHeaderPointerDown, _M3eBottomSheetElement_handleHeaderPointerMove, _M3eBottomSheetElement_handleHeaderPointerUp, _M3eBottomSheetElement_handleDocumentClick, _M3eBottomSheetElement_handleDocumentKeyDown, _M3eBottomSheetElement_handleWindowResize, _M3eBottomSheetElement_handleDragHandleClick, _M3eBottomSheetElement_handleDragHandleKeyDown, _M3eBottomSheetElement_getNextHigherDetent, _M3eBottomSheetElement_getNextLowerDetent, _M3eBottomSheetElement_getClosestDetent, _M3eBottomSheetElement_computeDetentHeight, _M3eBottomSheetElement_computeMaxHeight, _M3eBottomSheetElement_computeMinHeight, _M3eBottomSheetElement_computePeekHeight, _M3eBottomSheetElement_computeFitHeight, _M3eBottomSheetElement_handleSectionResize, _M3eBottomSheetElement_snapToDetent, _M3eBottomSheetElement_snapToHeight, _M3eBottomSheetElement_updateHeight, _M3eBottomSheetElement_applyInert, _M3eBottomSheetElement_restoreInert;
var M3eBottomSheetElement_1;
/**
 * A sheet used to show secondary content anchored to the bottom of the screen.
 *
 * @description
 * The `m3e-bottom-sheet` component implements a Material 3 bottom sheet with
 * gesture‑driven resizing, detent snapping, and adaptive motion. It behaves as
 * a heavy surface: transitions use non‑bouncy, decelerating motion.
 *
 * The sheet supports direct manipulation through vertical drag gestures.
 * Movement below an 8px threshold is ignored to ensure reliable tap‑to‑cycle
 * behavior on the handle. Once activated, dragging updates the sheet height
 * with friction near the minimum height to prevent abrupt collapse.
 *
 * When detents are defined, the sheet snaps to the nearest detent on release.
 * If hideable, downward gestures may dismiss the sheet using either velocity
 * or a small distance threshold below the collapsed detent. When no detents
 * are present, the sheet behaves as a simple open/hidden surface with a
 * bottom‑measured hide threshold.
 *
 * @example
 * The following example shows a modal bottom sheet with a drag handle,
 * three detents (`fit`, `half`, and `full`), and an action list. The sheet
 * is opened using a dedicated `m3e-bottom-sheet-trigger` associated by ID.
 * ```html
 * <m3e-button>
 *   <m3e-bottom-sheet-trigger for="bottomSheet">
 *     Open Bottom sheet
 *   </m3e-bottom-sheet-trigger>
 * </m3e-button>
 *
 * <m3e-bottom-sheet id="bottomSheet" modal handle hideable detents="fit half full">
 *   <m3e-action-list>
 *     <m3e-list-action autofocus>
 *       <m3e-bottom-sheet-action>Google Keep</m3e-bottom-sheet-action>
 *       <span slot="supporting-text">Add to a note</span>
 *     </m3e-list-action>
 *     <m3e-list-action>
 *       <m3e-bottom-sheet-action>Google Docs</m3e-bottom-sheet-action>
 *       <span slot="supporting-text">Embed in a document</span>
 *     </m3e-list-action>
 *   </m3e-action-list>
 * </m3e-bottom-sheet>
 * ```
 *
 * @tag m3e-bottom-sheet
 *
 * @slot - Renders the content of the sheet.
 * @slot header - Renders the header of the sheet.
 *
 * @attr detent - The zero‑based index of the detent the sheet should open to.
 * @attr detents - Detents (discrete height states) the sheet can snap to.
 * @attr handle - Whether to display a drag handle and enable the top region of the sheet as a gesture surface for dragging between detents.
 * @attr handle-label - The accessible label given to the drag handle.
 * @attr hideable - Whether the bottom sheet can hide when its swiped down.
 * @attr hide-friction - The friction coefficient to hide the sheet, or set it to the next closest expanded detent.
 * @attr modal - Whether the bottom sheet behaves as modal.
 * @attr open - Whether the bottom sheet is open.
 *
 * @fires opening - Emitted when the sheet begins to open.
 * @fires opened - Emitted when the sheet has opened.
 * @fires cancel - Emitted when the sheet is cancelled.
 * @fires closing - Emitted when the sheet begins to close.
 * @fires closed - Emitted when the sheet has closed.
 *
 * @cssprop --m3e-bottom-sheet-width - The width of the sheet.
 * @cssprop --m3e-bottom-sheet-max-width - The maximum width of the sheet.
 * @cssprop --m3e-bottom-sheet-container-color - The background color of the sheet container.
 * @cssprop --m3e-bottom-sheet-elevation - The elevation level when not modal.
 * @cssprop --m3e-bottom-sheet-modal-elevation - The elevation level when modal.
 * @cssprop --m3e-bottom-sheet-full-elevation - The elevation level when full height.
 * @cssprop --m3e-bottom-sheet-z-index - The z-index of the non-modal sheet.
 * @cssprop --m3e-bottom-sheet-minimized-container-shape - The border radius when minimized.
 * @cssprop --m3e-bottom-sheet-container-shape - The border radius of the sheet container.
 * @cssprop --m3e-bottom-sheet-full-container-shape - The border radius when full height.
 * @cssprop --m3e-bottom-sheet-scrim-color - The color of the scrim overlay.
 * @cssprop --m3e-bottom-sheet-scrim-opacity - The opacity of the scrim overlay.
 * @cssprop --m3e-bottom-sheet-peek-height - The visible height when minimized.
 * @cssprop --m3e-bottom-sheet-compact-top-space - The top space in compact mode.
 * @cssprop --m3e-bottom-sheet-top-space - The top space in standard mode.
 * @cssprop --m3e-bottom-sheet-padding-block - The vertical padding.
 * @cssprop --m3e-bottom-sheet-padding-inline - The horizontal padding.
 * @cssprop --m3e-bottom-sheet-handle-container-height - The height of the drag handle container.
 * @cssprop --m3e-bottom-sheet-handle-width - The width of the drag handle.
 * @cssprop --m3e-bottom-sheet-handle-height - The height of the drag handle.
 * @cssprop --m3e-bottom-sheet-handle-shape - The border radius of the handle.
 * @cssprop --m3e-bottom-sheet-handle-color - The color of the drag handle.
 * @cssprop --m3e-bottom-sheet-handle-focus-ring-offset - The offset of the focus ring around the handle.
 * @cssprop --m3e-bottom-sheet-color - The foreground (text) color of the sheet.
 * @cssprop --m3e-bottom-sheet-content-font-size - Font size for the sheet content.
 * @cssprop --m3e-bottom-sheet-content-font-weight - Font weight for the sheet content.
 * @cssprop --m3e-bottom-sheet-content-line-height - Line height for the sheet content.
 * @cssprop --m3e-bottom-sheet-content-tracking - Letter spacing (tracking) for the sheet content.
 * @cssprop --m3e-bottom-sheet-header-font-size - Font size for the sheet header.
 * @cssprop --m3e-bottom-sheet-header-font-weight - Font weight for the sheet header.
 * @cssprop --m3e-bottom-sheet-header-line-height - Line height for the sheet header.
 * @cssprop --m3e-bottom-sheet-header-tracking - Letter spacing (tracking) for the sheet header.
 */
let M3eBottomSheetElement = M3eBottomSheetElement_1 = class M3eBottomSheetElement extends EventAttribute(LitElement, "opening", "opened", "cancel", "closing", "closed") {
  constructor() {
    super(...arguments);
    _M3eBottomSheetElement_instances.add(this);
    /** @private */
    _M3eBottomSheetElement_documentClickHandler.set(this, e => __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleDocumentClick).call(this, e));
    /** @private */
    _M3eBottomSheetElement_documentKeyDownHandler.set(this, e => __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleDocumentKeyDown).call(this, e));
    /** @private */
    _M3eBottomSheetElement_windowResizeHandler.set(this, () => __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleWindowResize).call(this));
    /** @private */
    _M3eBottomSheetElement_inerts.set(this, new Array());
    /** @private */
    _M3eBottomSheetElement_velocityTracker.set(this, new VelocityTracker());
    /** @private */
    _M3eBottomSheetElement_scrollLockController.set(this, new ScrollLockController(this));
    /** @private */
    _M3eBottomSheetElement_resizeController.set(this, new ResizeController(this, {
      target: null,
      skipInitial: true,
      callback: x => __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleSectionResize).call(this, x)
    }));
    /** @private */
    _M3eBottomSheetElement_trigger.set(this, null);
    /** @private */
    _M3eBottomSheetElement_dragState.set(this, void 0);
    /** @private */
    _M3eBottomSheetElement_dragged.set(this, false);
    /** @private */
    _M3eBottomSheetElement_activeDetent.set(this, 0);
    /** @private */
    _M3eBottomSheetElement_requestDetent.set(this, void 0);
    /** @private */
    _M3eBottomSheetElement_cachedContentHeight.set(this, 0);
    /** @private */
    _M3eBottomSheetElement_cachedHeaderHeight.set(this, 0);
    /** @private */
    _M3eBottomSheetElement_snapAnimation.set(this, void 0);
    /**
     * Whether the bottom sheet behaves as modal.
     * @default false
     */
    this.modal = false;
    /**
     * Whether the bottom sheet is open.
     * @default false
     */
    this.open = false;
    /**
     * Whether to display a drag handle and enable the top region of the sheet as a gesture
     * surface for dragging between detents.
     * @default false
     */
    this.handle = false;
    /**
     * The accessible label given to the drag handle.
     * @default "Drag handle"
     */
    this.handleLabel = "Drag handle";
    /**
     * Detents (discrete height states) the sheet can snap to.
     * @default []
     */
    this.detents = [];
    /**
     * The zero‑based index of the detent the sheet should open to.
     * @default 0
     */
    this.detent = 0;
    /**
     * Whether the bottom sheet can hide when its swiped down.
     * @default false
     */
    this.hideable = false;
    /**
     * The friction coefficient to hide the sheet, or set it to the next closest expanded detent.
     * @default 0.5
     */
    this.hideFriction = 0.5;
  }
  /**
   * Shows the sheet.
   * @param {number} detent The zero‑based index of the detent the sheet should open to.
   */
  show(detent = this.detent) {
    if (!this.open) {
      __classPrivateFieldSet(this, _M3eBottomSheetElement_requestDetent, detent, "f");
      this.open = true;
    } else if (detent !== undefined && __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f") !== detent) {
      __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToDetent).call(this, detent);
    }
  }
  /** Hides the sheet. */
  hide() {
    this.open = false;
  }
  /**
   * Toggles the opened state of the sheet.
   * @param {number} detent The zero‑based index of the detent the sheet should open to.
   */
  toggle(detent) {
    if (this.open) {
      this.hide();
    } else {
      this.show(detent);
    }
  }
  /** Moves the sheet to the next detent. */
  cycle() {
    var _a;
    if (!this.open) {
      this.show();
    } else if (this.detents.length > 0) {
      if (__classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f") < this.detents.length - 1) {
        __classPrivateFieldSet(this, _M3eBottomSheetElement_activeDetent, (_a = __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f"), _a++, _a), "f");
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToDetent).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f"));
      } else if (this.hideable) {
        this.hide();
      }
    } else {
      this.hide();
    }
  }
  /** @inheritdoc */
  update(changedProperties) {
    super.update(changedProperties);
    if (changedProperties.has("modal")) {
      this.role = this.modal ? "dialog" : "region";
      this.ariaModal = this.modal ? "true" : null;
      this.popover = this.modal ? "manual" : null;
    }
  }
  /** @inheritdoc */
  firstUpdated(_changedProperties) {
    super.firstUpdated(_changedProperties);
    const content = this.shadowRoot?.querySelector(".content");
    if (content) {
      __classPrivateFieldSet(this, _M3eBottomSheetElement_cachedContentHeight, content.clientHeight, "f");
      __classPrivateFieldGet(this, _M3eBottomSheetElement_resizeController, "f").observe(content);
    }
    const header = this.shadowRoot?.querySelector(".header");
    if (header) {
      __classPrivateFieldSet(this, _M3eBottomSheetElement_cachedHeaderHeight, header.clientHeight, "f");
      __classPrivateFieldGet(this, _M3eBottomSheetElement_resizeController, "f").observe(header);
    }
  }
  /** @inheritdoc */
  updated(_changedProperties) {
    super.updated(_changedProperties);
    if (_changedProperties.has("open")) {
      if (this.open) {
        if (!this.dispatchEvent(new Event("opening", {
          cancelable: true
        }))) {
          this.open = false;
          return;
        }
        // Only one sheet can be open at a time.
        if (M3eBottomSheetElement_1.__openSheet !== this) {
          M3eBottomSheetElement_1.__openSheet?.hide();
        }
        M3eBottomSheetElement_1.__openSheet = this;
        this.inert = false;
        window.addEventListener("resize", __classPrivateFieldGet(this, _M3eBottomSheetElement_windowResizeHandler, "f"));
        // If there are detents (regardless of handle) open to the requested, default (or first) detent.
        // Otherwise, open to fit unless its greater than half.
        if (this.detents.length > 0) {
          __classPrivateFieldSet(this, _M3eBottomSheetElement_activeDetent, Math.min(Math.max(0, __classPrivateFieldGet(this, _M3eBottomSheetElement_requestDetent, "f") ?? this.detent), this.detents.length - 1), "f");
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, this.detents[__classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")]));
        } else {
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, Math.min(__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, "fit"), __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, "half")));
        }
        __classPrivateFieldSet(this, _M3eBottomSheetElement_requestDetent, undefined, "f");
      } else {
        if (!this.dispatchEvent(new Event("closing", {
          cancelable: true
        }))) {
          this.open = true;
          return;
        }
        // The following ensures layout is stable so that first tab isn't "skipped"
        requestAnimationFrame(() => this.inert = true);
        window.removeEventListener("resize", __classPrivateFieldGet(this, _M3eBottomSheetElement_windowResizeHandler, "f"));
        if (M3eBottomSheetElement_1.__openSheet === this) {
          M3eBottomSheetElement_1.__openSheet = undefined;
        }
      }
      if (this.modal) {
        if (this.open) {
          __classPrivateFieldSet(this, _M3eBottomSheetElement_trigger, document.activeElement, "f");
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_applyInert).call(this);
          __classPrivateFieldGet(this, _M3eBottomSheetElement_scrollLockController, "f").lock();
          this.showPopover();
          requestAnimationFrame(() => {
            document.addEventListener("click", __classPrivateFieldGet(this, _M3eBottomSheetElement_documentClickHandler, "f"));
            document.addEventListener("keydown", __classPrivateFieldGet(this, _M3eBottomSheetElement_documentKeyDownHandler, "f"));
          });
          let focusable = this.querySelector("[autofocus]");
          if (!focusable || !M3eInteractivityChecker.isFocusable(focusable)) {
            focusable = this.shadowRoot?.querySelector(".handle");
          }
          if (focusable) {
            focusWhenReady(focusable);
          }
        } else {
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToHeight).call(this, 0).then(() => {
            __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_restoreInert).call(this);
            __classPrivateFieldGet(this, _M3eBottomSheetElement_scrollLockController, "f").unlock();
            document.removeEventListener("click", __classPrivateFieldGet(this, _M3eBottomSheetElement_documentClickHandler, "f"));
            document.removeEventListener("keydown", __classPrivateFieldGet(this, _M3eBottomSheetElement_documentKeyDownHandler, "f"));
            this.hidePopover();
            if (__classPrivateFieldGet(this, _M3eBottomSheetElement_trigger, "f") instanceof HTMLElement) {
              __classPrivateFieldGet(this, _M3eBottomSheetElement_trigger, "f").focus();
            }
            __classPrivateFieldSet(this, _M3eBottomSheetElement_trigger, null, "f");
          });
        }
      }
      this.dispatchEvent(new Event(this.open ? "opened" : "closed"));
    }
  }
  /** @inheritdoc */
  render() {
    return html`<m3e-focus-trap ?disabled="${!this.modal}"><div class="base"><m3e-elevation class="elevation"></m3e-elevation><div class="header" @pointerdown="${__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleHeaderPointerDown)}" @pointermove="${__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleHeaderPointerMove)}" @pointerup="${__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleHeaderPointerUp)}">${this.handle ? html`<div class="handle-row"><div id="handle" class="handle" role="button" aria-label="${this.handleLabel}" tabindex="0" @click="${__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleDragHandleClick)}" @keydown="${__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleDragHandleKeyDown)}"><m3e-focus-ring class="focus-ring" for="handle"></m3e-focus-ring><div class="handle-touch" aria-hidden="true"></div></div></div>` : nothing}<slot name="header" @slotchange="${__classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_handleHeaderSlotChange)}"></slot></div><div class="body"><div class="content"><slot></slot></div></div></div></m3e-focus-trap>`;
  }
};
_M3eBottomSheetElement_documentClickHandler = new WeakMap();
_M3eBottomSheetElement_documentKeyDownHandler = new WeakMap();
_M3eBottomSheetElement_windowResizeHandler = new WeakMap();
_M3eBottomSheetElement_inerts = new WeakMap();
_M3eBottomSheetElement_velocityTracker = new WeakMap();
_M3eBottomSheetElement_scrollLockController = new WeakMap();
_M3eBottomSheetElement_resizeController = new WeakMap();
_M3eBottomSheetElement_trigger = new WeakMap();
_M3eBottomSheetElement_dragState = new WeakMap();
_M3eBottomSheetElement_dragged = new WeakMap();
_M3eBottomSheetElement_activeDetent = new WeakMap();
_M3eBottomSheetElement_requestDetent = new WeakMap();
_M3eBottomSheetElement_cachedContentHeight = new WeakMap();
_M3eBottomSheetElement_cachedHeaderHeight = new WeakMap();
_M3eBottomSheetElement_snapAnimation = new WeakMap();
_M3eBottomSheetElement_instances = new WeakSet();
_M3eBottomSheetElement_handleHeaderSlotChange = function _M3eBottomSheetElement_handleHeaderSlotChange(e) {
  this.classList.toggle("-has-header", hasAssignedNodes(e.target));
};
_M3eBottomSheetElement_handleHeaderPointerDown = function _M3eBottomSheetElement_handleHeaderPointerDown(e) {
  if (e.button === 2) return;
  if (e.target instanceof HTMLElement && M3eInteractivityChecker.isFocusable(e.target)) {
    return;
  }
  e.target.setPointerCapture(e.pointerId);
  e.target.style.cursor = "grabbing";
  __classPrivateFieldGet(this, _M3eBottomSheetElement_velocityTracker, "f").reset();
  __classPrivateFieldGet(this, _M3eBottomSheetElement_velocityTracker, "f").add(e.clientY);
  __classPrivateFieldSet(this, _M3eBottomSheetElement_dragState, {
    startY: e.clientY,
    startHeight: this.clientHeight,
    maxHeight: __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMaxHeight).call(this),
    minHeight: __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMinHeight).call(this)
  }, "f");
  __classPrivateFieldSet(this, _M3eBottomSheetElement_dragged, false, "f");
};
_M3eBottomSheetElement_handleHeaderPointerMove = function _M3eBottomSheetElement_handleHeaderPointerMove(e) {
  if (!__classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f")) {
    return;
  }
  const minDragThreshold = 8;
  if (Math.abs(e.clientY - __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").startY) <= minDragThreshold) {
    // Ignore movement under threshold
    return;
  }
  (e.getCoalescedEvents?.() ?? [e]).forEach(x => __classPrivateFieldGet(this, _M3eBottomSheetElement_velocityTracker, "f").add(x.clientY, e.timeStamp));
  let newHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").startHeight - (e.clientY - __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").startY);
  if (newHeight < __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").minHeight) {
    const overshoot = (__classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").minHeight - newHeight) * this.hideFriction;
    newHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").minHeight - overshoot;
  }
  __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, Math.min(__classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").maxHeight, newHeight));
  __classPrivateFieldSet(this, _M3eBottomSheetElement_dragged, true, "f");
};
_M3eBottomSheetElement_handleHeaderPointerUp = function _M3eBottomSheetElement_handleHeaderPointerUp(e) {
  if (!__classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f")) return;
  try {
    e.target.releasePointerCapture(e.pointerId);
    e.target.style.cursor = "";
    if (!__classPrivateFieldGet(this, _M3eBottomSheetElement_dragged, "f")) return;
    const significantVelocityThreshold = e.pointerType === "touch" ? 1200 : 500;
    const velocity = __classPrivateFieldGet(this, _M3eBottomSheetElement_velocityTracker, "f").getVelocity();
    __classPrivateFieldGet(this, _M3eBottomSheetElement_velocityTracker, "f").reset();
    if (this.hideable && velocity >= significantVelocityThreshold) {
      if (this.dispatchEvent(new Event("cancel", {
        cancelable: true
      }))) {
        this.hide();
      }
    } else if (Math.abs(velocity) >= significantVelocityThreshold) {
      if (this.detents.length > 0) {
        const nextDetent = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_getNextHigherDetent).call(this);
        if (nextDetent !== __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")) {
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToDetent).call(this, nextDetent);
        }
      } else {
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, "full"));
      }
    } else {
      const hideDistanceThreshold = 20;
      if (this.hideable) {
        const collapsed = __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").minHeight;
        if (this.clientHeight < collapsed - hideDistanceThreshold) {
          this.hide();
          return;
        }
      }
      if (this.detents.length > 0) {
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToDetent).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_getClosestDetent).call(this));
      } else if (this.clientHeight < __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").minHeight) {
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_dragState, "f").minHeight);
      }
    }
  } finally {
    __classPrivateFieldSet(this, _M3eBottomSheetElement_dragState, undefined, "f");
  }
};
_M3eBottomSheetElement_handleDocumentClick = function _M3eBottomSheetElement_handleDocumentClick(e) {
  if (this.open && this.modal && !e.composedPath().includes(this) && this.dispatchEvent(new Event("cancel", {
    cancelable: true
  }))) {
    this.hide();
  }
};
_M3eBottomSheetElement_handleDocumentKeyDown = function _M3eBottomSheetElement_handleDocumentKeyDown(e) {
  if (this.open && this.modal && e.key === "Escape" && !e.shiftKey && !e.ctrlKey) {
    e.preventDefault();
    if (this.dispatchEvent(new Event("cancel", {
      cancelable: true
    }))) {
      this.hide();
    }
  }
};
_M3eBottomSheetElement_handleWindowResize = function _M3eBottomSheetElement_handleWindowResize() {
  if (this.detents.length > 0 && this.detents[__classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")] === "half") {
    __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, "half"));
    return;
  }
  const maxHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMaxHeight).call(this);
  if (this.classList.contains("-full")) {
    __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, maxHeight);
  } else if (this.clientHeight > maxHeight) {
    __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, maxHeight);
  }
};
_M3eBottomSheetElement_handleDragHandleClick = function _M3eBottomSheetElement_handleDragHandleClick() {
  if (!__classPrivateFieldGet(this, _M3eBottomSheetElement_dragged, "f")) {
    this.cycle();
  }
  __classPrivateFieldSet(this, _M3eBottomSheetElement_dragged, false, "f");
};
_M3eBottomSheetElement_handleDragHandleKeyDown = function _M3eBottomSheetElement_handleDragHandleKeyDown(e) {
  if (e.defaultPrevented || !isModifierAllowed(e)) return;
  switch (e.key) {
    case "Up":
    case "ArrowUp":
      e.preventDefault();
      if (this.detents.length > 0) {
        const nextDetent = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_getNextHigherDetent).call(this);
        if (nextDetent !== __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")) {
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToDetent).call(this, nextDetent);
        }
      } else {
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, "full"));
      }
      break;
    case "Down":
    case "ArrowDown":
      e.preventDefault();
      if (this.detents.length > 0) {
        const nextDetent = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_getNextLowerDetent).call(this);
        if (nextDetent !== __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")) {
          __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToDetent).call(this, nextDetent);
        } else if (this.hideable) {
          this.hide();
        }
      } else if (this.hideable) {
        this.hide();
      }
      break;
  }
};
_M3eBottomSheetElement_getNextHigherDetent = function _M3eBottomSheetElement_getNextHigherDetent() {
  const currentHeight = this.clientHeight;
  let nextHeight = Infinity;
  let nextDetent = __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f");
  for (let i = 0; i < this.detents.length; i++) {
    if (i === __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")) continue;
    const detentHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, this.detents[i]);
    if (detentHeight > currentHeight && detentHeight < nextHeight) {
      nextHeight = detentHeight;
      nextDetent = i;
    }
  }
  return nextDetent;
};
_M3eBottomSheetElement_getNextLowerDetent = function _M3eBottomSheetElement_getNextLowerDetent() {
  const currentHeight = this.clientHeight;
  let nextHeight = -Infinity;
  let nextDetent = __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f");
  for (let i = 0; i < this.detents.length; i++) {
    if (i === __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")) continue;
    const detentHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, this.detents[i]);
    if (detentHeight < currentHeight && detentHeight > nextHeight) {
      nextHeight = detentHeight;
      nextDetent = i;
    }
  }
  return nextDetent;
};
_M3eBottomSheetElement_getClosestDetent = function _M3eBottomSheetElement_getClosestDetent() {
  const currentHeight = this.clientHeight;
  let closestDetent = __classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f");
  let closestDistance = Infinity;
  for (let i = 0; i < this.detents.length; i++) {
    const detentHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, this.detents[i]);
    const distance = Math.abs(currentHeight - detentHeight);
    if (distance < closestDistance) {
      closestDistance = distance;
      closestDetent = i;
    }
  }
  return closestDetent;
};
_M3eBottomSheetElement_computeDetentHeight = function _M3eBottomSheetElement_computeDetentHeight(detent) {
  switch (detent) {
    case "collapsed":
      return __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMinHeight).call(this);
    case "half":
      return __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMaxHeight).call(this) * 0.5;
    case "full":
      return __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMaxHeight).call(this);
    case "fit":
      return __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeFitHeight).call(this);
  }
  if (detent.endsWith("%")) {
    return __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMaxHeight).call(this) * (parseFloat(detent) / 100);
  }
  if (detent.endsWith("px")) {
    return parseFloat(detent);
  }
  return __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMinHeight).call(this);
};
_M3eBottomSheetElement_computeMaxHeight = function _M3eBottomSheetElement_computeMaxHeight() {
  const base = this.shadowRoot?.querySelector(".base");
  return window.innerHeight - (base ? computeCssSize(base, "var(--_bottom-sheet-top-space)") : 0);
};
_M3eBottomSheetElement_computeMinHeight = function _M3eBottomSheetElement_computeMinHeight() {
  return this.detents.includes("fit") && !this.detents.includes("collapsed") ? __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeFitHeight).call(this) : __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computePeekHeight).call(this);
};
_M3eBottomSheetElement_computePeekHeight = function _M3eBottomSheetElement_computePeekHeight() {
  const base = this.shadowRoot?.querySelector(".base");
  return __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedHeaderHeight, "f") + (base ? computeCssSize(base, "var(--_bottom-sheet-peek-height)") : 0);
};
_M3eBottomSheetElement_computeFitHeight = function _M3eBottomSheetElement_computeFitHeight() {
  const body = this.shadowRoot?.querySelector(".body");
  if (!body) {
    return 0;
  }
  const bodyStyle = getComputedStyle(body);
  return __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedHeaderHeight, "f") + __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedContentHeight, "f") + parseFloat(bodyStyle.paddingBlockStart) + parseFloat(bodyStyle.paddingBlockEnd);
};
_M3eBottomSheetElement_handleSectionResize = function _M3eBottomSheetElement_handleSectionResize(entries) {
  const oldCachedContentHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedContentHeight, "f");
  const oldCachedHeaderHeight = __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedHeaderHeight, "f");
  for (const entry of entries) {
    if (entry.target.classList.contains("content")) {
      __classPrivateFieldSet(this, _M3eBottomSheetElement_cachedContentHeight, (Array.isArray(entry.borderBoxSize) ? entry.borderBoxSize[0] : entry.borderBoxSize).blockSize, "f");
    } else if (entry.target.classList.contains("header")) {
      __classPrivateFieldSet(this, _M3eBottomSheetElement_cachedHeaderHeight, (Array.isArray(entry.borderBoxSize) ? entry.borderBoxSize[0] : entry.borderBoxSize).blockSize, "f");
    }
  }
  if (this.open && this.detents.length > 0 && (oldCachedContentHeight !== __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedContentHeight, "f") || oldCachedHeaderHeight !== __classPrivateFieldGet(this, _M3eBottomSheetElement_cachedHeaderHeight, "f"))) {
    switch (this.detents[__classPrivateFieldGet(this, _M3eBottomSheetElement_activeDetent, "f")]) {
      case "fit":
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeFitHeight).call(this));
        break;
      case "collapsed":
        __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMinHeight).call(this));
        break;
    }
  }
};
_M3eBottomSheetElement_snapToDetent = function _M3eBottomSheetElement_snapToDetent(detent) {
  if (detent >= 0 && detent < this.detents.length) {
    __classPrivateFieldSet(this, _M3eBottomSheetElement_activeDetent, detent, "f");
    __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_snapToHeight).call(this, __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeDetentHeight).call(this, this.detents[detent]));
  }
};
_M3eBottomSheetElement_snapToHeight = /** @private */
async function _M3eBottomSheetElement_snapToHeight(height) {
  if (__classPrivateFieldGet(this, _M3eBottomSheetElement_snapAnimation, "f")) {
    __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, this.clientHeight);
    __classPrivateFieldGet(this, _M3eBottomSheetElement_snapAnimation, "f")?.cancel();
    __classPrivateFieldSet(this, _M3eBottomSheetElement_snapAnimation, undefined, "f");
  }
  if (!prefersReducedMotion()) {
    __classPrivateFieldSet(this, _M3eBottomSheetElement_snapAnimation, this.animate([{
      "--_bottom-sheet-height": `${this.clientHeight}px`
    }, {
      "--_bottom-sheet-height": `${height}px`
    }], {
      duration: 250,
      easing: "cubic-bezier(0.2, 0.0, 0, 1.0)"
    }), "f");
    __classPrivateFieldGet(this, _M3eBottomSheetElement_snapAnimation, "f").onfinish = () => {
      __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, height);
      __classPrivateFieldSet(this, _M3eBottomSheetElement_snapAnimation, undefined, "f");
    };
    await __classPrivateFieldGet(this, _M3eBottomSheetElement_snapAnimation, "f").finished;
  } else {
    __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_updateHeight).call(this, height);
  }
};
_M3eBottomSheetElement_updateHeight = function _M3eBottomSheetElement_updateHeight(height) {
  this.style.setProperty("--_bottom-sheet-height", `${height}px`);
  this.classList.toggle("-full", height === __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computeMaxHeight).call(this));
  const content = this.shadowRoot?.querySelector(".content");
  if (content) {
    content.inert = height <= __classPrivateFieldGet(this, _M3eBottomSheetElement_instances, "m", _M3eBottomSheetElement_computePeekHeight).call(this);
  }
};
_M3eBottomSheetElement_applyInert = function _M3eBottomSheetElement_applyInert() {
  __classPrivateFieldGet(this, _M3eBottomSheetElement_inerts, "f").length = 0;
  for (let current = this; current.parentNode && current.parentNode !== document.documentElement; current = current.parentNode) {
    for (const sibling of current.parentNode.children) {
      if (sibling instanceof HTMLElement && sibling !== current && !sibling.inert) {
        sibling.inert = true;
        __classPrivateFieldGet(this, _M3eBottomSheetElement_inerts, "f").push(sibling);
      }
    }
  }
};
_M3eBottomSheetElement_restoreInert = function _M3eBottomSheetElement_restoreInert() {
  __classPrivateFieldGet(this, _M3eBottomSheetElement_inerts, "f").forEach(x => x.inert = false);
  __classPrivateFieldGet(this, _M3eBottomSheetElement_inerts, "f").length = 0;
};
(() => {
  if (document) {
    const lightDomStyle = new CSSStyleSheet();
    lightDomStyle.replaceSync(css`@property --_bottom-sheet-height { syntax: "<length>"; inherits: false; initial-value: 50vh; } m3e-bottom-sheet > [slot="header"] { margin-block-end: var(--m3e-bottom-sheet-padding-block, 0.5rem); margin-inline: var(--m3e-bottom-sheet-padding-inline, 1rem); }</length>`.toString());
    document.adoptedStyleSheets.push(lightDomStyle);
  }
})();
/** The styles of the element. */
M3eBottomSheetElement.styles = css`:host { display: block; position: fixed; left: 50%; top: calc(100vh - var(--_bottom-sheet-height)); margin: 0; padding: 0; outline: none; overflow: hidden; border: none; box-sizing: border-box; width: var(--m3e-bottom-sheet-width, 100%); max-width: var(--m3e-bottom-sheet-max-width, 40rem); height: var(--_bottom-sheet-height); color: var(--m3e-bottom-sheet-color, ${DesignToken.color.onSurface}); background-color: var(--m3e-bottom-sheet-container-color, ${DesignToken.color.surfaceContainerLow}); transition: ${unsafeCSS(`transform ${DesignToken.motion.duration.medium2} ${DesignToken.motion.easing.standardDecelerate},
        border-radius ${DesignToken.motion.duration.medium2} ${DesignToken.motion.easing.standard}`)}; } :host(:not([modal]):not(.-full)) .elevation { --m3e-elevation-level: var(--m3e-bottom-sheet-elevation, ${DesignToken.elevation.level1}); } :host([modal]:not(.-full)) .elevation { --m3e-elevation-level: var(--m3e-bottom-sheet-modal-elevation, ${DesignToken.elevation.level1}); } :host(.-full) .elevation { --m3e-elevation-level: var(--m3e-bottom-sheet-full-elevation, ${DesignToken.elevation.level1}); } :host(:not([modal])) { z-index: var(--m3e-bottom-sheet-z-index, 10); } :host(:not([modal]):not([open])), :host([modal]:not(:popover-open)) { border-radius: var(--m3e-bottom-sheet-minimized-container-shape, ${DesignToken.shape.corner.none}); transform: translateX(-50%) translateY(100%); } :host(:not([modal])[open]:not(.-full)), :host([modal]:not(.-full):popover-open) { border-radius: var(--m3e-bottom-sheet-container-shape, ${DesignToken.shape.corner.extraLargeTop}); } :host(:not([modal])[open].-full), :host([modal].-full:popover-open) { border-radius: var(--m3e-bottom-sheet-full-container-shape, ${DesignToken.shape.corner.extraLargeTop}); } :host(:not([modal])[open]), :host([modal]:popover-open) { transform: translateX(-50%) translateY(0); } :host([modal])::backdrop { transition: ${unsafeCSS(`background-color ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard}, 
        overlay ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard} allow-discrete,
        visibility ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard} allow-discrete`)}; } :host([modal]:popover-open)::backdrop { background-color: color-mix( in srgb, var(--m3e-bottom-sheet-scrim-color, ${DesignToken.color.scrim}) var(--m3e-bottom-sheet-scrim-opacity, 32%), transparent ); transition: ${unsafeCSS(`background-color ${DesignToken.motion.duration.long2} ${DesignToken.motion.easing.standard}, 
        overlay ${DesignToken.motion.duration.long2} ${DesignToken.motion.easing.standard} allow-discrete,
        visibility ${DesignToken.motion.duration.long2} ${DesignToken.motion.easing.standard} allow-discrete`)}; } @starting-style { :host([modal]:popover-open)::backdrop { background-color: color-mix( in srgb, var(--m3e-bottom-sheet-scrim-color, ${DesignToken.color.scrim}) 0%, transparent ); } } .base { display: flex; border-radius: inherit; flex-direction: column; height: 100%; --_bottom-sheet-peek-height: var(--m3e-bottom-sheet-peek-height, 0); --_bottom-sheet-top-space: var(--m3e-bottom-sheet-compact-top-space, 4.5rem); } @media (max-height: 640px) { .base { --_bottom-sheet-top-space: var(--m3e-bottom-sheet-top-space, 3.5rem); } } .body { flex: 1 1 auto; overflow-y: auto; scrollbar-width: ${DesignToken.scrollbar.thinWidth}; scrollbar-color: ${DesignToken.scrollbar.color}; padding-block-end: var(--m3e-bottom-sheet-padding-block, 0.5rem); padding-inline: var(--m3e-bottom-sheet-padding-inline, 1rem); font-size: var(--m3e-bottom-sheet-content-font-size, ${DesignToken.typescale.standard.body.medium.fontSize}); font-weight: var( --m3e-bottom-sheet-content-font-weight, ${DesignToken.typescale.standard.body.medium.fontWeight} ); line-height: var( --m3e-bottom-sheet-content-line-height, ${DesignToken.typescale.standard.body.medium.lineHeight} ); letter-spacing: var(--m3e-bottom-sheet-content-tracking, ${DesignToken.typescale.standard.body.medium.tracking}); } .content { height: fit-content; } :host(:not([handle]):not(.-has-header)) .header { display: none; } :host(:not([handle]):not(.-has-header)) .body, .header { border-top-left-radius: inherit; border-top-right-radius: inherit; padding-block-start: var(--m3e-bottom-sheet-padding-block, 0.5rem); font-size: var(--m3e-bottom-sheet-header-font-size, ${DesignToken.typescale.standard.title.large.fontSize}); font-weight: var(--m3e-bottom-sheet-header-font-weight, ${DesignToken.typescale.standard.title.large.fontWeight}); line-height: var(--m3e-bottom-sheet-header-line-height, ${DesignToken.typescale.standard.title.large.lineHeight}); letter-spacing: var(--m3e-bottom-sheet-header-tracking, ${DesignToken.typescale.standard.title.large.tracking}); } .header { cursor: grab; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); box-sizing: border-box; flex: none; display: flex; flex-direction: column; min-height: 3rem; --m3e-app-bar-container-color: var(--m3e-bottom-sheet-container-color, ${DesignToken.color.surfaceContainerLow}); } .handle-row { position: relative; flex: none; display: flex; align-items: center; justify-content: center; opacity: 1; visibility: visible; transition: ${unsafeCSS(`opacity ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard},
        padding ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard},
        height ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard},
        visibility ${DesignToken.motion.duration.short3} ${DesignToken.motion.easing.standard} allow-discrete`)}; height: var(--m3e-bottom-sheet-handle-container-height, 1.5rem); } .handle { position: relative; width: var(--m3e-bottom-sheet-handle-width, 2rem); height: var(--m3e-bottom-sheet-handle-height, 0.25rem); border-radius: var(--m3e-bottom-sheet-handle-shape, ${DesignToken.shape.corner.full}); background-color: var(--m3e-bottom-sheet-handle-color, ${DesignToken.color.onSurfaceVariant}); } .focus-ring { top: calc(0px - var(--m3e-bottom-sheet-handle-focus-ring-offset, 0.125rem)); left: calc(0px - var(--m3e-bottom-sheet-handle-focus-ring-offset, 0.125rem)); right: calc(0px - var(--m3e-bottom-sheet-handle-focus-ring-offset, 0.125rem)); bottom: calc(0px - var(--m3e-bottom-sheet-handle-focus-ring-offset, 0.125rem)); } .handle-touch { position: absolute; aspect-ratio: 1 / 1; height: 3rem; left: calc(0px - calc(calc(3rem - var(--m3e-bottom-sheet-handle-width, 2rem)) / 2)); right: calc(0px - calc(calc(3rem - var(--m3e-bottom-sheet-handle-width, 2rem)) / 2)); top: calc( 0px - calc( calc(3rem - var(--m3e-bottom-sheet-handle-container-height, 1.5rem)) - calc( var(--m3e-bottom-sheet-handle-height, 0.25rem) / 2 ) ) ); } @media (prefers-reduced-motion) { :host, :host([modal])::backdrop, :host([modal]:popover-open)::backdrop, .handle { transition: none; } } @media (forced-colors: active) { :host([modal])::backdrop, :host([modal]:popover-open)::backdrop { transition: none; } .base { border-style: solid; border-width: 1px; border-color: CanvasText; } .handle { background-color: ButtonText; } }`;
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eBottomSheetElement.prototype, "modal", void 0);
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eBottomSheetElement.prototype, "open", void 0);
__decorate([n({
  type: Boolean
})], M3eBottomSheetElement.prototype, "handle", void 0);
__decorate([n({
  attribute: "handle-label"
})], M3eBottomSheetElement.prototype, "handleLabel", void 0);
__decorate([n({
  attribute: "detents",
  converter: spaceSeparatedStringConverter
})], M3eBottomSheetElement.prototype, "detents", void 0);
__decorate([n({
  type: Number
})], M3eBottomSheetElement.prototype, "detent", void 0);
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eBottomSheetElement.prototype, "hideable", void 0);
__decorate([n({
  attribute: "hide-friction",
  type: Number
})], M3eBottomSheetElement.prototype, "hideFriction", void 0);
M3eBottomSheetElement = M3eBottomSheetElement_1 = __decorate([t$1("m3e-bottom-sheet")], M3eBottomSheetElement);

/**
 * An element, nested within a clickable element, used to trigger a bottom sheet.
 * @tag m3e-bottom-sheet-trigger
 *
 * @slot - Renders the content of the trigger.
 *
 * @attr detent - The zero‑based index of the detent the sheet should open to.
 * @attr secondary - Marks this trigger as a secondary trigger for accessibility. Secondary triggers do not receive ARIA ownership.
 */
let M3eBottomSheetTriggerElement = class M3eBottomSheetTriggerElement extends HtmlFor(ActionElementBase) {
  constructor() {
    super(...arguments);
    /**
     * Marks this trigger as a secondary trigger for accessibility. Secondary triggers do not receive ARIA ownership.
     * @default false
     */
    this.secondary = false;
  }
  /** @inheritdoc */
  attach(control) {
    if (control instanceof M3eBottomSheetElement) {
      super.attach(control);
      if (this.parentElement) {
        if (!control.modal) {
          if (!this.secondary) {
            this.parentElement.ariaExpanded = "false";
            addAriaReferencedId(this.parentElement, "aria-controls", control.id);
            if (this.parentElement.id) {
              addAriaReferencedId(control, "aria-labelledby", this.parentElement.id);
            }
          }
        } else {
          this.parentElement.ariaHasPopup = "dialog";
        }
      }
    }
  }
  /** @inheritdoc */
  detach() {
    if (this.control && this.parentElement && !this.secondary) {
      removeAriaReferencedId(this.parentElement, "aria-controls", this.control.id);
    }
    super.detach();
  }
  /** @inheritdoc */
  _onClick() {
    if (this.control instanceof M3eBottomSheetElement) {
      if (this.control.modal) {
        this.control.show(this.detent);
      } else {
        this.control.toggle(this.detent);
        if (!this.secondary && this.parentElement) {
          this.parentElement.ariaExpanded = `${this.control.open}`;
        }
      }
    }
  }
};
__decorate([n({
  type: Number
})], M3eBottomSheetTriggerElement.prototype, "detent", void 0);
__decorate([n({
  type: Boolean
})], M3eBottomSheetTriggerElement.prototype, "secondary", void 0);
M3eBottomSheetTriggerElement = __decorate([t$1("m3e-bottom-sheet-trigger")], M3eBottomSheetTriggerElement);

export { M3eBottomSheetActionElement, M3eBottomSheetElement, M3eBottomSheetTriggerElement };
