// @m3e/web/dist/core-bidi.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { isServer } from 'lit';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * Adapted from Angular Material CDK Directionality
 * Source: https://github.com/angular/components/blob/main/src/cdk/bidi/directionality.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _a, _M3eDirectionality_current, _M3eDirectionality_observers, _M3eDirectionality_updateDirection, _M3eDirectionality_handleChange;
/** Taken from `goog.i18n.bidi.isRtlLanguage`. */
const RTL_LOCALES = /^(ar|ckb|dv|he|iw|fa|nqo|ps|sd|ug|ur|yi|.*[-_](Adlm|Arab|Hebr|Nkoo|Rohg|Thaa))(?!.*[-_](Latn|Cyrl)($|-|_))($|-|_)/i;
/** Utility used to determine the directionality of the current document. */
class M3eDirectionality {
  /** The directionality of the current document. */
  static get current() {
    return __classPrivateFieldGet(this, _a, "f", _M3eDirectionality_current);
  }
  /**
   * Observes changes to directionality.
   * @param {()=>void} callback Function invoked when directionality changes.
   * @returns {()=>void} A function used to unobserve changes.
   */
  static observe(callback) {
    __classPrivateFieldGet(this, _a, "f", _M3eDirectionality_observers).push(callback);
    return () => {
      __classPrivateFieldGet(this, _a, "f", _M3eDirectionality_observers).splice(__classPrivateFieldGet(this, _a, "f", _M3eDirectionality_observers).indexOf(callback), 1);
    };
  }
}
_a = M3eDirectionality, _M3eDirectionality_updateDirection = function _M3eDirectionality_updateDirection() {
  const _dir = (document?.body?.dir || document?.documentElement?.dir)?.toLowerCase() || "auto";
  __classPrivateFieldSet(this, _a, _dir === "auto" && !isServer && navigator?.language ? RTL_LOCALES.test(navigator.language) ? "rtl" : "ltr" : _dir === "rtl" ? "rtl" : "ltr", "f", _M3eDirectionality_current);
}, _M3eDirectionality_handleChange = function _M3eDirectionality_handleChange() {
  __classPrivateFieldGet(this, _a, "m", _M3eDirectionality_updateDirection).call(this);
  __classPrivateFieldGet(this, _a, "f", _M3eDirectionality_observers).forEach(x => x());
};
/** @private */
_M3eDirectionality_current = {
  value: "ltr"
};
/** @private */
_M3eDirectionality_observers = {
  value: Array()
};
(() => {
  if (!isServer) {
    __classPrivateFieldGet(_a, _a, "m", _M3eDirectionality_updateDirection).call(_a);
    if (MutationObserver) {
      const init = {
        attributeFilter: ["dir"]
      };
      const observer = new MutationObserver(() => __classPrivateFieldGet(_a, _a, "m", _M3eDirectionality_handleChange).call(_a));
      if (document.body) {
        observer.observe(document.body, init);
      }
      if (document.documentElement) {
        observer.observe(document.documentElement, init);
      }
    }
    window.addEventListener("languagechange", () => __classPrivateFieldGet(_a, _a, "m", _M3eDirectionality_handleChange).call(_a));
  }
})();
globalThis.M3eDirectionality = M3eDirectionality;

export { M3eDirectionality };
