// @m3e/web/dist/core-a11y.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { isServer, html, css, LitElement } from 'lit';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/** The modifier keys. */
const ModifierKeys = ["alt", "shift", "ctrl", "meta"];

/** Specifies known key code values. */
var KeyCode;
(function (KeyCode) {
  KeyCode[KeyCode["Backspace"] = 8] = "Backspace";
  KeyCode[KeyCode["Tab"] = 9] = "Tab";
  KeyCode[KeyCode["Enter"] = 13] = "Enter";
  KeyCode[KeyCode["Shift"] = 16] = "Shift";
  KeyCode[KeyCode["Ctrl"] = 17] = "Ctrl";
  KeyCode[KeyCode["Alt"] = 18] = "Alt";
  KeyCode[KeyCode["PauseBreak"] = 19] = "PauseBreak";
  KeyCode[KeyCode["CapsLock"] = 20] = "CapsLock";
  KeyCode[KeyCode["Escape"] = 27] = "Escape";
  KeyCode[KeyCode["Space"] = 32] = "Space";
  KeyCode[KeyCode["PageUp"] = 33] = "PageUp";
  KeyCode[KeyCode["PageDown"] = 34] = "PageDown";
  KeyCode[KeyCode["End"] = 35] = "End";
  KeyCode[KeyCode["Home"] = 36] = "Home";
  KeyCode[KeyCode["LeftArrow"] = 37] = "LeftArrow";
  KeyCode[KeyCode["UpArrow"] = 38] = "UpArrow";
  KeyCode[KeyCode["RightArrow"] = 39] = "RightArrow";
  KeyCode[KeyCode["DownArrow"] = 40] = "DownArrow";
  KeyCode[KeyCode["Insert"] = 45] = "Insert";
  KeyCode[KeyCode["Delete"] = 46] = "Delete";
  KeyCode[KeyCode["Zero"] = 48] = "Zero";
  KeyCode[KeyCode["ClosedParen"] = 48] = "ClosedParen";
  KeyCode[KeyCode["One"] = 49] = "One";
  KeyCode[KeyCode["ExclamationMark"] = 49] = "ExclamationMark";
  KeyCode[KeyCode["Two"] = 50] = "Two";
  KeyCode[KeyCode["AtSign"] = 50] = "AtSign";
  KeyCode[KeyCode["Three"] = 51] = "Three";
  KeyCode[KeyCode["PoundSign"] = 51] = "PoundSign";
  KeyCode[KeyCode["Hash"] = 51] = "Hash";
  KeyCode[KeyCode["Four"] = 52] = "Four";
  KeyCode[KeyCode["DollarSign"] = 52] = "DollarSign";
  KeyCode[KeyCode["Five"] = 53] = "Five";
  KeyCode[KeyCode["PercentSign"] = 53] = "PercentSign";
  KeyCode[KeyCode["Six"] = 54] = "Six";
  KeyCode[KeyCode["Caret"] = 54] = "Caret";
  KeyCode[KeyCode["Hat"] = 54] = "Hat";
  KeyCode[KeyCode["Seven"] = 55] = "Seven";
  KeyCode[KeyCode["Ampersand"] = 55] = "Ampersand";
  KeyCode[KeyCode["Eight"] = 56] = "Eight";
  KeyCode[KeyCode["Star"] = 56] = "Star";
  KeyCode[KeyCode["Asterik"] = 56] = "Asterik";
  KeyCode[KeyCode["Nine"] = 57] = "Nine";
  KeyCode[KeyCode["OpenParen"] = 57] = "OpenParen";
  KeyCode[KeyCode["A"] = 65] = "A";
  KeyCode[KeyCode["B"] = 66] = "B";
  KeyCode[KeyCode["C"] = 67] = "C";
  KeyCode[KeyCode["D"] = 68] = "D";
  KeyCode[KeyCode["E"] = 69] = "E";
  KeyCode[KeyCode["F"] = 70] = "F";
  KeyCode[KeyCode["G"] = 71] = "G";
  KeyCode[KeyCode["H"] = 72] = "H";
  KeyCode[KeyCode["I"] = 73] = "I";
  KeyCode[KeyCode["J"] = 74] = "J";
  KeyCode[KeyCode["K"] = 75] = "K";
  KeyCode[KeyCode["L"] = 76] = "L";
  KeyCode[KeyCode["M"] = 77] = "M";
  KeyCode[KeyCode["N"] = 78] = "N";
  KeyCode[KeyCode["O"] = 79] = "O";
  KeyCode[KeyCode["P"] = 80] = "P";
  KeyCode[KeyCode["Q"] = 81] = "Q";
  KeyCode[KeyCode["R"] = 82] = "R";
  KeyCode[KeyCode["S"] = 83] = "S";
  KeyCode[KeyCode["T"] = 84] = "T";
  KeyCode[KeyCode["U"] = 85] = "U";
  KeyCode[KeyCode["V"] = 86] = "V";
  KeyCode[KeyCode["W"] = 87] = "W";
  KeyCode[KeyCode["X"] = 88] = "X";
  KeyCode[KeyCode["Y"] = 89] = "Y";
  KeyCode[KeyCode["Z"] = 90] = "Z";
  KeyCode[KeyCode["LeftWindowKey"] = 91] = "LeftWindowKey";
  KeyCode[KeyCode["RightWindowKey"] = 92] = "RightWindowKey";
  KeyCode[KeyCode["SelectKey"] = 93] = "SelectKey";
  KeyCode[KeyCode["Numpad0"] = 96] = "Numpad0";
  KeyCode[KeyCode["Numpad1"] = 97] = "Numpad1";
  KeyCode[KeyCode["Numpad2"] = 98] = "Numpad2";
  KeyCode[KeyCode["Numpad3"] = 99] = "Numpad3";
  KeyCode[KeyCode["Numpad4"] = 100] = "Numpad4";
  KeyCode[KeyCode["Numpad5"] = 101] = "Numpad5";
  KeyCode[KeyCode["Numpad6"] = 102] = "Numpad6";
  KeyCode[KeyCode["Numpad7"] = 103] = "Numpad7";
  KeyCode[KeyCode["Numpad8"] = 104] = "Numpad8";
  KeyCode[KeyCode["Numpad9"] = 105] = "Numpad9";
  KeyCode[KeyCode["Multiply"] = 106] = "Multiply";
  KeyCode[KeyCode["Add"] = 107] = "Add";
  KeyCode[KeyCode["Subtract"] = 109] = "Subtract";
  KeyCode[KeyCode["DecimalPoint"] = 110] = "DecimalPoint";
  KeyCode[KeyCode["Divide"] = 111] = "Divide";
  KeyCode[KeyCode["F1"] = 112] = "F1";
  KeyCode[KeyCode["F2"] = 113] = "F2";
  KeyCode[KeyCode["F3"] = 114] = "F3";
  KeyCode[KeyCode["F4"] = 115] = "F4";
  KeyCode[KeyCode["F5"] = 116] = "F5";
  KeyCode[KeyCode["F6"] = 117] = "F6";
  KeyCode[KeyCode["F7"] = 118] = "F7";
  KeyCode[KeyCode["F8"] = 119] = "F8";
  KeyCode[KeyCode["F9"] = 120] = "F9";
  KeyCode[KeyCode["F10"] = 121] = "F10";
  KeyCode[KeyCode["F11"] = 122] = "F11";
  KeyCode[KeyCode["F12"] = 123] = "F12";
  KeyCode[KeyCode["NumLock"] = 144] = "NumLock";
  KeyCode[KeyCode["ScrollLock"] = 145] = "ScrollLock";
  KeyCode[KeyCode["SemiColon"] = 186] = "SemiColon";
  KeyCode[KeyCode["Equals"] = 187] = "Equals";
  KeyCode[KeyCode["Comma"] = 188] = "Comma";
  KeyCode[KeyCode["Dash"] = 189] = "Dash";
  KeyCode[KeyCode["Period"] = 190] = "Period";
  KeyCode[KeyCode["UnderScore"] = 189] = "UnderScore";
  KeyCode[KeyCode["PlusSign"] = 187] = "PlusSign";
  KeyCode[KeyCode["ForwardSlash"] = 191] = "ForwardSlash";
  KeyCode[KeyCode["Tilde"] = 192] = "Tilde";
  KeyCode[KeyCode["GraveAccent"] = 192] = "GraveAccent";
  KeyCode[KeyCode["OpenBracket"] = 219] = "OpenBracket";
  KeyCode[KeyCode["ClosedBracket"] = 221] = "ClosedBracket";
  KeyCode[KeyCode["Quote"] = 222] = "Quote";
})(KeyCode || (KeyCode = {}));

/**
 * Resolves the key code for the specified `KeyboardEvent`.
 * @param {KeyboardEvent} e The `KeyboardEvent` for which to resolve the key code.
 * @returns {number} The key code for `e`.
 */
function getKeyCode(e) {
  return e.which || e.charCode || e.keyCode;
}

/**
 * Determines whether a modifier key is pressed.
 * @param {KeyboardEvent} e The `KeyboardEvent` to test.
 * @param {ModifierKey[]} modifiers The modifier keys to test.
 * @returns {boolean} A value indicating whether a modifier key is pressed.
 */
function hasModifierKey(e, ...modifiers) {
  return modifiers.length ? modifiers.some(x => e[`${x}Key`]) : e.altKey || e.shiftKey || e.ctrlKey || e.metaKey;
}

/**
 * Determines whether the specified modifier keys are allowed.
 * @param {KeyboardEvent} e The `KeyboardEvent` to test.
 * @param {ModifierKey[]} modifiers The allowed modifier keys.
 * @returns {boolean} A value indicating whether `modifiers` are allowed.
 */
function isModifierAllowed(e, ...modifiers) {
  return ModifierKeys.every(x => !hasModifierKey(e, x) || modifiers.includes(x));
}

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = globalThis,
  e$3 = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s = Symbol(),
  o$2 = new WeakMap();
let n$2 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$3 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$2.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$2.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$2 = t => new n$2("string" == typeof t ? t : t + "", void 0, s),
  S = (s, o) => {
    if (e$3) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$1 = e$3 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$2(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i,
    defineProperty: e$2,
    getOwnPropertyDescriptor: h,
    getOwnPropertyNames: r$1,
    getOwnPropertySymbols: o$1,
    getPrototypeOf: n$1
  } = Object,
  a = globalThis,
  c = a.trustedTypes,
  l = c ? c.emptyScript : "",
  p = a.reactiveElementPolyfillSupport,
  d = (t, s) => t,
  u = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f = (t, s) => !i(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    useDefault: false,
    hasChanged: f
  };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e$2(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties"))) return;
    const t = n$1(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t = this.properties,
        s = [...r$1(t), ...o$1(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$1(s));
    } else void 0 !== s && i.push(c$1(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d("elementProperties")] = new Map(), y[d("finalized")] = new Map(), p?.({
  ReactiveElement: y
}), (a.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    hasChanged: f
  },
  r = (t = o, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n(t) {
  return (e, o) => "object" == typeof o ? r(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const e$1 = (e, t, c) => (c.configurable = true, c.enumerable = true, Reflect.decorate && "object" != typeof t && Object.defineProperty(e, t, c), c);

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
function e(e, r) {
  return (n, s, i) => {
    const o = t => t.renderRoot?.querySelector(e) ?? null;
    return e$1(n, s, {
      get() {
        return o(this);
      }
    });
  };
}

/**
 * Determines whether an object has keys for a given type.
 * @template T The type to test.
 * @param {unknown} value The value to test.
 * @param {...keys: Array<keyof T>} keys The keys of `T` to test.
 * @returns {boolean} Whether `value` has all `keys`.
 */
function hasKeys(value, ...keys) {
  return typeof value === "object" && value !== null && keys.every(x => x in value);
}

/**
 * Determines whether a value is a `DisabledMixin`.
 * @param {unknown} value The value to test.
 * @returns {value is DisabledMixin} Whether `value` is a `DisabledMixin`.
 */
function isDisabledMixin(value) {
  return hasKeys(value, "disabled");
}
/**
 * Mixin to augment an element with behavior that supports a disabled state.
 * @template T The type of the base class.
 * @param {T} base The base class.
 * @param {boolean} [reflect=true] Whether the disabled property is reflected as an attribute.
 * @returns {Constructor<DisabledMixin> & T} A constructor that implements `DisabledMixin`.
 */
function Disabled(base, reflect = true) {
  class _DisabledMixin extends base {
    constructor() {
      super(...arguments);
      /**
       * Whether the element is disabled.
       * @default false
       */
      this.disabled = false;
    }
    /** @inheritdoc */
    update(changedProperties) {
      super.update(changedProperties);
      if (changedProperties.has("disabled") && this.role && this.role !== "none" && this.role !== "presentation") {
        this.ariaDisabled = this.disabled ? "true" : null;
      }
    }
  }
  __decorate([n({
    type: Boolean,
    reflect: reflect
  })], _DisabledMixin.prototype, "disabled", void 0);
  return _DisabledMixin;
}

var _ListManager_items, _ListManager_activeItem, _ListManager_onActiveItemChangeCallback;
/**
 * Utility for managing a list of items which supports activation.
 * @template T The type of managed item.
 */
class ListManager {
  constructor() {
    /** @private */_ListManager_items.set(this, new Array());
    /** @private */
    _ListManager_activeItem.set(this, null);
    /** @private */
    _ListManager_onActiveItemChangeCallback.set(this, void 0);
  }
  /** The items being managed. */
  get items() {
    return __classPrivateFieldGet(this, _ListManager_items, "f");
  }
  /** The active item. */
  get activeItem() {
    return __classPrivateFieldGet(this, _ListManager_activeItem, "f");
  }
  /**
   * Sets the items to manage.
   * @param {Array<T>} items The new items.
   * @returns {{ added: ReadonlyArray<T>; removed: ReadonlyArray<T> }} An object specifying added and removed items.
   */
  setItems(items) {
    const removed = this.items.filter(x => !items.includes(x));
    const added = items.filter(x => !this.items.includes(x));
    __classPrivateFieldSet(this, _ListManager_items, items, "f");
    if (this.activeItem && !this.items.includes(this.activeItem)) {
      this.updateActiveItem(null);
    }
    return {
      added,
      removed
    };
  }
  /**
   * Sets the active item.
   * @param {T | null | undefined} item The new active item.
   */
  setActiveItem(item) {
    if (this.activeItem !== item) {
      this.updateActiveItem(item);
      __classPrivateFieldGet(this, _ListManager_onActiveItemChangeCallback, "f")?.call(this);
    }
  }
  /**
   * Updates the active item.
   * @param {T | null | undefined} item The active item.
   */
  updateActiveItem(item) {
    __classPrivateFieldSet(this, _ListManager_activeItem, item ?? null, "f");
  }
  /**
   * Configures the list manager with a callback invoked when an item is activated.
   * @param {() => void} callback The callback invoked when an item is activated.
   * @returns {ListManager<T>} The configured list manager.
   */
  onActiveItemChange(callback) {
    __classPrivateFieldSet(this, _ListManager_onActiveItemChangeCallback, callback, "f");
    return this;
  }
}
_ListManager_items = new WeakMap(), _ListManager_activeItem = new WeakMap(), _ListManager_onActiveItemChangeCallback = new WeakMap();

/**
 * Adapted from Angular Material CDK KeyManager
 * Source: https://github.com/angular/components/blob/main/src/cdk/a11y/key-manager/typeahead.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _Typeahead_instances, _Typeahead_debounceInterval, _Typeahead_timeoutId, _Typeahead_callback, _Typeahead_skipPredicate, _Typeahead_pressedKeys, _Typeahead_selectedIndex, _Typeahead_items, _Typeahead_appendKey, _Typeahead_searchItems;
/** A symbol through which to access an element's textual content used for typeahead search. */
const typeaheadLabel = Symbol("typeaheadLabel");
/**
 * Determines whether a value is a `TypeaheadItem`.
 * @param {unknown} value The value to test.
 * @returns A value indicating whether `value` is an `TypeaheadItem`.
 */
function isTypeaheadItem(value) {
  return hasKeys(value, typeaheadLabel);
}
/**
 * Implements typeahead functionality which selects items based on keyboard input.
 * @template T The type of `TypeaheadItem`.
 */
class Typeahead {
  /**
   * Initializes a new instance of this class.
   * @param {TypeaheadOptions<T>} options Options that control typeahead behavior.
   */
  constructor(options) {
    _Typeahead_instances.add(this);
    /** @private */
    _Typeahead_debounceInterval.set(this, void 0);
    /** @private */
    _Typeahead_timeoutId.set(this, -1);
    /** @private */
    _Typeahead_callback.set(this, void 0);
    /** @private */
    _Typeahead_skipPredicate.set(this, void 0);
    /** @private */
    _Typeahead_pressedKeys.set(this, new Array());
    /** @private */
    _Typeahead_selectedIndex.set(this, -1);
    /** @private */
    _Typeahead_items.set(this, []);
    __classPrivateFieldSet(this, _Typeahead_debounceInterval, options.debounceInterval ?? 200, "f");
    __classPrivateFieldSet(this, _Typeahead_callback, options.callback, "f");
    __classPrivateFieldSet(this, _Typeahead_skipPredicate, options.skipPredicate, "f");
  }
  /** A value indicating whether the user is currently typing. */
  get isTyping() {
    return __classPrivateFieldGet(this, _Typeahead_pressedKeys, "f").length > 0;
  }
  /**
   * Sets the items to search.
   * @param {readonly T[]} items The items to search.
   */
  setItems(items) {
    __classPrivateFieldSet(this, _Typeahead_items, items, "f");
  }
  /**
   * Sets the index of the selected item.
   * @param {number} index The index of the selected item.
   */
  setSelectedIndex(index) {
    __classPrivateFieldSet(this, _Typeahead_selectedIndex, index, "f");
  }
  /** Resets the stored sequence of typed characters. */
  reset() {
    __classPrivateFieldGet(this, _Typeahead_pressedKeys, "f").length = 0;
  }
  /**
   * Sets the selected item depending on the key event passed in.
   * @param {KeyboardEvent} e The keyboard event to be used for determining which element should be active.
   */
  onKeyDown(e) {
    if (e.key && e.key.length === 1) {
      __classPrivateFieldGet(this, _Typeahead_instances, "m", _Typeahead_appendKey).call(this, e.key);
    } else {
      const keycode = getKeyCode(e);
      if (keycode >= KeyCode.A && keycode <= KeyCode.Z || keycode >= KeyCode.Zero && keycode <= KeyCode.Nine) {
        __classPrivateFieldGet(this, _Typeahead_instances, "m", _Typeahead_appendKey).call(this, String.fromCharCode(keycode));
      }
    }
  }
}
_Typeahead_debounceInterval = new WeakMap(), _Typeahead_timeoutId = new WeakMap(), _Typeahead_callback = new WeakMap(), _Typeahead_skipPredicate = new WeakMap(), _Typeahead_pressedKeys = new WeakMap(), _Typeahead_selectedIndex = new WeakMap(), _Typeahead_items = new WeakMap(), _Typeahead_instances = new WeakSet(), _Typeahead_appendKey = function _Typeahead_appendKey(key) {
  __classPrivateFieldGet(this, _Typeahead_pressedKeys, "f").push(key.toLocaleUpperCase());
  clearTimeout(__classPrivateFieldGet(this, _Typeahead_timeoutId, "f"));
  __classPrivateFieldSet(this, _Typeahead_timeoutId, setTimeout(() => __classPrivateFieldGet(this, _Typeahead_instances, "m", _Typeahead_searchItems).call(this), __classPrivateFieldGet(this, _Typeahead_debounceInterval, "f")), "f");
}, _Typeahead_searchItems = function _Typeahead_searchItems() {
  const term = __classPrivateFieldGet(this, _Typeahead_pressedKeys, "f").join("");
  for (let i = 1; i < __classPrivateFieldGet(this, _Typeahead_items, "f").length + 1; i++) {
    const index = (__classPrivateFieldGet(this, _Typeahead_selectedIndex, "f") + i) % __classPrivateFieldGet(this, _Typeahead_items, "f").length;
    const item = __classPrivateFieldGet(this, _Typeahead_items, "f")[index];
    const label = item[typeaheadLabel]?.().toLocaleUpperCase().trim();
    if (!__classPrivateFieldGet(this, _Typeahead_skipPredicate, "f")?.call(this, item) && label?.indexOf(term) === 0) {
      __classPrivateFieldGet(this, _Typeahead_callback, "f").call(this, item);
      break;
    }
  }
  this.reset();
};

/**
 * Adapted from Angular Material CDK KeyManager
 * Source: https://github.com/angular/components/blob/main/src/cdk/a11y/key-manager/list-key-manager.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _ListKeyManager_instances, _ListKeyManager_typeahead, _ListKeyManager_findFirst, _ListKeyManager_findLast, _ListKeyManager_findNext, _ListKeyManager_findPrevious, _ListKeyManager_findPreviousByIndex, _ListKeyManager_findNextByIndex;
/**
 * Utility for managing keyboard events for selectable lists.
 * @template T The type of managed item.
 */
class ListKeyManager extends ListManager {
  constructor() {
    super(...arguments);
    _ListKeyManager_instances.add(this);
    /** @private */
    _ListKeyManager_typeahead.set(this, void 0);
    /**
     * Whether the active item will wrap to the other end of
     * list when there are no more items in the given direction.
     * @default false
     */
    this.wrap = false;
    /**
     * Whether to activate the first and last items respectively when
     * the `HOME` or `END` key is pressed.
     * @default false
     */
    this.homeAndEnd = false;
    /**
     * Whether to activate every 10th (or configured) first/last item
     * in the up/down direction when the `PAGEUP` or `PAGEDOWN` key is pressed.
     * @default false
     */
    this.pageUpAndDown = false;
    /**
     * The number of items to skip when the `PAGEUP` or `PAGEDOWN` key is pressed.
     * @default 10
     */
    this.pageDelta = 10;
    /**
     * Whether to the list is oriented vertically.
     * @default false
     */
    this.vertical = false;
    /** The allowed modifier keys.
     * @default []
     */
    this.allowedModifiers = [];
    /**
     * A function used to skip items.
     * @param {T} item The item to test.
     * @returns {boolean} Whether `item` should be skipped.
     */
    this.skipPredicate = item => isDisabledMixin(item) && item.disabled;
    /**
     * The directionality.
     * @default "ltr"
     */
    this.directionality = "ltr";
  }
  /** @inheritdoc */
  setItems(items) {
    __classPrivateFieldGet(this, _ListKeyManager_typeahead, "f")?.setItems(items);
    return super.setItems(items);
  }
  /** @inheritdoc */
  updateActiveItem(item) {
    super.updateActiveItem(item);
    if (__classPrivateFieldGet(this, _ListKeyManager_typeahead, "f")) {
      __classPrivateFieldGet(this, _ListKeyManager_typeahead, "f").setSelectedIndex(item ? this.items.indexOf(item) : -1);
    }
  }
  /**
   * Configures the key manager to activate the first and last items respectively when the `HOME` or `END` key is pressed.
   * @param {boolean} [enabled = true] Whether to activate the first and last items respectively when
   * the `HOME` or `END` key is pressed.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withHomeAndEnd(enabled = true) {
    this.homeAndEnd = enabled;
    return this;
  }
  /**
   * Configures the key manager to page up and down when the `PAGEUP` or `PAGEDOWN` key is pressed.
   * @param {boolean} [enabled = true] Whether to activate page up and down when the `PAGEUP` or `PAGEDOWN` key is pressed.
   * @param {number} [pageDelta=10] The number of items to skip when the `PAGEUP` or `PAGEDOWN` key is pressed.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withPageUpAndDown(enabled = true, pageDelta = 10) {
    this.pageUpAndDown = enabled;
    this.pageDelta = pageDelta;
    return this;
  }
  /**
   * Configures wrapping mode, which determines whether the active item will wrap to the other end of list when there are no more items in the given direction.
   * @param {boolean} [enabled = true] Whether the active item will wrap to the other end of
   * list when there are no more items in the given direction.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withWrap(enabled = true) {
    this.wrap = enabled;
    return this;
  }
  /**
   * Configures whether to move the selection vertically.
   * @param {boolean} [enabled = true] Whether to move selection vertically.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withVerticalOrientation(enabled = true) {
    this.vertical = enabled;
    return this;
  }
  /**
   * Configured allowed modifier keys.
   * @param {ModifierKey[]} modifiers The allowed modifier keys.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withAllowedModifiers(...modifiers) {
    this.allowedModifiers = modifiers;
    return this;
  }
  /**
   * Configures whether typeahead is enabled.
   * @param {boolean} [enabled = true] Whether typeahead is enabled.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withTypeahead(enabled = true) {
    if (enabled) {
      __classPrivateFieldSet(this, _ListKeyManager_typeahead, new Typeahead({
        callback: item => this.setActiveItem(item)
      }), "f");
    } else {
      __classPrivateFieldSet(this, _ListKeyManager_typeahead, undefined, "f");
    }
    return this;
  }
  /**
   * Configures a function used to test whether an item should be skipped.
   * @param skipPredicate A function used to determine whether an item should be skipped.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withSkipPredicate(skipPredicate) {
    this.skipPredicate = skipPredicate;
    return this;
  }
  /**
   * Configures the directionality.
   * @param {"ltr" | "rtl"} directionality The directionality.
   * @returns {ListKeyManager<T>} The configured key manager.
   */
  withDirectionality(directionality) {
    this.directionality = directionality;
    return this;
  }
  /**
   * Sets the active item depending on the key event passed in.
   * @param {KeyboardEvent} e The keyboard event to be used for determining which element should be active.
   */
  onKeyDown(e) {
    if (e.defaultPrevented) return;
    const modifierAllowed = isModifierAllowed(e, ...this.allowedModifiers);
    switch (e.key) {
      case "Left":
      case "ArrowLeft":
        if (modifierAllowed && !this.vertical) {
          e.preventDefault();
          const prev = this.directionality === "ltr" ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findPrevious).call(this) : __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findNext).call(this);
          if (prev) {
            this.setActiveItem(prev);
          }
        }
        break;
      case "Up":
      case "ArrowUp":
        if (modifierAllowed) {
          e.preventDefault();
          const prev = this.directionality === "ltr" ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findPrevious).call(this) : __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findNext).call(this);
          if (prev) {
            this.setActiveItem(prev);
          }
        }
        break;
      case "Right":
      case "ArrowRight":
        if (modifierAllowed && !this.vertical) {
          e.preventDefault();
          const next = this.directionality === "ltr" ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findNext).call(this) : __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findPrevious).call(this);
          if (next) {
            this.setActiveItem(next);
          }
        }
        break;
      case "Down":
      case "ArrowDown":
        if (modifierAllowed) {
          e.preventDefault();
          const next = this.directionality === "ltr" ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findNext).call(this) : __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findPrevious).call(this);
          if (next) {
            this.setActiveItem(next);
          }
        }
        break;
      case "Home":
        if (modifierAllowed && this.homeAndEnd) {
          e.preventDefault();
          const first = __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findFirst).call(this);
          if (first) {
            this.setActiveItem(first);
          }
        }
        break;
      case "End":
        if (modifierAllowed && this.homeAndEnd) {
          e.preventDefault();
          const last = __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findLast).call(this);
          if (last) {
            this.setActiveItem(last);
          }
        }
        break;
      case "PageUp":
        if (modifierAllowed && this.pageUpAndDown) {
          e.preventDefault();
          const prev = (this.directionality === "ltr" ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findPreviousByIndex) : __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findNextByIndex))(this.activeItem ? Math.max(0, this.items.indexOf(this.activeItem) - this.pageDelta) : 0);
          if (prev) {
            this.setActiveItem(prev);
          }
        }
        break;
      case "PageDown":
        if (modifierAllowed && this.pageUpAndDown) {
          e.preventDefault();
          const next = (this.directionality === "ltr" ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findNextByIndex) : __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findPreviousByIndex))(this.activeItem ? Math.min(this.items.length - 1, this.items.indexOf(this.activeItem) + this.pageDelta) : this.items.length - 1);
          if (next) {
            this.setActiveItem(next);
          }
        }
        break;
      default:
        if (modifierAllowed || hasModifierKey(e, "shift")) {
          __classPrivateFieldGet(this, _ListKeyManager_typeahead, "f")?.onKeyDown(e);
        }
        break;
    }
  }
}
_ListKeyManager_typeahead = new WeakMap(), _ListKeyManager_instances = new WeakSet(), _ListKeyManager_findFirst = function _ListKeyManager_findFirst() {
  for (let i = 0; i < this.items.length; i++) {
    if (this.items[i] === this.activeItem) break;
    const test = this.items[i];
    if (!this.skipPredicate(test)) {
      return test;
    }
  }
  return null;
}, _ListKeyManager_findLast = function _ListKeyManager_findLast() {
  for (let i = this.items.length - 1; i >= 0; i--) {
    if (this.items[i] === this.activeItem) break;
    const test = this.items[i];
    if (!this.skipPredicate(test)) {
      return test;
    }
  }
  return null;
}, _ListKeyManager_findNext = function _ListKeyManager_findNext() {
  for (let i = 0; i < this.items.length; i++) {
    if (this.items[i] === this.activeItem) {
      for (let j = i + 1; j < this.items.length; j++) {
        const test = this.items[j];
        if (!this.skipPredicate(test)) {
          return test;
        }
      }
      break;
    }
  }
  return this.wrap ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findFirst).call(this) : null;
}, _ListKeyManager_findPrevious = function _ListKeyManager_findPrevious() {
  for (let i = 0; i < this.items.length; i++) {
    if (this.items[i] === this.activeItem) {
      for (let j = i - 1; j >= 0; j--) {
        const test = this.items[j];
        if (!this.skipPredicate(test)) {
          return test;
        }
      }
      break;
    }
  }
  return this.wrap ? __classPrivateFieldGet(this, _ListKeyManager_instances, "m", _ListKeyManager_findLast).call(this) : null;
}, _ListKeyManager_findPreviousByIndex = function _ListKeyManager_findPreviousByIndex(index) {
  for (let i = index; i >= 0; i--) {
    const test = this.items[i];
    if (!this.skipPredicate(test)) {
      return test;
    }
  }
  return null;
}, _ListKeyManager_findNextByIndex = function _ListKeyManager_findNextByIndex(index) {
  for (let i = index; i < this.items.length; i++) {
    const test = this.items[i];
    if (!this.skipPredicate(test)) {
      return test;
    }
  }
  return null;
};

var _FocusKeyManager_focusOptions;
/**
 * Utility for managing keyboard events for selectable lists whose items directly receive focus.
 * @template T The type of managed item.
 */
class FocusKeyManager extends ListKeyManager {
  constructor() {
    super(...arguments);
    /** @private */
    _FocusKeyManager_focusOptions.set(this, void 0);
  }
  /** @inheritdoc */
  setActiveItem(item) {
    super.setActiveItem(item);
    item?.focus(__classPrivateFieldGet(this, _FocusKeyManager_focusOptions, "f"));
  }
  /**
   * Configures the key manager with options used to focus items.
   * @param {FocusOptions} options Options used to focus items.
   * @returns {FocusKeyManager<T>} The configured key manager.
   */
  withOptions(options) {
    __classPrivateFieldSet(this, _FocusKeyManager_focusOptions, options, "f");
    return this;
  }
}
_FocusKeyManager_focusOptions = new WeakMap();

/**
 * Determines whether a value is a `CheckedMixin`.
 * @param {unknown} value The value to test.
 * @returns Whether `value` is a `CheckedMixin`.
 */
function isCheckedMixin(value) {
  return hasKeys(value, "checked");
}

/**
 * Determines whether a value is a `SelectedMixin`.
 * @param {unknown} value The value to test.
 * @returns Whether `value` is a `SelectedMixin`.
 */
function isSelectedMixin(value) {
  return hasKeys(value, "selected");
}

/**
 * Determines whether the state of an element is checked or selected.
 * @param {CheckedOrSelectedMixin} element The element to test.
 * @return {boolean} Whether `element` is checked or selected.
 */
function isCheckedOrSelected(element) {
  return isCheckedMixin(element) && element.checked || isSelectedMixin(element) && element.selected;
}
/**
 * Sets the checked or selected state of an element.
 * @param {CheckedOrSelectedMixin} element The element for which to set the checked or selected state.
 * @param {boolean} checkedOrSelected The checked or selected state.
 */
function checkOrSelect(element, checkedOrSelected) {
  if (isCheckedMixin(element)) {
    element.checked = checkedOrSelected;
  } else {
    element.selected = checkedOrSelected;
  }
}

var _RovingTabIndexManager_disableRovingTabIndex;
/**
 * Utility for managing keyboard events and a roving tab index for selectable lists whose items directly receive focus.
 * @template T The type of managed item.
 */
class RovingTabIndexManager extends FocusKeyManager {
  constructor() {
    super(...arguments);
    /** @private */
    _RovingTabIndexManager_disableRovingTabIndex.set(this, false);
  }
  /** @inheritdoc */
  updateActiveItem(item) {
    super.updateActiveItem(item);
    if (!__classPrivateFieldGet(this, _RovingTabIndexManager_disableRovingTabIndex, "f")) {
      item?.setAttribute("tabindex", "0");
      for (const other of this.items) {
        if (other !== item && other.hasAttribute("tabindex")) {
          other.setAttribute("tabindex", "-1");
        }
      }
    }
  }
  /** @inheritdoc */
  setItems(items) {
    const result = super.setItems(items);
    if (!__classPrivateFieldGet(this, _RovingTabIndexManager_disableRovingTabIndex, "f")) {
      for (const added of result.added) {
        if (added !== this.activeItem && !this.skipPredicate(added)) {
          added.setAttribute("tabindex", "-1");
        }
      }
    }
    return result;
  }
  /**
   * Configures whether roving tab index is disabled.
   * @param {boolean} [disabled=true] Whether the roving tab index is disabled.
   * @returns {RovingTabIndexManager<T>} The configured roving tab index manager.
   */
  disableRovingTabIndex(disabled = true) {
    if (disabled !== __classPrivateFieldGet(this, _RovingTabIndexManager_disableRovingTabIndex, "f")) {
      __classPrivateFieldSet(this, _RovingTabIndexManager_disableRovingTabIndex, disabled, "f");
      for (const item of this.items) {
        if (!this.skipPredicate(item)) {
          item?.setAttribute("tabindex", __classPrivateFieldGet(this, _RovingTabIndexManager_disableRovingTabIndex, "f") || item === this.activeItem ? "0" : "-1");
        }
      }
    }
    return this;
  }
}
_RovingTabIndexManager_disableRovingTabIndex = new WeakMap();

var _RadioKeyManager_disabled;
/**
 * Utility for managing keyboard events for selectable lists whose items behave like a radio.
 * @template T The type of managed item.
 */
class RadioKeyManager extends RovingTabIndexManager {
  constructor() {
    super(...arguments);
    /** @private */
    _RadioKeyManager_disabled.set(this, false);
  }
  /** A value indicating whether managed items are disabled. */
  get disabled() {
    return __classPrivateFieldGet(this, _RadioKeyManager_disabled, "f");
  }
  set disabled(value) {
    __classPrivateFieldSet(this, _RadioKeyManager_disabled, value, "f");
    this.items.forEach(x => x.disabled = value);
  }
  /** @inheritdoc */
  setItems(items) {
    if (this.disabled) {
      items.forEach(x => x.disabled = true);
    }
    const {
      added,
      removed
    } = super.setItems(items);
    if (added.length > 0 || removed.length > 0) {
      if (!this.activeItem) {
        this.updateActiveItem(added.find(x => !this.skipPredicate(x)) ?? null);
      }
      if (this.activeItem && (this.activeItem.disabled || !isCheckedOrSelected(this.activeItem))) {
        const checked = added.find(x => !this.skipPredicate(x) && isCheckedOrSelected(x));
        if (checked) {
          this.updateActiveItem(checked);
        }
      }
    }
    return {
      added,
      removed
    };
  }
}
_RadioKeyManager_disabled = new WeakMap();

var _SelectionManager_instances, _SelectionManager_onSelectedItemsChangeCallback, _SelectionManager_selectedItems, _SelectionManager_multi, _SelectionManager_enforceSingleSelect;
/** A symbol through which to access an element's selection manager. */
const selectionManager = Symbol("selectionManager");
/**
 * Utility for managing keyboard events for selectable lists where one or more items can be selected.
 * @template T The type of managed item.
 */
class SelectionManager extends RadioKeyManager {
  constructor() {
    super(...arguments);
    _SelectionManager_instances.add(this);
    /** @private */
    _SelectionManager_onSelectedItemsChangeCallback.set(this, void 0);
    /** @private */
    _SelectionManager_selectedItems.set(this, new Array());
    /** @private */
    _SelectionManager_multi.set(this, false);
  }
  /** A value indicating whether multiple items can be selected. */
  get multi() {
    return __classPrivateFieldGet(this, _SelectionManager_multi, "f");
  }
  set multi(value) {
    __classPrivateFieldSet(this, _SelectionManager_multi, value, "f");
    __classPrivateFieldGet(this, _SelectionManager_instances, "m", _SelectionManager_enforceSingleSelect).call(this, true);
  }
  /** The selected items. */
  get selectedItems() {
    return __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f");
  }
  /**
   * Selects or deselects the item based on the item's checked or selected state.
   * @param {T} item The item whose selection state has changed.
   */
  notifySelectionChange(item) {
    if (this.items.includes(item)) {
      if (isCheckedOrSelected(item)) {
        this.select(item);
      } else {
        this.deselect(item);
      }
    }
  }
  /**
   * Deselects the specified item.
   * @param {T} item The item to deselect.
   */
  deselect(item) {
    if (this.items.includes(item)) {
      if (isCheckedOrSelected(item)) {
        checkOrSelect(item, false);
      }
      const index = __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").indexOf(item);
      if (index >= 0) {
        __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").splice(index, 1);
        __classPrivateFieldGet(this, _SelectionManager_onSelectedItemsChangeCallback, "f")?.call(this);
      }
    }
  }
  /**
   * Updates the selected item.
   * @param {T | null | undefined} item The selected item.
   * @param {boolean} [activate=true] A value indicating whether to activate the item.
   */
  select(item, activate = true) {
    if (!item || this.items.includes(item)) {
      if (!this.multi) {
        for (const selected of __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f")) {
          if (selected !== item) {
            checkOrSelect(selected, false);
          }
        }
        __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").length = 0;
      }
      if (item) {
        __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").push(item);
        if (!isCheckedOrSelected(item)) {
          checkOrSelect(item, true);
        }
      }
      if (activate) {
        this.updateActiveItem(item);
      }
      __classPrivateFieldGet(this, _SelectionManager_onSelectedItemsChangeCallback, "f")?.call(this);
    }
  }
  /** @inheritdoc */
  setItems(items) {
    const {
      added,
      removed
    } = super.setItems(items);
    __classPrivateFieldSet(this, _SelectionManager_selectedItems, __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").filter(x => !removed.includes(x)), "f");
    __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").push(...added.filter(x => isCheckedOrSelected(x)));
    __classPrivateFieldGet(this, _SelectionManager_instances, "m", _SelectionManager_enforceSingleSelect).call(this);
    __classPrivateFieldGet(this, _SelectionManager_onSelectedItemsChangeCallback, "f")?.call(this);
    return {
      added,
      removed
    };
  }
  /**
   * Configures the selection manager with a callback invoked when selected items change.
   * @param {() => void} callback The callback invoked when selected items change.
   * @returns {SelectionManager<T>} The configured selection manager.
   */
  onSelectedItemsChange(callback) {
    __classPrivateFieldSet(this, _SelectionManager_onSelectedItemsChangeCallback, callback, "f");
    return this;
  }
}
_SelectionManager_onSelectedItemsChangeCallback = new WeakMap(), _SelectionManager_selectedItems = new WeakMap(), _SelectionManager_multi = new WeakMap(), _SelectionManager_instances = new WeakSet(), _SelectionManager_enforceSingleSelect = function _SelectionManager_enforceSingleSelect(emit = false) {
  if (!this.multi && __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").length > 1) {
    for (let i = 1; i < __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").length; i++) {
      checkOrSelect(__classPrivateFieldGet(this, _SelectionManager_selectedItems, "f")[i], false);
    }
    __classPrivateFieldGet(this, _SelectionManager_selectedItems, "f").length = 1;
    if (emit) {
      __classPrivateFieldGet(this, _SelectionManager_onSelectedItemsChangeCallback, "f")?.call(this);
    }
  }
};

/**
 * Adapted from Angular Material CDK AriaDescriber
 * Source: https://github.com/angular/components/blob/main/src/cdk/a11y/aria-describer/aria-reference.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
/**
 * Gets a list of IDs referenced by an element for the specified ARIA attribute.
 * @param {Element} element The element from which to get referenced IDs.
 * @param {`aria-${string}`} attribute The ARIA attribute from which to get referenced IDs.
 * @returns {string[]} a list of IDs referenced by `element` for `attribute`.
 */
function getAriaReferenceIds(element, attribute) {
  return element.getAttribute(attribute)?.match(/\S+/g) ?? [];
}
/**
 * Adds an ID to the list of IDs referenced by an element for a given ARIA attribute.
 * @param {Element} element The element to which to add a referenced ID.
 * @param {`aria-${string}`} attribute The ARIA attribute to which to add an ID.
 * @param {string} id The ID to add.
 */
function addAriaReferencedId(element, attribute, id) {
  id = id.trim();
  const ids = getAriaReferenceIds(element, attribute);
  if (!ids.some(existingId => existingId.trim() === id)) {
    ids.push(id);
    element.setAttribute(attribute, ids.join(" "));
  }
}
/**
 * Removes an ID from a list of IDs referenced by an element for the specified ARIA attribute.
 * @param {Element} element The element from which to remove an ID.
 * @param {`aria-${string}`} attribute The ARIA attribute from which to remove an ID.
 * @param {string} id The ID to remove.
 */
function removeAriaReferencedId(element, attribute, id) {
  id = id.trim();
  const ids = getAriaReferenceIds(element, attribute).filter(val => val !== id);
  if (ids.length > 0) {
    element.setAttribute(attribute, ids.join(" "));
  } else {
    element.removeAttribute(attribute);
  }
}

/**
 * Visually hides an element.
 * @param {CSSStyleDeclaration} style The style of the element to visually hide.
 */
function visuallyHide(style) {
  style.position = "absolute";
  style.appearance = "none";
  style.visibility = "hidden";
  style.border = "0";
  style.outline = "0";
  style.overflow = "hidden";
  style.left = "0";
  style.height = "1px";
  style.width = "1px";
  style.margin = "-1px";
  style.padding = "0";
  style.whiteSpace = "nowrap";
}

/**
 * Adapted from Angular Material CDK ARIA Describer
 * Source: https://github.com/angular/components/blob/main/src/cdk/a11y/aria-describer/aria-describer.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _a$2, _M3eAriaDescriber_nextId, _M3eAriaDescriber_messageContainers, _M3eAriaDescriber_createMessageContainer, _M3eAriaDescriber_createMessageElement;
/** Utility for generating visually hidden elements that convey descriptive messages using `aria-describedby`. */
class M3eAriaDescriber {
  /**
   * Adds an `aria-describedby` reference to a hidden element that contains the message.
   * @param {Element} element The element for which to provide a visually hidden description.
   * @param {string} message The message used to describe `element`.
   * @param {string} [role="tooltip"] The ARIA role of the message.
   */
  static describe(element, message, role = "tooltip") {
    if (isServer) {
      return;
    }
    const rootNode = element.getRootNode();
    if (!(rootNode instanceof ShadowRoot || rootNode instanceof Document)) {
      return;
    }
    let container = __classPrivateFieldGet(this, _a$2, "f", _M3eAriaDescriber_messageContainers).get(rootNode);
    if (!container) {
      container = {
        containerElement: __classPrivateFieldGet(this, _a$2, "m", _M3eAriaDescriber_createMessageContainer).call(this),
        messageRegistry: new Map()
      };
      // Append to body if document, otherwise, append as a child of shadow root.
      (rootNode instanceof Document ? rootNode.body : rootNode).appendChild(container.containerElement);
      __classPrivateFieldGet(this, _a$2, "f", _M3eAriaDescriber_messageContainers).set(rootNode, container);
    }
    const key = `${role}:${message}`;
    let entry = container.messageRegistry.get(key);
    if (!entry) {
      entry = {
        messageElement: __classPrivateFieldGet(this, _a$2, "m", _M3eAriaDescriber_createMessageElement).call(this, message, role),
        count: 0
      };
      container.containerElement.appendChild(entry.messageElement);
      container.messageRegistry.set(key, entry);
    }
    entry.count++;
    addAriaReferencedId(element, "aria-describedby", entry.messageElement.id);
  }
  /**
   * Removes an `aria-describedby` reference to a hidden element that contains the message.
   * @param {Element} element The element from which to remove a visually hidden description.
   * @param {string} message The message used to describe `element`.
   * @param {string} [role="tooltip"] The ARIA role of the message.
   */
  static removeDescription(element, message, role = "tooltip") {
    if (isServer) {
      return;
    }
    const rootNode = element.getRootNode();
    const container = __classPrivateFieldGet(this, _a$2, "f", _M3eAriaDescriber_messageContainers).get(rootNode);
    if (!container) {
      return;
    }
    const key = `${role}:${message}`;
    const entry = container.messageRegistry.get(key);
    if (!entry) {
      return;
    }
    entry.count--;
    removeAriaReferencedId(element, "aria-describedby", entry.messageElement.id);
    // If there are no more elements referencing the message, remove it from the DOM.
    if (entry.count <= 0) {
      entry.messageElement.remove();
      container.messageRegistry.delete(key);
    }
    // If there are no more registered messages, remove the container from the DOM.
    if (container.messageRegistry.size == 0) {
      container.containerElement?.remove();
      __classPrivateFieldGet(this, _a$2, "f", _M3eAriaDescriber_messageContainers).delete(rootNode);
    }
  }
}
_a$2 = M3eAriaDescriber, _M3eAriaDescriber_createMessageContainer = function _M3eAriaDescriber_createMessageContainer() {
  const messageContainer = document.createElement("div");
  messageContainer.classList.add("m3e-describedby-message-container");
  visuallyHide(messageContainer.style);
  return messageContainer;
}, _M3eAriaDescriber_createMessageElement = function _M3eAriaDescriber_createMessageElement(message, role) {
  var _b, _c;
  const messageElement = document.createElement("span");
  messageElement.id = `m3e-describedby-message-${__classPrivateFieldSet(this, _a$2, (_c = __classPrivateFieldGet(this, _a$2, "f", _M3eAriaDescriber_nextId), _b = _c++, _c), "f", _M3eAriaDescriber_nextId), _b}`;
  messageElement.role = role;
  messageElement.ariaAtomic = "true";
  messageElement.innerText = message.trim();
  return messageElement;
};
/** @private */
_M3eAriaDescriber_nextId = {
  value: 0
};
/** @private */
_M3eAriaDescriber_messageContainers = {
  value: new Map()
};
globalThis.M3eAriaDescriber = M3eAriaDescriber;

var _a$1, _M3eInteractivityChecker_cannotFocusParent;
/** Utility for checking the interactivity of an element, such as whether it is focusable. */
class M3eInteractivityChecker {
  /**
   * Determines whether a given element can receive focus.
   * @param {Element} element The element to test.
   * @param {readonly Element[]} [parents = undefined] The known parent elements to test. The default value is `undefined`.
   * @returns {boolean} Whether `element` can receive focus.
   */
  static isFocusable(element, parents) {
    if (element.matches(":is(button,input,select,textarea,object,:is(a,area)[href],[tabindex],[contenteditable=true]):not(:disabled,[disabled])")) {
      return !__classPrivateFieldGet(this, _a$1, "m", _M3eInteractivityChecker_cannotFocusParent).call(this, parents);
    }
    if (!element.localName.includes("-") || !element.matches(":not(:disabled,[disabled])") || element.getAttribute("aria-disabled") === "true") {
      return false;
    }
    if (element.shadowRoot?.delegatesFocus) {
      return !__classPrivateFieldGet(this, _a$1, "m", _M3eInteractivityChecker_cannotFocusParent).call(this, parents);
    }
    return false;
  }
  /**
   * Finds interactive elements that descend from the specified element.
   * @param {HTMLElement} element The `HTMLElement` to search.
   * @returns {HTMLElement[]} The interactive elements that descend from `element`.
   */
  static findInteractiveElements(element) {
    const elements = new Array();
    const walker = element.ownerDocument.createTreeWalker(element, NodeFilter.SHOW_ELEMENT);
    for (let node = walker.nextNode(); node; node = walker.nextNode()) {
      const element = walker.currentNode;
      if (this.isFocusable(element)) {
        elements.push(element);
      }
    }
    return elements;
  }
}
_a$1 = M3eInteractivityChecker, _M3eInteractivityChecker_cannotFocusParent = function _M3eInteractivityChecker_cannotFocusParent(parents) {
  return parents?.some(x => x.matches(":is([aria-hidden='true'],:disabled,[disabled])")) ?? false;
};
globalThis.M3eInteractivityChecker = M3eInteractivityChecker;

var _M3eFocusTrapElement_instances, _M3eFocusTrapElement_onFocus, _M3eFocusTrapElement_getFirstAndLastFocusableChild, _M3eFocusTrapElement_walkTree, _M3eFocusTrapElement_walkShadowRoot, _M3eFocusTrapElement_walkSlot;
/**
 * A non-visual element used to trap focus within nested content.
 * @tag m3e-focus-trap
 *
 * @slot - Renders content for which to trap focus.
 *
 * @attr disabled - Disables the focus trap.
 */
let M3eFocusTrapElement = class M3eFocusTrapElement extends Disabled(LitElement) {
  constructor() {
    super(...arguments);
    _M3eFocusTrapElement_instances.add(this);
  }
  /** @inheritdoc */
  render() {
    const trap = html`<div class="trap" .inert="${this.disabled}" tabindex="0" aria-hidden="true" @focus="${__classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_onFocus)}"></div>`;
    return html`${trap}<slot></slot>${trap}`;
  }
};
_M3eFocusTrapElement_instances = new WeakSet();
_M3eFocusTrapElement_onFocus = function _M3eFocusTrapElement_onFocus(e) {
  function matchesRelatedTarget(relatedTarget, test) {
    return relatedTarget === test || relatedTarget instanceof HTMLElement && relatedTarget.shadowRoot !== null && relatedTarget.shadowRoot.contains(test);
  }
  const [first, last] = __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_getFirstAndLastFocusableChild).call(this);
  const isFirst = e?.target === this._firstTrap;
  const fromFirst = matchesRelatedTarget(e.relatedTarget, first);
  const fromLast = matchesRelatedTarget(e.relatedTarget, last);
  const fromOutside = !fromFirst && !fromLast;
  if (!isFirst && fromLast || isFirst && fromOutside) {
    first?.focus();
    return;
  }
  if (isFirst && fromFirst || !isFirst && fromOutside) {
    last?.focus();
  }
};
_M3eFocusTrapElement_getFirstAndLastFocusableChild = function _M3eFocusTrapElement_getFirstAndLastFocusableChild() {
  let first = null;
  let last = null;
  __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_walkTree).call(this, this, [], (element, parents) => {
    if (M3eInteractivityChecker.isFocusable(element, parents) && element instanceof HTMLElement && element.tabIndex >= 0) {
      first = first ?? element;
      last = element;
    }
  });
  return [first, last];
};
_M3eFocusTrapElement_walkTree = function _M3eFocusTrapElement_walkTree(root, parents, callback) {
  parents = [...parents, root];
  for (const node of root.childNodes) {
    if (node.nodeType !== Node.ELEMENT_NODE) continue;
    const element = node;
    callback(element, parents);
    if (element.shadowRoot) {
      __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_walkShadowRoot).call(this, element, parents, callback);
    } else if (element instanceof HTMLSlotElement) {
      __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_walkSlot).call(this, element, parents, callback);
    } else if (element.hasChildNodes()) {
      __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_walkTree).call(this, element, parents, callback);
    }
  }
};
_M3eFocusTrapElement_walkShadowRoot = function _M3eFocusTrapElement_walkShadowRoot(root, parents, callback) {
  for (const node of root.shadowRoot?.childNodes ?? []) {
    if (node.nodeType !== Node.ELEMENT_NODE) continue;
    const element = node;
    callback(element, parents);
    if (!element.hasChildNodes()) continue;
    __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_walkTree).call(this, element, parents, callback);
  }
};
_M3eFocusTrapElement_walkSlot = function _M3eFocusTrapElement_walkSlot(slot, parents, callback) {
  for (const node of slot.assignedElements()) {
    callback(node, parents);
    if (!node.hasChildNodes()) continue;
    __classPrivateFieldGet(this, _M3eFocusTrapElement_instances, "m", _M3eFocusTrapElement_walkTree).call(this, node, parents, callback);
  }
};
/** The styles of the element. */
M3eFocusTrapElement.styles = css`:host { display: contents; } .trap { outline: none; }`;
__decorate([e(".trap")], M3eFocusTrapElement.prototype, "_firstTrap", void 0);
M3eFocusTrapElement = __decorate([t$1("m3e-focus-trap")], M3eFocusTrapElement);

/**
 * Adapted from Angular Material CDK Live Announcer
 * Source: https://github.com/angular/components/blob/main/src/cdk/a11y/live-announcer/live-announcer.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _a, _M3eLiveAnnouncer_nextId, _M3eLiveAnnouncer_liveElement, _M3eLiveAnnouncer_previousTimeout, _M3eLiveAnnouncer_currentPromise, _M3eLiveAnnouncer_currentResolve, _M3eLiveAnnouncer_createLiveElement;
/** Utility for announcing messages to screen readers. */
class M3eLiveAnnouncer {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  static announce(message, ...args) {
    if (isServer) {
      return Promise.resolve();
    }
    __classPrivateFieldSet(this, _a, __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement) ?? __classPrivateFieldGet(this, _a, "m", _M3eLiveAnnouncer_createLiveElement).call(this), "f", _M3eLiveAnnouncer_liveElement);
    let politeness;
    let duration;
    if (args.length === 1 && typeof args[0] === "number") {
      duration = args[0];
    } else {
      [politeness, duration] = args;
    }
    this.clear();
    clearTimeout(__classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_previousTimeout));
    __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement).setAttribute("aria-live", politeness ?? "polite");
    for (const modal of document.querySelectorAll("m3e-dialog")) {
      addAriaReferencedId(modal, "aria-owns", __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement).id);
    }
    __classPrivateFieldSet(this, _a, __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_currentPromise) ?? new Promise(resolve => __classPrivateFieldSet(this, _a, resolve, "f", _M3eLiveAnnouncer_currentResolve)), "f", _M3eLiveAnnouncer_currentPromise);
    clearTimeout(__classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_previousTimeout));
    __classPrivateFieldSet(this, _a, setTimeout(() => {
      if (!__classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement)) {
        return;
      }
      __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement).textContent = message;
      if (duration !== undefined) {
        __classPrivateFieldSet(this, _a, setTimeout(() => this.clear(), duration), "f", _M3eLiveAnnouncer_previousTimeout);
      }
      __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_currentResolve)?.call(this);
      __classPrivateFieldSet(this, _a, __classPrivateFieldSet(this, _a, undefined, "f", _M3eLiveAnnouncer_currentResolve), "f", _M3eLiveAnnouncer_currentPromise);
    }, 100), "f", _M3eLiveAnnouncer_previousTimeout);
    return __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_currentPromise);
  }
  /**
   * Clears the current text from the announcer element. Can be used to prevent
   * screen readers from reading the text out again while the user is going
   * through page landmarks.
   */
  static clear() {
    if (__classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement)) {
      __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_liveElement).textContent = "";
    }
  }
}
_a = M3eLiveAnnouncer, _M3eLiveAnnouncer_createLiveElement = function _M3eLiveAnnouncer_createLiveElement() {
  var _b, _c;
  const prev = document.getElementsByClassName("m3e-live-announcer");
  for (let i = 0; i < prev.length; i++) {
    prev[i].remove();
  }
  const liveAnnouncer = document.createElement("div");
  liveAnnouncer.classList.add("m3e-live-announcer");
  liveAnnouncer.setAttribute("aria-atomic", "true");
  liveAnnouncer.setAttribute("aria-live", "polite");
  liveAnnouncer.id = `m3e-live-announcer-${__classPrivateFieldSet(this, _a, (_c = __classPrivateFieldGet(this, _a, "f", _M3eLiveAnnouncer_nextId), _b = _c++, _c), "f", _M3eLiveAnnouncer_nextId), _b}`;
  visuallyHide(liveAnnouncer.style);
  document.body.append(liveAnnouncer);
  return liveAnnouncer;
};
/** @private */
_M3eLiveAnnouncer_nextId = {
  value: 0
};
/** @private */
_M3eLiveAnnouncer_liveElement = {
  value: void 0
};
/** @private */
_M3eLiveAnnouncer_previousTimeout = {
  value: void 0
};
/** @private */
_M3eLiveAnnouncer_currentPromise = {
  value: void 0
};
/** @private */
_M3eLiveAnnouncer_currentResolve = {
  value: void 0
};
globalThis.M3eLiveAnnouncer = M3eLiveAnnouncer;

export { FocusKeyManager, KeyCode, ListKeyManager, ListManager, M3eAriaDescriber, M3eFocusTrapElement, M3eInteractivityChecker, M3eLiveAnnouncer, ModifierKeys, RadioKeyManager, RovingTabIndexManager, SelectionManager, Typeahead, addAriaReferencedId, getAriaReferenceIds, getKeyCode, hasModifierKey, isModifierAllowed, isTypeaheadItem, removeAriaReferencedId, selectionManager, typeaheadLabel };
