// @m3e/web/dist/drawer-container.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { unsafeCSS, css, LitElement, html } from 'lit';
import { DesignToken, ResizeController, hasAssignedNodes, HtmlFor, ActionElementBase } from '@m3e/web/core';
import { M3eBreakpointObserver, Breakpoint } from '@m3e/web/core/layout';
import { addAriaReferencedId, removeAriaReferencedId } from '@m3e/web/core/a11y';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = globalThis,
  e$1 = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s = Symbol(),
  o$2 = new WeakMap();
let n$2 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$1 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$2.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$2.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$3 = t => new n$2("string" == typeof t ? t : t + "", void 0, s),
  S = (s, o) => {
    if (e$1) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$1 = e$1 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$3(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i,
    defineProperty: e,
    getOwnPropertyDescriptor: h,
    getOwnPropertyNames: r$2,
    getOwnPropertySymbols: o$1,
    getPrototypeOf: n$1
  } = Object,
  a = globalThis,
  c = a.trustedTypes,
  l = c ? c.emptyScript : "",
  p = a.reactiveElementPolyfillSupport,
  d = (t, s) => t,
  u = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f = (t, s) => !i(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    useDefault: false,
    hasChanged: f
  };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties"))) return;
    const t = n$1(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t = this.properties,
        s = [...r$2(t), ...o$1(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$1(s));
    } else void 0 !== s && i.push(c$1(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d("elementProperties")] = new Map(), y[d("finalized")] = new Map(), p?.({
  ReactiveElement: y
}), (a.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    hasChanged: f
  },
  r$1 = (t = o, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n(t) {
  return (e, o) => "object" == typeof o ? r$1(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
function r(r) {
  return n({
    ...r,
    state: true,
    attribute: false
  });
}

/**
 * Component design tokens that control `M3eDrawerContainerElement`.
 * @internal
 */
const DrawerContainerToken = {
  containerColor: unsafeCSS(`var(--m3e-drawer-container-color, ${DesignToken.color.surface})`),
  containerElevation: unsafeCSS(`var(--m3e-drawer-container-elevation, ${DesignToken.elevation.level0})`),
  containerWidth: unsafeCSS(`var(--m3e-drawer-container-width, 22.5rem)`),
  scrimOpacity: unsafeCSS("var(--m3e-drawer-container-scrim-opacity, 32%)"),
  cornerShape: unsafeCSS(`var(--m3e-modal-drawer-corner-shape, ${DesignToken.shape.corner.large})`),
  modalContainerColor: unsafeCSS(`var(--m3e-modal-drawer-container-color, ${DesignToken.color.surfaceContainerLow})`),
  modalContainerElevation: unsafeCSS(`var(--m3e-modal-drawer-elevation, ${DesignToken.elevation.level1})`),
  dividerColor: unsafeCSS(`var(--m3e-drawer-divider-color, ${DesignToken.color.outline})`),
  dividerThickness: unsafeCSS("var(--m3e-drawer-divider-thickness, 1px)")
};

/**
 * Styles for `M3eDrawerContainerElement`.
 * @internal
 */
const DrawerContainerStyle = css`:host { position: relative; overflow: hidden; flex: 1 1 auto; } .start, .end { position: absolute; top: 0; bottom: 0; z-index: 3; outline: none; overflow-x: hidden; overflow-y: auto; scrollbar-width: ${DesignToken.scrollbar.thinWidth}; scrollbar-color: ${DesignToken.scrollbar.color}; box-sizing: border-box; background-color: ${DrawerContainerToken.containerColor}; box-shadow: ${DrawerContainerToken.containerElevation}; transition: ${unsafeCSS(`margin-inline ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard},
      visibility ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard} allow-discrete,
      background-color ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard},
      box-shadow ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard}`)}; } .start { inset-inline-start: 0; } .end { inset-inline-end: 0; } :host(:not([start])) .start { visibility: hidden; margin-inline-start: calc(0px - var(--_start-drawer-size, ${DrawerContainerToken.containerWidth})); } :host([start]) .start { margin-inline-start: 0; } :host(:not([end])) .end { visibility: hidden; margin-inline-end: calc(0px - var(--_end-drawer-size, ${DrawerContainerToken.containerWidth})); } :host([end]) .end { margin-inline-end: 0; } ::slotted([slot="start"]), ::slotted([slot="end"]) { height: 100%; width: ${DrawerContainerToken.containerWidth}; box-sizing: border-box; } .content { position: relative; height: 100%; overflow: auto; margin-inline-start: 0; margin-inline-end: 0; transition: ${unsafeCSS(`margin-inline ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard}`)}; } :host(.-start-push[start]) .content, :host(.-start-side[start]) .content { margin-inline-start: var(--_start-drawer-size, ${DrawerContainerToken.containerWidth}); } :host(.-end-push[end]) .content, :host(.-end-side[end]) .content { margin-inline-end: var(--_end-drawer-size, ${DrawerContainerToken.containerWidth}); } .scrim { display: block; position: absolute; left: 0; right: 0; top: 0; bottom: 0; z-index: 1; background-color: ${DesignToken.color.scrim}; opacity: 0; visibility: hidden; transition: ${unsafeCSS(`opacity ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard}, 
        visibility ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard} allow-discrete`)}; } @starting-style { .scrim { opacity: 0; } } :host(.-start-push[start]) .scrim, :host(.-end-push[end]) .scrim, :host(.-start-over[start]) .scrim, :host(.-end-over[end]) .scrim { visibility: visible; opacity: ${DrawerContainerToken.scrimOpacity}; } :host(.-start-push) .start, :host(.-start-over) .start { border-start-end-radius: ${DrawerContainerToken.cornerShape}; border-end-end-radius: ${DrawerContainerToken.cornerShape}; } :host(.-end-push) .end, :host(.-end-over) .end { border-start-start-radius: ${DrawerContainerToken.cornerShape}; border-end-start-radius: ${DrawerContainerToken.cornerShape}; } :host(.-start-push) .start, :host(.-end-push) .end, :host(.-start-over) .start, :host(.-end-over) .end { background-color: ${DrawerContainerToken.modalContainerColor}; box-shadow: ${DrawerContainerToken.modalContainerElevation}; } :host([start-divider]) .start { border-inline-end-color: transparent; border-inline-end-width: ${DrawerContainerToken.dividerThickness}; border-inline-end-style: solid; box-sizing: border-box; } :host([start-divider].-start-side[start]:not(.-end-push[end]):not(.-end-over[end])) .start { border-inline-end-color: ${DrawerContainerToken.dividerColor}; } :host([end-divider]) .end { border-inline-start-color: transparent; border-inline-start-width: ${DrawerContainerToken.dividerThickness}; border-inline-start-style: solid; box-sizing: border-box; } :host([end-divider].-end-side[end]:not(.-start-push[start]):not(.-start-over[start])) .end { border-inline-start-color: ${DrawerContainerToken.dividerColor}; } :host(.-no-animate) .start, :host(.-no-animate) .end, :host(.-no-animate) .content { transition: none; } @media (forced-colors: active) { .start, .end { background-color: Canvas; box-shadow: unset; transition: ${unsafeCSS(`margin ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard},
      visibility ${DesignToken.motion.duration.medium4} ${DesignToken.motion.easing.standard} allow-discrete`)}; } :host(.-start-push) .start, :host(.-end-push) .end, :host(.-start-over) .start, :host(.-end-over) .end { background-color: Canvas; box-shadow: unset; border-color: CanvasText; } .start, .end { border-style: solid; border-color: Canvas; border-width: 1px; } .start { border-inline-start-style: none; } .end { border-inline-end-style: none; } :host([start-divider].-start-side[start]:not(.-end-push[end]):not(.-end-over[end])) .start { border-inline-end-color: GrayText; } :host([end-divider].-end-side[end]:not(.-start-push[start]):not(.-start-over[start])) .end { border-inline-start-color: GrayText; } } @media (prefers-reduced-motion) { .start, .end, .content, .scrim { transition: none; } }`;

var _M3eDrawerContainerElement_instances, _M3eDrawerContainerElement_breakpointUnobserve, _M3eDrawerContainerElement_disableStartFocusTrap, _M3eDrawerContainerElement_disableEndFocusTrap, _M3eDrawerContainerElement_resizeController, _M3eDrawerContainerElement_handleScrimClick, _M3eDrawerContainerElement_handleStartSlotChange, _M3eDrawerContainerElement_handleEndSlotChange, _M3eDrawerContainerElement_handleDrawerResize, _M3eDrawerContainerElement_clearMode, _M3eDrawerContainerElement_updateMode;
/**
 * A container for one or two sliding drawers.
 *
 * @description
 * A responsive layout container that manages left and right drawers alongside main content.
 * Supports `over`, `push`, `side`, and `auto` modes, adapting drawer behavior based on breakpoint size.
 * Encodes spatial hierarchy, motion transitions, and accessibility semantics for modal, dismissible,
 * and permanent navigation.
 *
 * @example
 * The following example illustrates a typical drawer layout.
 * ```html
 * <m3e-drawer-container>
 *  <nav slot="start">
 *    <!-- Start drawer content -->
 *  </nav>
 *  <main>
 *    <!-- Main content -->
 *  </main>
 *  <aside slot="end">
 *    <!-- End drawer content -->
 *  </aside>
 * <m3e-drawer-container>
 * ```
 *
 * @tag m3e-drawer-container
 *
 * @slot - Renders the main content.
 * @slot start - Renders the start drawer.
 * @slot end - Renders the end drawer.
 *
 * @attr end - Whether the end drawer is open.
 * @attr end-mode - The behavior mode of the end drawer.
 * @attr end-divider - Whether to show a divider between the end drawer and content for `side` mode.
 * @attr start - Whether the start drawer is open.
 * @attr start-mode - The behavior mode of the start drawer.
 * @attr start-divider - Whether to show a divider between the start drawer and content for `side` mode.
 *
 * @fires change - Emitted when the state of the start or end drawers change.
 *
 * @cssprop --m3e-drawer-container-color - The background color of the drawer container.
 * @cssprop --m3e-drawer-container-elevation - The elevation level of the drawer container.
 * @cssprop --m3e-drawer-container-width - The width of the drawer container.
 * @cssprop --m3e-drawer-container-scrim-opacity - The opacity of the scrim behind the drawer.
 * @cssprop --m3e-modal-drawer-start-shape - The shape of the drawer’s start edge (typically left in LTR).
 * @cssprop --m3e-modal-drawer-end-shape - The shape of the drawer’s end edge (typically right in LTR).
 * @cssprop --m3e-modal-drawer-container-color - The background color of the modal drawer container.
 * @cssprop --m3e-modal-drawer-elevation - The elevation level of the modal drawer container.
 * @cssprop --m3e-drawer-divider-color - The color of the divider between drawer sections.
 * @cssprop --m3e-drawer-divider-thickness - The thickness of the divider line.
 */
let M3eDrawerContainerElement = class M3eDrawerContainerElement extends LitElement {
  constructor() {
    super(...arguments);
    _M3eDrawerContainerElement_instances.add(this);
    /** @private */
    this._startMode = "side";
    /** @private */
    this._endMode = "side";
    /** @private */
    _M3eDrawerContainerElement_breakpointUnobserve.set(this, void 0);
    /** @private */
    _M3eDrawerContainerElement_disableStartFocusTrap.set(this, false);
    /** @private */
    _M3eDrawerContainerElement_disableEndFocusTrap.set(this, false);
    /** @private */
    _M3eDrawerContainerElement_resizeController.set(this, new ResizeController(this, {
      target: null,
      skipInitial: true,
      callback: entries => __classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_handleDrawerResize).call(this, entries)
    }));
    /**
     * Whether the start drawer is open.
     * @default false
     */
    this.start = false;
    /**
     * The behavior mode of the start drawer.
     * @default "side"
     */
    this.startMode = "side";
    /**
     * Whether to show a divider between the start drawer and content for `side` mode.
     * @default "side"
     */
    this.startDivider = false;
    /**
     * Whether the end drawer is open.
     * @default false
     */
    this.end = false;
    /**
     * The behavior mode of the end drawer.
     * @default "side"
     */
    this.endMode = "side";
    /**
     * Whether to show a divider between the end drawer and content for `side` mode.
     * @default "side"
     */
    this.endDivider = false;
  }
  /** @inheritdoc */
  connectedCallback() {
    super.connectedCallback();
    this.classList.add("-no-animate");
  }
  /** @inheritdoc */
  disconnectedCallback() {
    super.disconnectedCallback();
    __classPrivateFieldGet(this, _M3eDrawerContainerElement_breakpointUnobserve, "f")?.call(this);
    __classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_clearMode).call(this);
  }
  /** @inheritdoc */
  update(changedProperties) {
    super.update(changedProperties);
    if (changedProperties.has("startMode") || changedProperties.has("endMode")) {
      __classPrivateFieldGet(this, _M3eDrawerContainerElement_breakpointUnobserve, "f")?.call(this);
      if (this.startMode === "auto" || this.endMode === "auto") {
        __classPrivateFieldSet(this, _M3eDrawerContainerElement_breakpointUnobserve, M3eBreakpointObserver.observe([Breakpoint.XSmall, Breakpoint.Small], matches => __classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_updateMode).call(this, matches, true)), "f");
      } else {
        __classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_updateMode).call(this);
      }
    }
    if (changedProperties.has("start")) {
      if (this.start && this.end && this._endMode !== "side") {
        this.end = false;
        this.dispatchEvent(new Event("change", {
          bubbles: true
        }));
      }
    } else if (changedProperties.has("end")) {
      if (this.end && this.start && this._startMode !== "side") {
        this.start = false;
        this.dispatchEvent(new Event("change", {
          bubbles: true
        }));
      }
    }
  }
  /** @inheritdoc */
  firstUpdated(_changedProperties) {
    super.firstUpdated(_changedProperties);
    let drawer = this.shadowRoot?.querySelector(".start");
    if (drawer) __classPrivateFieldGet(this, _M3eDrawerContainerElement_resizeController, "f").observe(drawer);
    drawer = this.shadowRoot?.querySelector(".end");
    if (drawer) __classPrivateFieldGet(this, _M3eDrawerContainerElement_resizeController, "f").observe(drawer);
  }
  /** @inheritdoc */
  render() {
    return html`<div class="start"><m3e-focus-trap ?disabled="${!this.start || this._startMode === "side" || __classPrivateFieldGet(this, _M3eDrawerContainerElement_disableStartFocusTrap, "f")}"><slot name="start" @slotchange="${__classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_handleStartSlotChange)}"></slot></m3e-focus-trap></div><div class="content" .inert="${(this._startMode !== "side" || this._endMode !== "side") && (this.start || this.end)}"><slot></slot></div><div class="scrim" @click="${__classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_handleScrimClick)}"></div><div class="end"><m3e-focus-trap ?disabled="${!this.end || this._endMode === "side" || __classPrivateFieldGet(this, _M3eDrawerContainerElement_disableEndFocusTrap, "f")}"><slot name="end" @slotchange="${__classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_handleEndSlotChange)}"></slot></m3e-focus-trap></div>`;
  }
};
_M3eDrawerContainerElement_breakpointUnobserve = new WeakMap();
_M3eDrawerContainerElement_disableStartFocusTrap = new WeakMap();
_M3eDrawerContainerElement_disableEndFocusTrap = new WeakMap();
_M3eDrawerContainerElement_resizeController = new WeakMap();
_M3eDrawerContainerElement_instances = new WeakSet();
_M3eDrawerContainerElement_handleScrimClick = function _M3eDrawerContainerElement_handleScrimClick() {
  if (this._startMode !== "side") {
    this.start = false;
  }
  if (this._endMode !== "side") {
    this.end = false;
  }
  this.dispatchEvent(new Event("change", {
    bubbles: true
  }));
};
_M3eDrawerContainerElement_handleStartSlotChange = function _M3eDrawerContainerElement_handleStartSlotChange(e) {
  __classPrivateFieldSet(this, _M3eDrawerContainerElement_disableStartFocusTrap, !hasAssignedNodes(e.target), "f");
};
_M3eDrawerContainerElement_handleEndSlotChange = function _M3eDrawerContainerElement_handleEndSlotChange(e) {
  __classPrivateFieldSet(this, _M3eDrawerContainerElement_disableEndFocusTrap, !hasAssignedNodes(e.target), "f");
};
_M3eDrawerContainerElement_handleDrawerResize = function _M3eDrawerContainerElement_handleDrawerResize(entries) {
  for (const entry of entries) {
    const borderSize = Array.isArray(entry.borderBoxSize) ? entry.borderBoxSize[0] : entry.borderBoxSize;
    if (entry.target.classList.contains("start")) {
      this.style.setProperty("--_start-drawer-size", `${borderSize.inlineSize}px`);
    } else if (entry.target.classList.contains("end")) {
      this.style.setProperty("--_end-drawer-size", `${borderSize.inlineSize}px`);
    }
  }
  if (this.classList.contains("-no-animate")) {
    // Force synchronous layout flush
    void this.offsetTop;
    this.classList.remove("-no-animate");
  }
};
_M3eDrawerContainerElement_clearMode = function _M3eDrawerContainerElement_clearMode() {
  this.classList.remove("-start-over");
  this.classList.remove("-start-push");
  this.classList.remove("-start-side");
  this.classList.remove("-end-over");
  this.classList.remove("-end-push");
  this.classList.remove("-end-side");
};
_M3eDrawerContainerElement_updateMode = function _M3eDrawerContainerElement_updateMode(breakpoints, autoClose = false) {
  let autoCloseStart = false,
    autoCloseEnd = false;
  if (this.startMode === "auto") {
    if (breakpoints?.get(Breakpoint.Medium)) {
      this._startMode = "side";
    } else if (breakpoints?.get(Breakpoint.Small)) {
      autoCloseStart = this._startMode === "side" && this.start;
      this._startMode = "push";
    } else if (breakpoints?.get(Breakpoint.XSmall)) {
      autoCloseStart = this._startMode !== "over" && this.start;
      this._startMode = "over";
    } else {
      this._startMode = "side";
    }
  } else {
    this._startMode = this.startMode;
  }
  if (this.endMode === "auto") {
    if (breakpoints?.get(Breakpoint.Medium)) {
      this._endMode = "side";
    } else if (breakpoints?.get(Breakpoint.Small)) {
      autoCloseEnd = this._endMode === "side" && this.end;
      this._endMode = "push";
    } else if (breakpoints?.get(Breakpoint.XSmall)) {
      autoCloseEnd = this._endMode !== "over" && this.end;
      this._endMode = "over";
    } else {
      this._endMode = "side";
    }
  } else {
    this._endMode = this.endMode;
  }
  __classPrivateFieldGet(this, _M3eDrawerContainerElement_instances, "m", _M3eDrawerContainerElement_clearMode).call(this);
  this.classList.add(`-start-${this._startMode}`);
  this.classList.add(`-end-${this._endMode}`);
  if (autoClose && (autoCloseStart || autoCloseEnd)) {
    if (autoCloseStart) {
      this.start = false;
    }
    if (autoCloseEnd) {
      this.end = false;
    }
    this.dispatchEvent(new Event("change", {
      bubbles: true
    }));
  }
};
/** The styles of the element. */
M3eDrawerContainerElement.styles = DrawerContainerStyle;
__decorate([r()], M3eDrawerContainerElement.prototype, "_startMode", void 0);
__decorate([r()], M3eDrawerContainerElement.prototype, "_endMode", void 0);
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eDrawerContainerElement.prototype, "start", void 0);
__decorate([n({
  attribute: "start-mode",
  reflect: true
})], M3eDrawerContainerElement.prototype, "startMode", void 0);
__decorate([n({
  attribute: "start-divider",
  type: Boolean,
  reflect: true
})], M3eDrawerContainerElement.prototype, "startDivider", void 0);
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eDrawerContainerElement.prototype, "end", void 0);
__decorate([n({
  attribute: "end-mode",
  reflect: true
})], M3eDrawerContainerElement.prototype, "endMode", void 0);
__decorate([n({
  attribute: "end-divider",
  type: Boolean,
  reflect: true
})], M3eDrawerContainerElement.prototype, "endDivider", void 0);
M3eDrawerContainerElement = __decorate([t$1("m3e-drawer-container")], M3eDrawerContainerElement);

var _M3eDrawerToggleElement_instances, _M3eDrawerToggleElement_drawerContainerChangeHandler, _M3eDrawerToggleElement_togglingDrawer, _M3eDrawerToggleElement_handleDrawerContainerChange, _M3eDrawerToggleElement_toggleDrawer, _M3eDrawerToggleElement_updateToggle;
/**
 * An element, nested within a clickable element, used to toggle the opened state of a drawer.
 *
 * @example
 * The following example illustrates the use of a `m3e-drawer-toggle`, nested inside a `m3e-icon-button` component,
 * which toggles the open state of the start drawer.
 *
 * ```html
 * <m3e-icon-button slot="leading-icon" aria-label="Menu" toggle selected>
 *   <m3e-drawer-toggle for="startDrawer"></m3e-drawer-toggle>
 *   <m3e-icon name="menu"></m3e-icon>
 *   <m3e-icon slot="selected" name="menu_open"></m3e-icon>
 * </m3e-icon-button>
 *
 * <m3e-drawer-container start>
 *   <nav slot="start" id="startDrawer" aria-label="Navigation">
 *     <!-- Start drawer content -->
 *   </nav>
 *   <!-- Container content -->
 * </m3e-drawer-container>
 * ```
 *
 * @tag m3e-drawer-toggle
 */
let M3eDrawerToggleElement = class M3eDrawerToggleElement extends HtmlFor(ActionElementBase) {
  constructor() {
    super(...arguments);
    _M3eDrawerToggleElement_instances.add(this);
    /** @private */
    _M3eDrawerToggleElement_drawerContainerChangeHandler.set(this, () => __classPrivateFieldGet(this, _M3eDrawerToggleElement_instances, "m", _M3eDrawerToggleElement_handleDrawerContainerChange).call(this));
    /** @private */
    _M3eDrawerToggleElement_togglingDrawer.set(this, false);
  }
  /** @inheritdoc */
  attach(control) {
    super.attach(control);
    if (this.htmlFor && this.parentElement) {
      addAriaReferencedId(this.parentElement, "aria-controls", this.htmlFor);
    }
    const container = control.closest("m3e-drawer-container");
    if (container) {
      container.addEventListener("change", __classPrivateFieldGet(this, _M3eDrawerToggleElement_drawerContainerChangeHandler, "f"));
      __classPrivateFieldGet(this, _M3eDrawerToggleElement_instances, "m", _M3eDrawerToggleElement_updateToggle).call(this, container, true);
    }
  }
  /** @inheritdoc */
  detach() {
    if (this.parentElement) {
      if (this.htmlFor) {
        removeAriaReferencedId(this.parentElement, "aria-controls", this.htmlFor);
      }
      this.parentElement.ariaExpanded = null;
    }
    this.control?.closest("m3e-drawer-container")?.removeEventListener("change", __classPrivateFieldGet(this, _M3eDrawerToggleElement_drawerContainerChangeHandler, "f"));
    super.detach();
  }
  /** @inheritdoc */
  _onClick() {
    __classPrivateFieldGet(this, _M3eDrawerToggleElement_instances, "m", _M3eDrawerToggleElement_toggleDrawer).call(this);
  }
};
_M3eDrawerToggleElement_drawerContainerChangeHandler = new WeakMap();
_M3eDrawerToggleElement_togglingDrawer = new WeakMap();
_M3eDrawerToggleElement_instances = new WeakSet();
_M3eDrawerToggleElement_handleDrawerContainerChange = function _M3eDrawerToggleElement_handleDrawerContainerChange() {
  if (this.control && !__classPrivateFieldGet(this, _M3eDrawerToggleElement_togglingDrawer, "f")) {
    const container = this.control.closest("m3e-drawer-container");
    if (container) {
      __classPrivateFieldGet(this, _M3eDrawerToggleElement_instances, "m", _M3eDrawerToggleElement_updateToggle).call(this, container, true);
    }
  }
};
_M3eDrawerToggleElement_toggleDrawer = /** @private */
async function _M3eDrawerToggleElement_toggleDrawer() {
  if (!this.parentElement || !this.control) {
    return;
  }
  const container = this.control.closest("m3e-drawer-container");
  if (container) {
    __classPrivateFieldSet(this, _M3eDrawerToggleElement_togglingDrawer, true, "f");
    try {
      if (this.control.slot === "start") {
        container.start = !container.start;
      } else if (this.control.slot === "end") {
        container.end = !container.end;
      }
      if (container.isUpdatePending) {
        await container.updateComplete;
      }
      await __classPrivateFieldGet(this, _M3eDrawerToggleElement_instances, "m", _M3eDrawerToggleElement_updateToggle).call(this, container, false);
    } finally {
      __classPrivateFieldSet(this, _M3eDrawerToggleElement_togglingDrawer, false, "f");
    }
  }
};
_M3eDrawerToggleElement_updateToggle = /** @private */
async function _M3eDrawerToggleElement_updateToggle(container, syncToggle) {
  if (!this.parentElement || !this.control) {
    return;
  }
  let selected = false;
  if (this.control.slot === "start") {
    selected = container.start;
  } else if (this.control.slot === "end") {
    selected = container.end;
  }
  this.parentElement.ariaExpanded = `${selected}`;
  if (syncToggle && this.parentElement.hasAttribute("toggle")) {
    this.parentElement.toggleAttribute("selected", selected);
  }
  if (this.parentElement instanceof LitElement) {
    // Wait for update and remove aria-pressed due to use of aria-expanded.
    await this.parentElement.updateComplete;
    this.parentElement.ariaPressed = null;
  }
};
M3eDrawerToggleElement = __decorate([t$1("m3e-drawer-toggle")], M3eDrawerToggleElement);

export { M3eDrawerContainerElement, M3eDrawerToggleElement };
