// @m3e/web/dist/expansion-panel.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { css, LitElement, html, unsafeCSS, nothing } from 'lit';
import { DesignToken, KeyboardClick, Focusable, Disabled, AttachInternals, Role, EventAttribute, hasAssignedNodes } from '@m3e/web/core';
import { M3eDirectionality } from '@m3e/web/core/bidi';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = globalThis,
  e$3 = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s = Symbol(),
  o$2 = new WeakMap();
let n$2 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$3 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$2.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$2.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$2 = t => new n$2("string" == typeof t ? t : t + "", void 0, s),
  S = (s, o) => {
    if (e$3) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$1 = e$3 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$2(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i,
    defineProperty: e$2,
    getOwnPropertyDescriptor: h,
    getOwnPropertyNames: r$1,
    getOwnPropertySymbols: o$1,
    getPrototypeOf: n$1
  } = Object,
  a = globalThis,
  c = a.trustedTypes,
  l = c ? c.emptyScript : "",
  p = a.reactiveElementPolyfillSupport,
  d = (t, s) => t,
  u = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f = (t, s) => !i(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    useDefault: false,
    hasChanged: f
  };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e$2(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties"))) return;
    const t = n$1(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t = this.properties,
        s = [...r$1(t), ...o$1(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$1(s));
    } else void 0 !== s && i.push(c$1(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d("elementProperties")] = new Map(), y[d("finalized")] = new Map(), p?.({
  ReactiveElement: y
}), (a.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    hasChanged: f
  },
  r = (t = o, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n(t) {
  return (e, o) => "object" == typeof o ? r(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const e$1 = (e, t, c) => (c.configurable = true, c.enumerable = true, Reflect.decorate && "object" != typeof t && Object.defineProperty(e, t, c), c);

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
function e(e, r) {
  return (n, s, i) => {
    const o = t => t.renderRoot?.querySelector(e) ?? null;
    return e$1(n, s, {
      get() {
        return o(this);
      }
    });
  };
}

var _M3eAccordionElement_instances, _M3eAccordionElement_panels, _M3eAccordionElement_handleSlotChange, _M3eAccordionElement_handleOpening;
/**
 * Combines multiple expansion panels in to an accordion.
 *
 * @description
 * The `m3e-accordion` component organizes multiple expansion panels into a coordinated, accessible group.
 * It supports single or multiple open panels via the `multi` attribute, and provides expressive theming
 * and shape control for grouped layouts. The accordion manages open/close state across its child panels,
 * enabling interactive disclosure patterns for complex content.
 *
 * @example
 * The following example illustrates the basic use of the `m3e-accordion` and `m3e-expansion-panel` components.
 *
 * ```html
 * <m3e-accordion>
 *   <m3e-expansion-panel>
 *     <span slot="header">Panel 1</span>
 *     I am content for the first expansion panel
 *   </m3e-expansion-panel>
 *   <m3e-expansion-panel>
 *     <span slot="header">Panel 2</span>
 *     I am content for the second expansion panel
 *   </m3e-expansion-panel>
 * </m3e-accordion>
 * ```
 *
 * @tag m3e-accordion
 *
 * @slot - Renders the panels of the accordion.
 *
 * @attr multi - Whether multiple expansion panels can be open at the same time.
 */
let M3eAccordionElement = class M3eAccordionElement extends LitElement {
  constructor() {
    super(...arguments);
    _M3eAccordionElement_instances.add(this);
    /** @private */
    _M3eAccordionElement_panels.set(this, []);
    /**
     * Whether multiple expansion panels can be open at the same time.
     * @default false
     */
    this.multi = false;
  }
  /** The panels of the accordion. */
  get panels() {
    return __classPrivateFieldGet(this, _M3eAccordionElement_panels, "f");
  }
  /** @inheritdoc */
  render() {
    return html`<slot @slotchange="${__classPrivateFieldGet(this, _M3eAccordionElement_instances, "m", _M3eAccordionElement_handleSlotChange)}" @opening="${__classPrivateFieldGet(this, _M3eAccordionElement_instances, "m", _M3eAccordionElement_handleOpening)}"></slot>`;
  }
};
_M3eAccordionElement_panels = new WeakMap();
_M3eAccordionElement_instances = new WeakSet();
_M3eAccordionElement_handleSlotChange = function _M3eAccordionElement_handleSlotChange() {
  __classPrivateFieldSet(this, _M3eAccordionElement_panels, [...this.querySelectorAll("m3e-expansion-panel")], "f");
  if (this.multi) {
    return;
  }
  let alreadyOpen = false;
  for (const panel of __classPrivateFieldGet(this, _M3eAccordionElement_panels, "f").filter(x => !x.open)) {
    if (alreadyOpen) {
      panel.open = false;
    }
    alreadyOpen = true;
  }
};
_M3eAccordionElement_handleOpening = function _M3eAccordionElement_handleOpening(e) {
  if (!this.multi) {
    for (const panel of __classPrivateFieldGet(this, _M3eAccordionElement_panels, "f").filter(x => x !== e.target && x.open)) {
      panel.open = false;
    }
  }
};
/** The styles of the element. */
M3eAccordionElement.styles = css`:host { display: block; } ::slotted(m3e-expansion-panel) { --m3e-expansion-panel-container-color: ${DesignToken.color.surface}; --m3e-expansion-panel-elevation: ${DesignToken.elevation.level2}; --m3e-expansion-panel-open-shape: ${DesignToken.shape.corner.medium}; --_expansion-panel-open-spacing: 1rem; } ::slotted(m3e-expansion-panel:first-of-type:last-of-type) { --m3e-expansion-panel-shape: ${DesignToken.shape.corner.medium}; } ::slotted(m3e-expansion-panel:first-of-type:not(:last-of-type)) { --m3e-expansion-panel-shape: ${DesignToken.shape.corner.value.medium} ${DesignToken.shape.corner.value.medium} ${DesignToken.shape.corner.value.none} ${DesignToken.shape.corner.value.none}; } ::slotted(m3e-expansion-panel:last-of-type:not(:first-of-type)) { --m3e-expansion-panel-shape: ${DesignToken.shape.corner.value.none} ${DesignToken.shape.corner.value.none} ${DesignToken.shape.corner.value.medium} ${DesignToken.shape.corner.value.medium}; }`;
__decorate([n({
  type: Boolean
})], M3eAccordionElement.prototype, "multi", void 0);
M3eAccordionElement = __decorate([t$1("m3e-accordion")], M3eAccordionElement);

/**
 * Component design tokens for `M3eExpansionHeaderElement`.
 * @internal
 */
const ExpansionHeaderToken = {
  collapsedHeight: unsafeCSS("var(--m3e-expansion-header-collapsed-height, 3rem)"),
  expandedHeight: unsafeCSS("var(--m3e-expansion-header-expanded-height, 4rem)"),
  paddingLeft: unsafeCSS("var(--m3e-expansion-header-padding-left, 1.5rem)"),
  paddingRight: unsafeCSS("var(--m3e-expansion-header-padding-right, 1.5rem)"),
  spacing: unsafeCSS("var(--m3e-expansion-header-spacing, 0.5rem)"),
  toggleIconSize: unsafeCSS("var(--m3e-expansion-header-toggle-icon-size, 1.5rem)"),
  fontSize: unsafeCSS(`var(--m3e-expansion-header-font-size, ${DesignToken.typescale.standard.title.medium.fontSize})`),
  fontWeight: unsafeCSS(`var(--m3e-expansion-header-font-weight, ${DesignToken.typescale.standard.title.medium.fontWeight})`),
  lineHeight: unsafeCSS(`var(--m3e-expansion-header-line-height, ${DesignToken.typescale.standard.title.medium.lineHeight})`),
  tracking: unsafeCSS(`var(--m3e-expansion-header-tracking, ${DesignToken.typescale.standard.title.medium.tracking})`)
};

/**
 * Styles for `M3eExpansionHeaderElement`.
 * @internal
 */
const ExpansionHeaderStyle = css`:host { display: block; border-radius: inherit; outline: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); transition: ${unsafeCSS(`height var(--m3e-collapsible-animation-duration, ${DesignToken.motion.duration.medium1})
        ${DesignToken.motion.easing.standard}`)}; } .base { box-sizing: border-box; width: 100%; height: 100%; display: flex; border-radius: inherit; outline: none; position: relative; padding-inline-start: ${ExpansionHeaderToken.paddingLeft}; padding-inline-end: ${ExpansionHeaderToken.paddingRight}; font-size: ${ExpansionHeaderToken.fontSize}; font-weight: ${ExpansionHeaderToken.fontWeight}; line-height: ${ExpansionHeaderToken.lineHeight}; letter-spacing: ${ExpansionHeaderToken.tracking}; column-gap: ${ExpansionHeaderToken.spacing}; } :host(:not(:disabled)) { cursor: pointer; } :host(:not([aria-expanded="true"])) { height: ${ExpansionHeaderToken.collapsedHeight}; } :host([aria-expanded="true"]) { height: ${ExpansionHeaderToken.expandedHeight}; } :host(:not(:focus-visible)) .state-layer { --m3e-state-layer-focus-color: transparent; } :host([aria-expanded="true"]) .state-layer { --m3e-state-layer-hover-color: transparent; } :host([aria-expanded="true"]) [part="background"], .content { flex: 1 1 auto; display: flex; align-items: center; } .toggle { display: flex; align-items: center; justify-content: center; vertical-align: middle; font-size: ${ExpansionHeaderToken.toggleIconSize}; transition: ${unsafeCSS(`transform var(--m3e-collapsible-animation-duration, ${DesignToken.motion.duration.medium1})
        ${DesignToken.motion.easing.standard}`)}; } :host([toggle-direction="vertical"][aria-expanded="true"]) .toggle { transform: rotate(180deg); } :host([toggle-direction="horizontal"][aria-expanded="true"]) .toggle { transform: rotate(var(--_expansion-header-horizontal-expanded-toggle-rotation, 90deg)); } :host([toggle-position="before"]) .toggle { margin-inline-start: calc(0px - ${ExpansionHeaderToken.spacing}); } :host([toggle-position="after"]) .toggle { margin-inline-end: calc(0px - ${ExpansionHeaderToken.spacing}); } :host([hide-toggle]) .toggle { display: none; } ::slotted([slot="toggle-icon"]) { font-size: inherit !important; flex: none; } ::slotted(svg[slot="toggle-icon"]) { width: 1em; height: 1em; } @media (prefers-reduced-motion) { :host, .toggle { transition: none; } }`;

/**
 * Component design tokens for `M3eExpansionPanelElement`.
 * @internal
 */
const ExpansionPanelToken = {
  textColor: unsafeCSS(`var(--m3e-expansion-panel-text-color, ${DesignToken.color.onSurface})`),
  disabledTextColor: unsafeCSS(`var(--m3e-expansion-panel-disabled-text-color, ${DesignToken.color.onSurface})`),
  disabledTextOpacity: unsafeCSS(`var(--m3e-expansion-panel-disabled-text-opacity, 38%)`),
  containerColor: unsafeCSS("var(--m3e-expansion-panel-container-color)"),
  collapsedElevation: unsafeCSS("var(--m3e-expansion-panel-elevation)"),
  collapsedShape: unsafeCSS("var(--m3e-expansion-panel-shape)"),
  expandedElevation: unsafeCSS("var(--m3e-expansion-panel-open-elevation, var(--m3e-expansion-panel-elevation))"),
  expandedShape: unsafeCSS("var(--m3e-expansion-panel-open-shape, var(--m3e-expansion-panel-shape))"),
  expandedSpace: unsafeCSS("var(--_expansion-panel-open-spacing)"),
  contentPadding: unsafeCSS("var(--m3e-expansion-panel-content-padding, 0 1.5rem 1rem 1.5rem)"),
  actionsSpacing: unsafeCSS("var(--m3e-expansion-panel-actions-spacing, 0.5rem)"),
  actionsPadding: unsafeCSS("var(--m3e-expansion-panel-actions-padding, 1rem 1.5rem 1rem 1.5rem)"),
  actionsDividerThickness: unsafeCSS("var(--m3e-expansion-panel-actions-divider-thickness, var(--m3e-divider-thickness, 1px))"),
  actionsDividerColor: unsafeCSS(`var(--m3e-expansion-panel-actions-divider-color, var(--m3e-divider-color, ${DesignToken.color.outlineVariant}))`)
};

/**
 * Styles for `M3eExpansionPanelElement`.
 * @internal
 */
const ExpansionPanelStyle = css`:host { display: block; } .base { background-color: ${ExpansionPanelToken.containerColor}; transition: ${unsafeCSS(`box-shadow var(--m3e-collapsible-animation-duration, ${DesignToken.motion.duration.medium1})
        ${DesignToken.motion.easing.standard}`)}; } :host(:not(:disabled)) .base { color: ${ExpansionPanelToken.textColor}; } :host(:disabled) .base { color: color-mix( in srgb, ${ExpansionPanelToken.disabledTextColor} ${ExpansionPanelToken.disabledTextOpacity}, transparent ); } :host(:not([open])) .base { box-shadow: ${ExpansionPanelToken.collapsedElevation}; border-radius: ${ExpansionPanelToken.collapsedShape}; } :host([open]) .base { box-shadow: ${ExpansionPanelToken.expandedElevation}; border-radius: ${ExpansionPanelToken.expandedShape}; margin-block: ${ExpansionPanelToken.expandedSpace}; } .toggle { display: flex; align-items: center; justify-content: center; font-size: ${ExpansionHeaderToken.toggleIconSize}; } ::slotted([slot="toggle-icon"]) { font-size: inherit !important; flex: none; } .toggle svg, ::slotted(svg[slot="toggle-icon"]) { width: 1em; height: 1em; } .content { padding: ${ExpansionPanelToken.contentPadding}; } :host(:not(.-has-actions)) .actions { display: none; } .actions { padding: ${ExpansionPanelToken.actionsPadding}; border-top-style: solid; border-top-width: ${ExpansionPanelToken.actionsDividerThickness}; border-top-color: ${ExpansionPanelToken.actionsDividerColor}; } ::slotted([slot="actions"]) { flex: none; display: flex; align-items: center; column-gap: ${ExpansionPanelToken.actionsSpacing}; } ::slotted([slot="actions"][end]) { justify-content: flex-end; } @media (prefers-reduced-motion) { .base { transition: none; } } @media (forced-colors: active) { .base { border: 1px solid CanvasText; } :host(:disabled) .base { color: GrayText; } .actions { border-top-color: GrayText; } }`;

var _M3eExpansionHeaderElement_instances, _M3eExpansionHeaderElement_renderToggle;
/**
 * A button used to toggle the expanded state of an expansion panel.
 *
 * @tag m3e-expansion-header
 *
 * @slot - Renders the content of the header.
 * @slot toggle-icon - Renders the icon of the expansion toggle.
 *
 * @attr hide-toggle - Whether to hide the expansion toggle.
 * @attr toggle-direction - The direction of the expansion toggle.
 * @attr toggle-position - The position of the expansion toggle.
 *
 * @fires click - Emitted when the element is clicked.
 *
 * @cssprop --m3e-expansion-header-collapsed-height - Height of the header when the panel is collapsed.
 * @cssprop --m3e-expansion-header-expanded-height - Height of the header when the panel is expanded.
 * @cssprop --m3e-expansion-header-padding-left - Left padding inside the header.
 * @cssprop --m3e-expansion-header-padding-right - Right padding inside the header.
 * @cssprop --m3e-expansion-header-spacing - Spacing between header elements.
 * @cssprop --m3e-expansion-header-toggle-icon-size - Size of the toggle icon (e.g. chevron).
 * @cssprop --m3e-expansion-header-font-size - The font size of the header text.
 * @cssprop --m3e-expansion-header-font-weight - The font weight of the header text.
 * @cssprop --m3e-expansion-header-line-height - The line height of the header text.
 * @cssprop --m3e-expansion-header-tracking - Letter spacing (tracking) of the header text.
 */
let M3eExpansionHeaderElement = class M3eExpansionHeaderElement extends KeyboardClick(Focusable(Disabled(AttachInternals(Role(LitElement, "button"), true)))) {
  constructor() {
    super(...arguments);
    _M3eExpansionHeaderElement_instances.add(this);
    /**
     * The direction of the expansion toggle.
     * @default "vertical"
     */
    this.toggleDirection = "vertical";
    /**
     * The position of the expansion toggle.
     * @default "after"
     */
    this.togglePosition = "after";
    /**
     * Whether to hide the expansion toggle.
     * @default false
     */
    this.hideToggle = false;
  }
  /** @inheritdoc */
  firstUpdated(_changedProperties) {
    super.firstUpdated(_changedProperties);
    [this._focusRing, this._stateLayer].forEach(x => x?.attach(this));
  }
  /** @inheritdoc */
  render() {
    return html`<div class="base"><m3e-focus-ring class="focus-ring" ?disabled="${this.disabled}"></m3e-focus-ring><m3e-state-layer class="state-layer" ?disabled="${this.disabled}"></m3e-state-layer>${this.togglePosition === "before" ? __classPrivateFieldGet(this, _M3eExpansionHeaderElement_instances, "m", _M3eExpansionHeaderElement_renderToggle).call(this) : nothing}<div class="content"><slot></slot></div>${this.togglePosition === "after" ? __classPrivateFieldGet(this, _M3eExpansionHeaderElement_instances, "m", _M3eExpansionHeaderElement_renderToggle).call(this) : nothing}</div>`;
  }
};
_M3eExpansionHeaderElement_instances = new WeakSet();
_M3eExpansionHeaderElement_renderToggle = function _M3eExpansionHeaderElement_renderToggle() {
  return html`<div class="toggle" aria-hidden="true"><slot name="toggle-icon"></slot></div>`;
};
/** The styles of the element. */
M3eExpansionHeaderElement.styles = ExpansionHeaderStyle;
__decorate([e(".focus-ring")], M3eExpansionHeaderElement.prototype, "_focusRing", void 0);
__decorate([e(".state-layer")], M3eExpansionHeaderElement.prototype, "_stateLayer", void 0);
__decorate([n({
  attribute: "toggle-direction",
  reflect: true
})], M3eExpansionHeaderElement.prototype, "toggleDirection", void 0);
__decorate([n({
  attribute: "toggle-position",
  reflect: true
})], M3eExpansionHeaderElement.prototype, "togglePosition", void 0);
__decorate([n({
  attribute: "hide-toggle",
  type: Boolean,
  reflect: true
})], M3eExpansionHeaderElement.prototype, "hideToggle", void 0);
M3eExpansionHeaderElement = __decorate([t$1("m3e-expansion-header")], M3eExpansionHeaderElement);

/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */
var _M3eExpansionPanelElement_instances, _M3eExpansionPanelElement_id, _M3eExpansionPanelElement_contentId, _M3eExpansionPanelElement_headerId, _M3eExpansionPanelElement_directionalitySubscription, _M3eExpansionPanelElement_renderToggleIcon, _M3eExpansionPanelElement_handleHeaderClick, _M3eExpansionPanelElement_handleKeyDown, _M3eExpansionPanelElement_handleCollapsibleEvent, _M3eExpansionPanelElement_handleActionsSlotChange, _M3eExpansionPanelElement_updateHeaderToggleRotation;
var M3eExpansionPanelElement_1;
/**
 * An expandable details-summary view.
 *
 * @description
 * The `m3e-expansion-panel` component provides an accessible, animated details-summary view for
 * organizing content in collapsible sections. It supports custom header, content, actions, and
 * toggle icon slots, and offers configurable toggle position and direction. The panel responds to
 * open/close states, emits lifecycle events, and supports rich theming via CSS custom properties
 * for elevation, shape, spacing, and color.
 *
 * @example
 * The following example illustrates the basic use of the `m3e-accordion` and `m3e-expansion-panel` components.
 *
 * ```html
 * <m3e-accordion>
 *   <m3e-expansion-panel>
 *     <span slot="header">Panel 1</span>
 *     I am content for the first expansion panel
 *   </m3e-expansion-panel>
 *   <m3e-expansion-panel>
 *     <span slot="header">Panel 2</span>
 *     I am content for the second expansion panel
 *   </m3e-expansion-panel>
 * </m3e-accordion>
 * ```
 *
 * @tag m3e-expansion-panel
 *
 * @slot - Renders the detail of the panel.
 * @slot actions - Renders the actions bar of the panel.
 * @slot header - Renders the header content.
 * @slot toggle-icon - Renders the expansion toggle icon.
 *
 * @attr disabled - Whether the element is disabled.
 * @attr hide-toggle - Whether to hide the expansion toggle.
 * @attr open - Whether the panel is expanded.
 * @attr toggle-direction - The direction of the expansion toggle.
 * @attr toggle-position - The position of the expansion toggle.
 *
 * @fires opening - Emitted when the expansion panel begins to open.
 * @fires opened - Emitted when the expansion panel has opened.
 * @fires closing - Emitted when the expansion panel begins to close.
 * @fires closed - Emitted when the expansion panel has closed.
 *
 * @cssprop --m3e-expansion-header-collapsed-height - Height of the header when the panel is collapsed.
 * @cssprop --m3e-expansion-header-expanded-height - Height of the header when the panel is expanded.
 * @cssprop --m3e-expansion-header-padding-left - Left padding inside the header.
 * @cssprop --m3e-expansion-header-padding-right - Right padding inside the header.
 * @cssprop --m3e-expansion-header-spacing - Spacing between header elements.
 * @cssprop --m3e-expansion-header-toggle-icon-size - Size of the toggle icon (e.g. chevron).
 * @cssprop --m3e-expansion-header-font-size - The font size of the header text.
 * @cssprop --m3e-expansion-header-font-weight - The font weight of the header text.
 * @cssprop --m3e-expansion-header-line-height - The line height of the header text.
 * @cssprop --m3e-expansion-header-tracking - Letter spacing (tracking) of the header text.
 * @cssprop --m3e-expansion-panel-text-color - Color of the panel's text content.
 * @cssprop --m3e-expansion-panel-disabled-text-color - Color of the panel's text content, when disabled.
 * @cssprop --m3e-expansion-panel-disabled-text-opacity - Opacity of the panel's text content, when disabled.
 * @cssprop --m3e-expansion-panel-container-color - Background color of the panel container.
 * @cssprop --m3e-expansion-panel-elevation - Elevation level when the panel is collapsed.
 * @cssprop --m3e-expansion-panel-shape - Shape (e.g. border radius) of the panel when collapsed.
 * @cssprop --m3e-expansion-panel-open-elevation - Elevation level when the panel is expanded.
 * @cssprop --m3e-expansion-panel-open-shape - Shape (e.g. border radius) of the panel when expanded.
 * @cssprop --m3e-expansion-panel-content-padding - Padding around the panel's content area.
 * @cssprop --m3e-expansion-panel-actions-spacing - Spacing between action buttons or elements.
 * @cssprop --m3e-expansion-panel-actions-padding - Padding around the actions section.
 * @cssprop --m3e-expansion-panel-actions-divider-thickness - Thickness of the divider above actions.
 * @cssprop --m3e-expansion-panel-actions-divider-color - Color of the divider above actions.
 */
let M3eExpansionPanelElement = M3eExpansionPanelElement_1 = class M3eExpansionPanelElement extends EventAttribute(Disabled(AttachInternals(LitElement, true)), "opening", "opened", "closing", "closed") {
  constructor() {
    super(...arguments);
    _M3eExpansionPanelElement_instances.add(this);
    /** @private */
    _M3eExpansionPanelElement_id.set(this, M3eExpansionPanelElement_1.__nextId++);
    /** @private */
    _M3eExpansionPanelElement_contentId.set(this, `m3e-expansion-panel-${__classPrivateFieldGet(this, _M3eExpansionPanelElement_id, "f")}-content`);
    /** @private */
    _M3eExpansionPanelElement_headerId.set(this, `m3e-expansion-panel-${__classPrivateFieldGet(this, _M3eExpansionPanelElement_id, "f")}-header`);
    /** @private */
    _M3eExpansionPanelElement_directionalitySubscription.set(this, void 0);
    /**
     * Whether the panel is expanded.
     * @default false
     */
    this.open = false;
    /**
     * The direction of the expansion toggle.
     * @default "vertical"
     */
    this.toggleDirection = "vertical";
    /**
     * The position of the expansion toggle.
     * @default "after"
     */
    this.togglePosition = "after";
    /**
     * Whether to hide the expansion toggle.
     * @default false
     */
    this.hideToggle = false;
  }
  /** @inheritdoc */
  focus(options) {
    this._header?.focus(options);
  }
  /** @inheritdoc */
  blur() {
    this._header?.blur();
  }
  /** @inheritdoc */
  click() {
    this._header?.click();
  }
  /** @inheritdoc */
  connectedCallback() {
    super.connectedCallback();
    __classPrivateFieldSet(this, _M3eExpansionPanelElement_directionalitySubscription, M3eDirectionality.observe(() => {
      this.requestUpdate();
      __classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_updateHeaderToggleRotation).call(this);
    }), "f");
  }
  /** @inheritdoc */
  disconnectedCallback() {
    super.disconnectedCallback();
    __classPrivateFieldGet(this, _M3eExpansionPanelElement_directionalitySubscription, "f")?.call(this);
  }
  /** @inheritdoc */
  firstUpdated(_changedProperties) {
    super.firstUpdated(_changedProperties);
    __classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_updateHeaderToggleRotation).call(this);
  }
  /** @inheritdoc */
  render() {
    return html`<div class="base"><m3e-expansion-header id="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_headerId, "f")}" toggle-direction="${this.toggleDirection}" toggle-position="${this.togglePosition}" ?hide-toggle="${this.hideToggle}" ?disabled="${this.disabled}" aria-expanded="${this.open}" aria-controls="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_contentId, "f")}" @click="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleHeaderClick)}" @keydown="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleKeyDown)}"><div slot="toggle-icon" class="toggle"><slot name="toggle-icon">${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_renderToggleIcon).call(this)}</slot></div><slot name="header"></slot></m3e-expansion-header><m3e-collapsible id="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_contentId, "f")}" role="region" aria-labelledby="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_headerId, "f")}" ?open="${this.open}" @opening="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleCollapsibleEvent)}" @opened="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleCollapsibleEvent)}" @closing="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleCollapsibleEvent)}" @closed="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleCollapsibleEvent)}"><div class="content"><slot></slot></div><div class="actions"><slot name="actions" @slotchange="${__classPrivateFieldGet(this, _M3eExpansionPanelElement_instances, "m", _M3eExpansionPanelElement_handleActionsSlotChange)}"></slot></div></m3e-collapsible></div>`;
  }
};
_M3eExpansionPanelElement_id = new WeakMap();
_M3eExpansionPanelElement_contentId = new WeakMap();
_M3eExpansionPanelElement_headerId = new WeakMap();
_M3eExpansionPanelElement_directionalitySubscription = new WeakMap();
_M3eExpansionPanelElement_instances = new WeakSet();
_M3eExpansionPanelElement_renderToggleIcon = function _M3eExpansionPanelElement_renderToggleIcon() {
  return this.toggleDirection === "vertical" ? html`<svg viewBox="0 -960 960 960" fill="currentColor"><path d="M480-344 240-584l56-56 184 184 184-184 56 56-240 240Z"/></svg>` : M3eDirectionality.current === "ltr" ? html`<svg viewBox="0 -960 960 960" fill="currentColor"><path d="M504-480 320-664l56-56 240 240-240 240-56-56 184-184Z"/></svg>` : html`<svg viewBox="0 -960 960 960" fill="currentColor"><path d="M560-240 320-480l240-240 56 56-184 184 184 184-56 56Z"/></svg>`;
};
_M3eExpansionPanelElement_handleHeaderClick = function _M3eExpansionPanelElement_handleHeaderClick() {
  this.open = !this.open;
};
_M3eExpansionPanelElement_handleKeyDown = function _M3eExpansionPanelElement_handleKeyDown(e) {
  switch (e.key) {
    case "ArrowUp":
      {
        e.preventDefault();
        const accordion = this.closest("m3e-accordion");
        if (accordion) {
          const panels = [...accordion.panels].reverse();
          const index = panels.indexOf(this);
          (panels.find((x, i) => !x.disabled && i > index) ?? panels.find((x, i) => !x.disabled && i < index))?._header?.focus();
        } else {
          this.open = false;
        }
      }
      break;
    case "ArrowDown":
      {
        e.preventDefault();
        const accordion = this.closest("m3e-accordion");
        if (accordion) {
          const index = accordion.panels.indexOf(this);
          (accordion.panels.find((x, i) => !x.disabled && i > index) ?? accordion.panels.find((x, i) => !x.disabled && i < index))?._header?.focus();
        } else {
          this.open = true;
        }
      }
      break;
  }
};
_M3eExpansionPanelElement_handleCollapsibleEvent = function _M3eExpansionPanelElement_handleCollapsibleEvent(e) {
  e.stopPropagation();
  this.dispatchEvent(new Event(e.type, {
    bubbles: true
  }));
};
_M3eExpansionPanelElement_handleActionsSlotChange = function _M3eExpansionPanelElement_handleActionsSlotChange(e) {
  this.classList.toggle("-has-actions", hasAssignedNodes(e.target));
};
_M3eExpansionPanelElement_updateHeaderToggleRotation = function _M3eExpansionPanelElement_updateHeaderToggleRotation() {
  if (M3eDirectionality.current === "rtl") {
    this._header?.style.setProperty("--_expansion-header-horizontal-expanded-toggle-rotation", "-90deg");
  } else {
    this._header?.style.removeProperty("--_expansion-header-horizontal-expanded-toggle-rotation");
  }
};
/** The styles of the element. */
M3eExpansionPanelElement.styles = ExpansionPanelStyle;
/** @private */
M3eExpansionPanelElement.__nextId = 0;
__decorate([e("m3e-expansion-header")], M3eExpansionPanelElement.prototype, "_header", void 0);
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eExpansionPanelElement.prototype, "open", void 0);
__decorate([n({
  attribute: "toggle-direction",
  reflect: true
})], M3eExpansionPanelElement.prototype, "toggleDirection", void 0);
__decorate([n({
  attribute: "toggle-position",
  reflect: true
})], M3eExpansionPanelElement.prototype, "togglePosition", void 0);
__decorate([n({
  attribute: "hide-toggle",
  type: Boolean,
  reflect: true
})], M3eExpansionPanelElement.prototype, "hideToggle", void 0);
M3eExpansionPanelElement = M3eExpansionPanelElement_1 = __decorate([t$1("m3e-expansion-panel")], M3eExpansionPanelElement);

export { M3eAccordionElement, M3eExpansionHeaderElement, M3eExpansionPanelElement };
