// @m3e/web/dist/paginator.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { LitElement, html, css, nothing } from 'lit';
import { EventAttribute, Role, DesignToken } from '@m3e/web/core';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = globalThis,
  e$1 = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s = Symbol(),
  o$2 = new WeakMap();
let n$2 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$1 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$2.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$2.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$2 = t => new n$2("string" == typeof t ? t : t + "", void 0, s),
  S = (s, o) => {
    if (e$1) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$1 = e$1 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$2(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i,
    defineProperty: e,
    getOwnPropertyDescriptor: h,
    getOwnPropertyNames: r$1,
    getOwnPropertySymbols: o$1,
    getPrototypeOf: n$1
  } = Object,
  a = globalThis,
  c = a.trustedTypes,
  l = c ? c.emptyScript : "",
  p = a.reactiveElementPolyfillSupport,
  d = (t, s) => t,
  u = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f = (t, s) => !i(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    useDefault: false,
    hasChanged: f
  };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties"))) return;
    const t = n$1(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t = this.properties,
        s = [...r$1(t), ...o$1(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$1(s));
    } else void 0 !== s && i.push(c$1(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d("elementProperties")] = new Map(), y[d("finalized")] = new Map(), p?.({
  ReactiveElement: y
}), (a.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    hasChanged: f
  },
  r = (t = o, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n(t) {
  return (e, o) => "object" == typeof o ? r(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * Adapted from Angular Material Paginator
 * Source: https://github.com/angular/components/blob/main/src/material/paginator/paginator.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _M3ePaginatorElement_instances, _M3ePaginatorElement_pageSizeLabelId, _M3ePaginatorElement_defaultRangeLabelFormatter, _M3ePaginatorElement_renderPageSize, _M3ePaginatorElement_parsePageSizes, _M3ePaginatorElement_renderRangeActions, _M3ePaginatorElement_handleSelectChange, _M3ePaginatorElement_emitPageEvent;
var M3ePaginatorElement_1;
/**
 * Provides navigation for paged information, typically used with a table.
 *
 * @description
 * The `m3e-paginator` component is a compact, accessible paginator control for navigating
 * paged data sets. It provides semantic first/previous/next/last navigation controls and an
 * optional page-size selector that integrates with Material design tokens and `m3e-form-field` variants
 * to ensure consistent typography, density, and spacing across applications.
 *
 * @example
 * The following example illustrates use of the `m3e-paginator`. In this example, there are 300 total
 * records and the first/last navigation controls are shown.
 * ```html
 * <m3e-paginator show-first-last-buttons length="300"></m3e-paginator>
 * ```
 *
 * @tag m3e-paginator
 *
 * @slot first-page-icon - Slot for a custom first-page icon.
 * @slot previous-page-icon - Slot for a custom previous-page icon.
 * @slot next-page-icon - Slot for a custom next-page icon.
 * @slot last-page-icon - Slot for a custom last-page icon.
 *
 * @attr disabled - Whether the element is disabled.
 * @attr first-page-label - The accessible label given to the button used to move to the first page.
 * @attr hide-page-size - Whether to hide page size selection.
 * @attr items-per-page-label - The label for the page size selector.
 * @attr last-page-label - The accessible label given to the button used to move to the last page.
 * @attr length - The length of the total number of items which are being paginated.
 * @attr next-page-label - The accessible label given to the button used to move to the next page.
 * @attr page-index - The zero-based page index of the displayed list of items.
 * @attr page-size - The number of items to display in a page.
 * @attr page-sizes - A comma separated list of available page sizes.
 * @attr page-size-variant - The appearance variant of the page size field.
 * @attr previous-page-label - The accessible label given to the button used to move to the previous page.
 * @attr show-first-last-buttons - Whether to show first/last buttons.
 *
 * @fires page - Emitted when a user selects a different page size or navigates to another page.
 *
 * @cssprop --m3e-paginator-font-size - The font size used for paginator text.
 * @cssprop --m3e-paginator-font-weight - The font weight used for paginator text.
 * @cssprop --m3e-paginator-line-height - The line height used for paginator text.
 * @cssprop --m3e-paginator-tracking - The letter-spacing used for paginator text.
 *
 */
let M3ePaginatorElement = M3ePaginatorElement_1 = class M3ePaginatorElement extends EventAttribute(Role(LitElement, "group"), "page") {
  constructor() {
    super(...arguments);
    _M3ePaginatorElement_instances.add(this);
    /* @private */
    _M3ePaginatorElement_pageSizeLabelId.set(this, `m3e-paginator-page-size-label-${M3ePaginatorElement_1.__nextId++}`);
    /* @private */
    _M3ePaginatorElement_defaultRangeLabelFormatter.set(this, (pageIndex, pageSize, length) => {
      length = Math.max(length, 0);
      if (pageSize === "all") return "";
      if (length === 0 || pageSize <= 0) return `0 of ${length}`;
      const startIndex = pageIndex * pageSize;
      const endIndex = startIndex < length ? Math.min(startIndex + pageSize, length) : startIndex + pageSize;
      return `${startIndex + 1} - ${endIndex} of ${length}`;
    });
    /**
     * Whether the element is disabled.
     * @default false
     */
    this.disabled = false;
    /**
     * The zero-based page index of the displayed list of items.
     * @default 0
     */
    this.pageIndex = 0;
    /**
     * The length of the total number of items which are being paginated.
     * @default 0
     */
    this.length = 0;
    /**
     * The number of items to display in a page.
     * @default 50
     */
    this.pageSize = 50;
    /**
     * A comma separated list of available page sizes.
     * @default "5,10,25,50,100"
     */
    this.pageSizes = "5,10,25,50,100";
    /**
     * Whether to hide page size selection.
     * @default false
     */
    this.hidePageSize = false;
    /**
     * Whether to show first/last buttons.
     * @default false
     */
    this.showFirstLastButtons = false;
    /**
     * The label for the page size selector.
     * @default "Items per page:"
     */
    this.itemsPerPageLabel = "Items per page:";
    /**
     * The accessible label given to the button used to move to the previous page.
     * @default "Previous page"
     */
    this.previousPageLabel = "Previous page";
    /**
     * The accessible label given to the button used to move to the next page.
     * @default "Next page"
     */
    this.nextPageLabel = "Next page";
    /**
     * The accessible label given to the button used to move to the first page.
     * @default "First page"
     */
    this.firstPageLabel = "First page";
    /**
     * The accessible label given to the button used to move to the last page.
     * @default "Last page"
     */
    this.lastPageLabel = "Last page";
    /**
     * The appearance variant of the page size field.
     * @default "outlined"
     */
    this.pageSizeVariant = "outlined";
  }
  /** The total number of pages. */
  get pageCount() {
    return !this.pageSize || this.pageSize === "all" ? 0 : Math.ceil(this.length / this.pageSize);
  }
  /** Whether a previous page can be displayed. */
  get hasPreviousPage() {
    return this.pageSize !== "all" && this.pageSize > 0 && this.pageIndex >= 1;
  }
  /** Whether a next page can be displayed. */
  get hasNextPage() {
    return this.pageSize !== "all" && this.pageSize > 0 && this.pageIndex < this.pageCount - 1;
  }
  /** Move to the first page. */
  firstPage() {
    if (this.hasPreviousPage) {
      const previousPageIndex = this.pageIndex;
      this.pageIndex = 0;
      __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_emitPageEvent).call(this, previousPageIndex);
    }
  }
  /** Move to the previous page. */
  previousPage() {
    if (this.hasPreviousPage) {
      const previousPageIndex = this.pageIndex;
      this.pageIndex--;
      __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_emitPageEvent).call(this, previousPageIndex);
    }
  }
  /** Move to the next page. */
  nextPage() {
    if (this.hasNextPage) {
      const previousPageIndex = this.pageIndex;
      this.pageIndex++;
      __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_emitPageEvent).call(this, previousPageIndex);
    }
  }
  /** Move to the last page. */
  lastPage() {
    if (this.hasNextPage) {
      const previousPageIndex = this.pageIndex;
      this.pageIndex = this.pageCount - 1;
      __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_emitPageEvent).call(this, previousPageIndex);
    }
  }
  /** @inheritdoc */
  update(changedProperties) {
    super.update(changedProperties);
    if (changedProperties.has("pageIndex")) {
      const pageSizes = __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_parsePageSizes).call(this);
      if (!pageSizes.includes(this.pageSize)) {
        this.pageSizes = [...pageSizes, this.pageSize].join(",");
        this.pageSizes = __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_parsePageSizes).call(this).join(",");
      }
    }
  }
  /** @inheritdoc */
  render() {
    return html`<div class="outer"><div class="inner">${__classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_renderPageSize).call(this)} ${__classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_renderRangeActions).call(this)}</div></div>`;
  }
};
_M3ePaginatorElement_pageSizeLabelId = new WeakMap();
_M3ePaginatorElement_defaultRangeLabelFormatter = new WeakMap();
_M3ePaginatorElement_instances = new WeakSet();
_M3ePaginatorElement_renderPageSize = function _M3ePaginatorElement_renderPageSize() {
  const sizes = __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_parsePageSizes).call(this);
  return this.hidePageSize ? nothing : html`<div id="${__classPrivateFieldGet(this, _M3ePaginatorElement_pageSizeLabelId, "f")}" class="items-per-page-label" aria-live="polite">${this.itemsPerPageLabel}</div><m3e-form-field class="form-field" variant="${this.pageSizeVariant}" hide-subscript="always"><m3e-select class="select" aria-labelledby="${__classPrivateFieldGet(this, _M3ePaginatorElement_pageSizeLabelId, "f")}" hide-selection-indicator ?disabled="${this.disabled || sizes.length <= 1}" @change="${__classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_handleSelectChange)}">${sizes.map(x => html`<m3e-option value="${x}" ?selected="${x === this.pageSize}">${x === "all" ? "All" : x}</m3e-option>`)}</m3e-select></m3e-form-field>`;
};
_M3ePaginatorElement_parsePageSizes = function _M3ePaginatorElement_parsePageSizes() {
  const sizes = this.pageSizes.split(",").map(x => x.trim()).filter(x => x !== "").map(x => x === "all" ? x : Number(x));
  const numericSizes = sizes.filter(x => x !== "all");
  numericSizes.sort((a, b) => a - b);
  return sizes.some(x => x === "all") ? [...numericSizes, "all"] : numericSizes;
};
_M3ePaginatorElement_renderRangeActions = function _M3ePaginatorElement_renderRangeActions() {
  const rangeLabelFormatter = this.rangeLabelFormatter ?? __classPrivateFieldGet(this, _M3ePaginatorElement_defaultRangeLabelFormatter, "f");
  return this.pageSize === "all" ? nothing : html`<div class="range-actions"><div class="range-label">${rangeLabelFormatter(this.pageIndex, this.pageSize, this.length)}</div>${!this.showFirstLastButtons ? nothing : html`<m3e-icon-button id="firstPageButton" aria-label="${this.firstPageLabel}" ?disabled="${this.disabled || !this.hasPreviousPage}" @click="${this.firstPage}"><slot name="first-page-icon"><svg viewBox="0 -960 960 960" fill="currentColor"><path d="M240-240v-480h80v480h-80Zm440 0L440-480l240-240 56 56-184 184 184 184-56 56Z"/></svg></slot></m3e-icon-button><m3e-tooltip for="firstPageButton" position="above">${this.firstPageLabel}</m3e-tooltip>`}<m3e-icon-button id="previousPageButton" aria-label="${this.previousPageLabel}" ?disabled="${this.disabled || !this.hasPreviousPage}" @click="${this.previousPage}"><slot name="previous-page-icon"><svg viewBox="0 -960 960 960" fill="currentColor"><path d="M560-240 320-480l240-240 56 56-184 184 184 184-56 56Z"/></svg></slot></m3e-icon-button><m3e-tooltip for="previousPageButton" position="above">${this.previousPageLabel}</m3e-tooltip><m3e-icon-button id="nextPageButton" aria-label="${this.nextPageLabel}" ?disabled="${this.disabled || !this.hasNextPage}" @click="${this.nextPage}"><slot name="next-page-icon"><svg viewBox="0 -960 960 960" fill="currentColor"><path d="M504-480 320-664l56-56 240 240-240 240-56-56 184-184Z"/></svg></slot></m3e-icon-button><m3e-tooltip for="nextPageButton" position="above">${this.nextPageLabel}</m3e-tooltip>${!this.showFirstLastButtons ? nothing : html`<m3e-icon-button id="lastPageButton" aria-label="${this.lastPageLabel}" ?disabled="${this.disabled || !this.hasNextPage}" @click="${this.lastPage}"><slot name="last-page-icon"><svg viewBox="0 -960 960 960" fill="currentColor"><path d="m280-240-56-56 184-184-184-184 56-56 240 240-240 240Zm360 0v-480h80v480h-80Z"/></svg></slot></m3e-icon-button><m3e-tooltip for="lastPageButton" position="above">${this.lastPageLabel}</m3e-tooltip>`}</div>`;
};
_M3ePaginatorElement_handleSelectChange = function _M3ePaginatorElement_handleSelectChange(e) {
  const option = e.target?.selected[0];
  if (!option) return;
  const pageSize = option.value === "all" ? "all" : Number(option.value);
  if (pageSize !== this.pageSize) {
    const previousPageIndex = this.pageIndex;
    this.pageIndex = 0;
    this.pageSize = pageSize;
    __classPrivateFieldGet(this, _M3ePaginatorElement_instances, "m", _M3ePaginatorElement_emitPageEvent).call(this, previousPageIndex);
  }
};
_M3ePaginatorElement_emitPageEvent = function _M3ePaginatorElement_emitPageEvent(previousPageIndex) {
  this.dispatchEvent(new CustomEvent("page", {
    detail: {
      previousPageIndex: previousPageIndex,
      pageIndex: this.pageIndex,
      pageSize: this.pageSize,
      length: this.length
    }
  }));
};
/** The styles of the element. */
M3ePaginatorElement.styles = css`:host { display: block; } .outer { display: flex; font-size: var(--m3e-paginator-font-size, ${DesignToken.typescale.standard.body.small.fontSize}); font-weight: var(--m3e-paginator-font-weight, ${DesignToken.typescale.standard.body.small.fontWeight}); line-height: var(--m3e-paginator-line-height, ${DesignToken.typescale.standard.body.small.lineHeight}); letter-spacing: var(--m3e-paginator-tracking, ${DesignToken.typescale.standard.body.small.tracking}); } .inner { display: flex; align-items: center; justify-content: flex-end; padding-inline: 0.5rem; width: 100%; } .form-field { --md-sys-density-scale: -2; --m3e-form-field-font-size: var(--m3e-paginator-font-size, ${DesignToken.typescale.standard.body.small.fontSize}); --m3e-form-field-font-weight: var( --m3e-paginator-font-weight, ${DesignToken.typescale.standard.body.small.fontWeight} ); --m3e-form-field-line-height: var( --m3e-paginator-line-height, ${DesignToken.typescale.standard.body.small.lineHeight} ); --m3e-form-field-tracking: var(--m3e-paginator-tracking, ${DesignToken.typescale.standard.body.small.tracking}); } .items-per-page-label { display: flex; align-items: center; margin-inline-end: 0.5rem; } .form-field { min-width: auto; width: 6rem; margin-inline: 0.25rem; } .range-label { margin-inline: 1.5rem 2rem; } :host([hide-page-size]) .range-label { margin-inline-start: unset; } .range-actions { display: flex; align-items: center; } ::slotted([slot="first-page-icon"]), ::slotted([slot="previous-page-icon"]), ::slotted([slot="next-page-icon"]), ::slotted([slot="last-page-icon"]), svg { width: 1em; font-size: var(--m3e-icon-button-medium-icon-size, 1.5rem) !important; } :host(:dir(rtl)) svg { transform: rotate(180deg); }`;
/* @private */
M3ePaginatorElement.__nextId = 0;
__decorate([n({
  type: Boolean,
  reflect: true
})], M3ePaginatorElement.prototype, "disabled", void 0);
__decorate([n({
  attribute: "page-index",
  type: Number
})], M3ePaginatorElement.prototype, "pageIndex", void 0);
__decorate([n({
  type: Number
})], M3ePaginatorElement.prototype, "length", void 0);
__decorate([n({
  attribute: "page-size",
  converter: value => value === "all" ? "all" : Number(value)
})], M3ePaginatorElement.prototype, "pageSize", void 0);
__decorate([n({
  attribute: "page-sizes"
})], M3ePaginatorElement.prototype, "pageSizes", void 0);
__decorate([n({
  attribute: "hide-page-size",
  type: Boolean,
  reflect: true
})], M3ePaginatorElement.prototype, "hidePageSize", void 0);
__decorate([n({
  attribute: "show-first-last-buttons",
  type: Boolean
})], M3ePaginatorElement.prototype, "showFirstLastButtons", void 0);
__decorate([n({
  attribute: "items-per-page-label"
})], M3ePaginatorElement.prototype, "itemsPerPageLabel", void 0);
__decorate([n({
  attribute: "previous-page-label"
})], M3ePaginatorElement.prototype, "previousPageLabel", void 0);
__decorate([n({
  attribute: "next-page-label"
})], M3ePaginatorElement.prototype, "nextPageLabel", void 0);
__decorate([n({
  attribute: "first-page-label"
})], M3ePaginatorElement.prototype, "firstPageLabel", void 0);
__decorate([n({
  attribute: "last-page-label"
})], M3ePaginatorElement.prototype, "lastPageLabel", void 0);
__decorate([n({
  attribute: "page-size-variant"
})], M3ePaginatorElement.prototype, "pageSizeVariant", void 0);
M3ePaginatorElement = M3ePaginatorElement_1 = __decorate([t$1("m3e-paginator")], M3ePaginatorElement);

export { M3ePaginatorElement };
