// @m3e/web/dist/progress-indicator.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { LitElement, css, html, nothing, svg } from 'lit';
import { AttachInternals, Role, DesignToken, ResizeController, AnimationLoopController, resolveFragmentUrl, safeStyleMap } from '@m3e/web/core';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$3 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$2 = globalThis,
  e$4 = t$2.ShadowRoot && (void 0 === t$2.ShadyCSS || t$2.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s$1 = Symbol(),
  o$4 = new WeakMap();
let n$3 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s$1) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$4 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$4.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$4.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$3 = t => new n$3("string" == typeof t ? t : t + "", void 0, s$1),
  S$1 = (s, o) => {
    if (e$4) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t$2.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$2 = e$4 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$3(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i$2,
    defineProperty: e$3,
    getOwnPropertyDescriptor: h$1,
    getOwnPropertyNames: r$2,
    getOwnPropertySymbols: o$3,
    getPrototypeOf: n$2
  } = Object,
  a$1 = globalThis,
  c$1 = a$1.trustedTypes,
  l$1 = c$1 ? c$1.emptyScript : "",
  p$1 = a$1.reactiveElementPolyfillSupport,
  d$1 = (t, s) => t,
  u$1 = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l$1 : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f$1 = (t, s) => !i$2(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u$1,
    reflect: false,
    useDefault: false,
    hasChanged: f$1
  };
Symbol.metadata ??= Symbol("metadata"), a$1.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e$3(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h$1(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d$1("elementProperties"))) return;
    const t = n$2(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d$1("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d$1("properties"))) {
      const t = this.properties,
        s = [...r$2(t), ...o$3(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$2(s));
    } else void 0 !== s && i.push(c$2(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S$1(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u$1).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u$1;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f$1)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d$1("elementProperties")] = new Map(), y[d$1("finalized")] = new Map(), p$1?.({
  ReactiveElement: y
}), (a$1.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o$2 = {
    attribute: true,
    type: String,
    converter: u$1,
    reflect: false,
    hasChanged: f$1
  },
  r$1 = (t = o$2, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n$1(t) {
  return (e, o) => "object" == typeof o ? r$1(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = globalThis,
  i$1 = t$1.trustedTypes,
  s = i$1 ? i$1.createPolicy("lit-html", {
    createHTML: t => t
  }) : void 0,
  e$2 = "$lit$",
  h = `lit$${Math.random().toFixed(9).slice(2)}$`,
  o$1 = "?" + h,
  n = `<${o$1}>`,
  r = document,
  l = () => r.createComment(""),
  c = t => null === t || "object" != typeof t && "function" != typeof t,
  a = Array.isArray,
  u = t => a(t) || "function" == typeof t?.[Symbol.iterator],
  d = "[ \t\n\f\r]",
  f = /<(?:(!--|\/[^a-zA-Z])|(\/?[a-zA-Z][^>\s]*)|(\/?$))/g,
  v = /-->/g,
  _ = />/g,
  m = RegExp(`>|${d}(?:([^\\s"'>=/]+)(${d}*=${d}*(?:[^ \t\n\f\r"'\`<>=]|("|')|))|$)`, "g"),
  p = /'/g,
  g = /"/g,
  $ = /^(?:script|style|textarea|title)$/i,
  T = Symbol.for("lit-noChange"),
  E = Symbol.for("lit-nothing"),
  A = new WeakMap(),
  C = r.createTreeWalker(r, 129);
function P(t, i) {
  if (!a(t) || !t.hasOwnProperty("raw")) throw Error("invalid template strings array");
  return void 0 !== s ? s.createHTML(i) : i;
}
const V = (t, i) => {
  const s = t.length - 1,
    o = [];
  let r,
    l = 2 === i ? "<svg>" : 3 === i ? "<math>" : "",
    c = f;
  for (let i = 0; i < s; i++) {
    const s = t[i];
    let a,
      u,
      d = -1,
      y = 0;
    for (; y < s.length && (c.lastIndex = y, u = c.exec(s), null !== u);) y = c.lastIndex, c === f ? "!--" === u[1] ? c = v : void 0 !== u[1] ? c = _ : void 0 !== u[2] ? ($.test(u[2]) && (r = RegExp("</" + u[2], "g")), c = m) : void 0 !== u[3] && (c = m) : c === m ? ">" === u[0] ? (c = r ?? f, d = -1) : void 0 === u[1] ? d = -2 : (d = c.lastIndex - u[2].length, a = u[1], c = void 0 === u[3] ? m : '"' === u[3] ? g : p) : c === g || c === p ? c = m : c === v || c === _ ? c = f : (c = m, r = void 0);
    const x = c === m && t[i + 1].startsWith("/>") ? " " : "";
    l += c === f ? s + n : d >= 0 ? (o.push(a), s.slice(0, d) + e$2 + s.slice(d) + h + x) : s + h + (-2 === d ? i : x);
  }
  return [P(t, l + (t[s] || "<?>") + (2 === i ? "</svg>" : 3 === i ? "</math>" : "")), o];
};
class N {
  constructor({
    strings: t,
    _$litType$: s
  }, n) {
    let r;
    this.parts = [];
    let c = 0,
      a = 0;
    const u = t.length - 1,
      d = this.parts,
      [f, v] = V(t, s);
    if (this.el = N.createElement(f, n), C.currentNode = this.el.content, 2 === s || 3 === s) {
      const t = this.el.content.firstChild;
      t.replaceWith(...t.childNodes);
    }
    for (; null !== (r = C.nextNode()) && d.length < u;) {
      if (1 === r.nodeType) {
        if (r.hasAttributes()) for (const t of r.getAttributeNames()) if (t.endsWith(e$2)) {
          const i = v[a++],
            s = r.getAttribute(t).split(h),
            e = /([.?@])?(.*)/.exec(i);
          d.push({
            type: 1,
            index: c,
            name: e[2],
            strings: s,
            ctor: "." === e[1] ? H : "?" === e[1] ? I : "@" === e[1] ? L : k
          }), r.removeAttribute(t);
        } else t.startsWith(h) && (d.push({
          type: 6,
          index: c
        }), r.removeAttribute(t));
        if ($.test(r.tagName)) {
          const t = r.textContent.split(h),
            s = t.length - 1;
          if (s > 0) {
            r.textContent = i$1 ? i$1.emptyScript : "";
            for (let i = 0; i < s; i++) r.append(t[i], l()), C.nextNode(), d.push({
              type: 2,
              index: ++c
            });
            r.append(t[s], l());
          }
        }
      } else if (8 === r.nodeType) if (r.data === o$1) d.push({
        type: 2,
        index: c
      });else {
        let t = -1;
        for (; -1 !== (t = r.data.indexOf(h, t + 1));) d.push({
          type: 7,
          index: c
        }), t += h.length - 1;
      }
      c++;
    }
  }
  static createElement(t, i) {
    const s = r.createElement("template");
    return s.innerHTML = t, s;
  }
}
function S(t, i, s = t, e) {
  if (i === T) return i;
  let h = void 0 !== e ? s._$Co?.[e] : s._$Cl;
  const o = c(i) ? void 0 : i._$litDirective$;
  return h?.constructor !== o && (h?._$AO?.(false), void 0 === o ? h = void 0 : (h = new o(t), h._$AT(t, s, e)), void 0 !== e ? (s._$Co ??= [])[e] = h : s._$Cl = h), void 0 !== h && (i = S(t, h._$AS(t, i.values), h, e)), i;
}
class M {
  constructor(t, i) {
    this._$AV = [], this._$AN = void 0, this._$AD = t, this._$AM = i;
  }
  get parentNode() {
    return this._$AM.parentNode;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  u(t) {
    const {
        el: {
          content: i
        },
        parts: s
      } = this._$AD,
      e = (t?.creationScope ?? r).importNode(i, true);
    C.currentNode = e;
    let h = C.nextNode(),
      o = 0,
      n = 0,
      l = s[0];
    for (; void 0 !== l;) {
      if (o === l.index) {
        let i;
        2 === l.type ? i = new R(h, h.nextSibling, this, t) : 1 === l.type ? i = new l.ctor(h, l.name, l.strings, this, t) : 6 === l.type && (i = new z(h, this, t)), this._$AV.push(i), l = s[++n];
      }
      o !== l?.index && (h = C.nextNode(), o++);
    }
    return C.currentNode = r, e;
  }
  p(t) {
    let i = 0;
    for (const s of this._$AV) void 0 !== s && (void 0 !== s.strings ? (s._$AI(t, s, i), i += s.strings.length - 2) : s._$AI(t[i])), i++;
  }
}
class R {
  get _$AU() {
    return this._$AM?._$AU ?? this._$Cv;
  }
  constructor(t, i, s, e) {
    this.type = 2, this._$AH = E, this._$AN = void 0, this._$AA = t, this._$AB = i, this._$AM = s, this.options = e, this._$Cv = e?.isConnected ?? true;
  }
  get parentNode() {
    let t = this._$AA.parentNode;
    const i = this._$AM;
    return void 0 !== i && 11 === t?.nodeType && (t = i.parentNode), t;
  }
  get startNode() {
    return this._$AA;
  }
  get endNode() {
    return this._$AB;
  }
  _$AI(t, i = this) {
    t = S(this, t, i), c(t) ? t === E || null == t || "" === t ? (this._$AH !== E && this._$AR(), this._$AH = E) : t !== this._$AH && t !== T && this._(t) : void 0 !== t._$litType$ ? this.$(t) : void 0 !== t.nodeType ? this.T(t) : u(t) ? this.k(t) : this._(t);
  }
  O(t) {
    return this._$AA.parentNode.insertBefore(t, this._$AB);
  }
  T(t) {
    this._$AH !== t && (this._$AR(), this._$AH = this.O(t));
  }
  _(t) {
    this._$AH !== E && c(this._$AH) ? this._$AA.nextSibling.data = t : this.T(r.createTextNode(t)), this._$AH = t;
  }
  $(t) {
    const {
        values: i,
        _$litType$: s
      } = t,
      e = "number" == typeof s ? this._$AC(t) : (void 0 === s.el && (s.el = N.createElement(P(s.h, s.h[0]), this.options)), s);
    if (this._$AH?._$AD === e) this._$AH.p(i);else {
      const t = new M(e, this),
        s = t.u(this.options);
      t.p(i), this.T(s), this._$AH = t;
    }
  }
  _$AC(t) {
    let i = A.get(t.strings);
    return void 0 === i && A.set(t.strings, i = new N(t)), i;
  }
  k(t) {
    a(this._$AH) || (this._$AH = [], this._$AR());
    const i = this._$AH;
    let s,
      e = 0;
    for (const h of t) e === i.length ? i.push(s = new R(this.O(l()), this.O(l()), this, this.options)) : s = i[e], s._$AI(h), e++;
    e < i.length && (this._$AR(s && s._$AB.nextSibling, e), i.length = e);
  }
  _$AR(t = this._$AA.nextSibling, i) {
    for (this._$AP?.(false, true, i); t !== this._$AB;) {
      const i = t.nextSibling;
      t.remove(), t = i;
    }
  }
  setConnected(t) {
    void 0 === this._$AM && (this._$Cv = t, this._$AP?.(t));
  }
}
class k {
  get tagName() {
    return this.element.tagName;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  constructor(t, i, s, e, h) {
    this.type = 1, this._$AH = E, this._$AN = void 0, this.element = t, this.name = i, this._$AM = e, this.options = h, s.length > 2 || "" !== s[0] || "" !== s[1] ? (this._$AH = Array(s.length - 1).fill(new String()), this.strings = s) : this._$AH = E;
  }
  _$AI(t, i = this, s, e) {
    const h = this.strings;
    let o = false;
    if (void 0 === h) t = S(this, t, i, 0), o = !c(t) || t !== this._$AH && t !== T, o && (this._$AH = t);else {
      const e = t;
      let n, r;
      for (t = h[0], n = 0; n < h.length - 1; n++) r = S(this, e[s + n], i, n), r === T && (r = this._$AH[n]), o ||= !c(r) || r !== this._$AH[n], r === E ? t = E : t !== E && (t += (r ?? "") + h[n + 1]), this._$AH[n] = r;
    }
    o && !e && this.j(t);
  }
  j(t) {
    t === E ? this.element.removeAttribute(this.name) : this.element.setAttribute(this.name, t ?? "");
  }
}
class H extends k {
  constructor() {
    super(...arguments), this.type = 3;
  }
  j(t) {
    this.element[this.name] = t === E ? void 0 : t;
  }
}
class I extends k {
  constructor() {
    super(...arguments), this.type = 4;
  }
  j(t) {
    this.element.toggleAttribute(this.name, !!t && t !== E);
  }
}
class L extends k {
  constructor(t, i, s, e, h) {
    super(t, i, s, e, h), this.type = 5;
  }
  _$AI(t, i = this) {
    if ((t = S(this, t, i, 0) ?? E) === T) return;
    const s = this._$AH,
      e = t === E && s !== E || t.capture !== s.capture || t.once !== s.once || t.passive !== s.passive,
      h = t !== E && (s === E || e);
    e && this.element.removeEventListener(this.name, this, s), h && this.element.addEventListener(this.name, this, t), this._$AH = t;
  }
  handleEvent(t) {
    "function" == typeof this._$AH ? this._$AH.call(this.options?.host ?? this.element, t) : this._$AH.handleEvent(t);
  }
}
class z {
  constructor(t, i, s) {
    this.element = t, this.type = 6, this._$AN = void 0, this._$AM = i, this.options = s;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AI(t) {
    S(this, t);
  }
}
const j = t$1.litHtmlPolyfillSupport;
j?.(N, R), (t$1.litHtmlVersions ??= []).push("3.3.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = {
    ATTRIBUTE: 1},
  e$1 = t => (...e) => ({
    _$litDirective$: t,
    values: e
  });
class i {
  constructor(t) {}
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AT(t, e, i) {
    this._$Ct = t, this._$AM = e, this._$Ci = i;
  }
  _$AS(t, e) {
    return this.update(t, e);
  }
  update(t, e) {
    return this.render(...e);
  }
}

/**
 * @license
 * Copyright 2018 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const e = e$1(class extends i {
  constructor(t$1) {
    if (super(t$1), t$1.type !== t.ATTRIBUTE || "class" !== t$1.name || t$1.strings?.length > 2) throw Error("`classMap()` can only be used in the `class` attribute and must be the only part in the attribute.");
  }
  render(t) {
    return " " + Object.keys(t).filter(s => t[s]).join(" ") + " ";
  }
  update(s, [i]) {
    if (void 0 === this.st) {
      this.st = new Set(), void 0 !== s.strings && (this.nt = new Set(s.strings.join(" ").split(/\s/).filter(t => "" !== t)));
      for (const t in i) i[t] && !this.nt?.has(t) && this.st.add(t);
      return this.render(i);
    }
    const r = s.element.classList;
    for (const t of this.st) t in i || (r.remove(t), this.st.delete(t));
    for (const t in i) {
      const s = !!i[t];
      s === this.st.has(t) || this.nt?.has(t) || (s ? (r.add(t), this.st.add(t)) : (r.remove(t), this.st.delete(t)));
    }
    return T;
  }
});

/**
 * @license
 * Copyright 2018 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = o => o ?? E;

/** A base implementation for an element used to convey progress. This class must be inherited. */
class ProgressElementIndicatorBase extends AttachInternals(Role(LitElement, "progressbar"), true) {
  constructor() {
    super(...arguments);
    /**
     * A fractional value, between 0 and `max`, indicating progress.
     * @default 0
     */
    this.value = 0;
    /**
     * The maximum progress value.
     * @default 100
     */
    this.max = 100;
    /**
     * The appearance of the indicator.
     * @default "flat"
     */
    this.variant = "flat";
  }
  /** @inheritdoc */
  connectedCallback() {
    super.connectedCallback();
    this.ariaValueMin = "0";
  }
  /** @inheritdoc */
  update(changedProperties) {
    super.update(changedProperties);
    if (changedProperties.has("value")) {
      this.ariaValueNow = `${this.value}`;
    }
    if (changedProperties.has("max")) {
      this.ariaValueMax = `${this.max}`;
    }
  }
}
/** The styles of the element. */
ProgressElementIndicatorBase.styles = css`@media (forced-colors: active) { :host { --m3e-progress-indicator-color: CanvasText; --m3e-progress-indicator-track-color: Canvas; } }`;
__decorate([n$1({
  type: Number,
  reflect: true
})], ProgressElementIndicatorBase.prototype, "value", void 0);
__decorate([n$1({
  type: Number
})], ProgressElementIndicatorBase.prototype, "max", void 0);
__decorate([n$1({
  reflect: true
})], ProgressElementIndicatorBase.prototype, "variant", void 0);

var _M3eCircularProgressIndicatorElement_instances, _M3eCircularProgressIndicatorElement_maskId, _M3eCircularProgressIndicatorElement_diameter, _M3eCircularProgressIndicatorElement_strokeWidth, _M3eCircularProgressIndicatorElement_amplitude, _M3eCircularProgressIndicatorElement_wavelength, _M3eCircularProgressIndicatorElement_spinnerActiveTrack, _M3eCircularProgressIndicatorElement_spinnerTrack, _M3eCircularProgressIndicatorElement_resizeController, _M3eCircularProgressIndicatorElement_indeterminateWavyAnimationLoop, _M3eCircularProgressIndicatorElement_renderFlatIndicator, _M3eCircularProgressIndicatorElement_renderWavyIndicator, _M3eCircularProgressIndicatorElement_renderResizeObservedElements, _M3eCircularProgressIndicatorElement_renderContent, _M3eCircularProgressIndicatorElement_updateDiameterAndStroke, _M3eCircularProgressIndicatorElement_updateAmplitudeAndWavelength, _M3eCircularProgressIndicatorElement_updateWavyIndeterminateSpinner, _M3eCircularProgressIndicatorElement_computeWavyIndeterminateSweep, _M3eCircularProgressIndicatorElement_sizeToDegrees, _M3eCircularProgressIndicatorElement_degreesToRadians, _M3eCircularProgressIndicatorElement_polarToCartesian, _M3eCircularProgressIndicatorElement_computeCircle, _M3eCircularProgressIndicatorElement_drawArc, _M3eCircularProgressIndicatorElement_drawWavyArc;
var M3eCircularProgressIndicatorElement_1;
const WAVY_INDETERMINATE_DURATION = 1.575;
/**
 * A circular indicator of progress and activity.
 *
 * @description
 * The `m3e-circular-progress-indicator` component displays a circular progress spinner for
 * tracking the completion of a task or process. It supports determinate and indeterminate
 * modes, and can be customized with CSS custom properties for diameter, stroke width, and
 * color. The component is accessible, animates smoothly, and adapts to various use cases including
 * loading and activity indication.
 *
 * @example
 * The following example illustrates a determinate circular progress indicator.
 * ```html
 * <m3e-circular-progress-indicator value="30"></m3e-circular-progress-indicator>
 * ```
 * @example
 * The next example illustrates an indeterminate circular progress indicator using the `indeterminate` attribute.
 * ```html
 * <m3e-circular-progress-indicator indeterminate></m3e-circular-progress-indicator>
 * ```
 *
 * @tag m3e-circular-progress-indicator
 *
 * @slot - Renders the content inside the progress indicator.
 *
 * @attr indeterminate - Whether to show something is happening without conveying progress.
 * @attr max - The maximum progress value.
 * @attr value - A fractional value, between 0 and `max`, indicating progress.
 * @attr variant - The appearance of the indicator.
 *
 * @cssprop --m3e-circular-flat-progress-indicator-diameter - Diameter of the `flat` variant.
 * @cssprop --m3e-circular-wavy-progress-indicator-diameter - Diameter of the `wavy` variant.
 * @cssprop --m3e-circular-wavy-progress-indicator-amplitude - Amplitude of the `wavy` variant.
 * @cssprop --m3e-circular-wavy-progress-indicator-wavelength - Wavelength of the `wavy` variant.
 * @cssprop --m3e-circular-progress-indicator-thickness - Thickness of the progress indicator.
 * @cssprop --m3e-progress-indicator-track-color - Track color of the progress indicator (background).
 * @cssprop --m3e-progress-indicator-color - Color of the progress indicator (foreground).
 */
let M3eCircularProgressIndicatorElement = M3eCircularProgressIndicatorElement_1 = class M3eCircularProgressIndicatorElement extends ProgressElementIndicatorBase {
  constructor() {
    super(...arguments);
    _M3eCircularProgressIndicatorElement_instances.add(this);
    /** @private */
    _M3eCircularProgressIndicatorElement_maskId.set(this, `m3e-circular-progress-mask-${M3eCircularProgressIndicatorElement_1.__nextMaskId++}`);
    /** @private */
    _M3eCircularProgressIndicatorElement_diameter.set(this, 0);
    /** @private */
    _M3eCircularProgressIndicatorElement_strokeWidth.set(this, 0);
    /** @private */
    _M3eCircularProgressIndicatorElement_amplitude.set(this, 0);
    /** @private */
    _M3eCircularProgressIndicatorElement_wavelength.set(this, 0);
    /** @private */
    _M3eCircularProgressIndicatorElement_spinnerActiveTrack.set(this, void 0);
    /** @private */
    _M3eCircularProgressIndicatorElement_spinnerTrack.set(this, void 0);
    /** @private */
    _M3eCircularProgressIndicatorElement_resizeController.set(this, new ResizeController(this, {
      skipInitial: true,
      target: null,
      callback: () => {
        __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_updateDiameterAndStroke).call(this);
        __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_updateAmplitudeAndWavelength).call(this);
      }
    }));
    /** @private */
    _M3eCircularProgressIndicatorElement_indeterminateWavyAnimationLoop.set(this, new AnimationLoopController(this, (_, t) => __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_updateWavyIndeterminateSpinner).call(this, t)));
    /**
     * Whether to show something is happening without conveying progress.
     * @default false
     */
    this.indeterminate = false;
  }
  /** @inheritdoc */
  update(changedProperties) {
    super.update(changedProperties);
    if (changedProperties.has("indeterminate")) {
      this.ariaValueNow = this.indeterminate ? null : `${this.value}`;
    }
  }
  /** @inheritdoc */
  firstUpdated(_changedProperties) {
    super.firstUpdated(_changedProperties);
    const diameterAndStroke = this.shadowRoot?.querySelector(".diameter-and-stroke");
    if (diameterAndStroke) {
      __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_updateDiameterAndStroke).call(this);
      __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_resizeController, "f").observe(diameterAndStroke);
    }
    const amplitudeAndWavelength = this.shadowRoot?.querySelector(".amplitude-and-wavelength");
    if (amplitudeAndWavelength) {
      __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_updateDiameterAndStroke).call(this);
      __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_resizeController, "f").observe(amplitudeAndWavelength);
    }
  }
  /** @inheritdoc */
  updated(_changedProperties) {
    super.updated(_changedProperties);
    if (_changedProperties.has("indeterminate")) {
      __classPrivateFieldSet(this, _M3eCircularProgressIndicatorElement_spinnerActiveTrack, this.shadowRoot?.querySelector(".spinner.active-track"), "f");
      __classPrivateFieldSet(this, _M3eCircularProgressIndicatorElement_spinnerTrack, this.shadowRoot?.querySelector(".spinner.track"), "f");
      if (this.indeterminate) {
        __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_indeterminateWavyAnimationLoop, "f").start();
      } else {
        __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_indeterminateWavyAnimationLoop, "f").stop();
      }
    }
  }
  /** @inheritdoc */
  render() {
    return this.variant === "wavy" ? __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderWavyIndicator).call(this) : __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderFlatIndicator).call(this);
  }
};
_M3eCircularProgressIndicatorElement_maskId = new WeakMap();
_M3eCircularProgressIndicatorElement_diameter = new WeakMap();
_M3eCircularProgressIndicatorElement_strokeWidth = new WeakMap();
_M3eCircularProgressIndicatorElement_amplitude = new WeakMap();
_M3eCircularProgressIndicatorElement_wavelength = new WeakMap();
_M3eCircularProgressIndicatorElement_spinnerActiveTrack = new WeakMap();
_M3eCircularProgressIndicatorElement_spinnerTrack = new WeakMap();
_M3eCircularProgressIndicatorElement_resizeController = new WeakMap();
_M3eCircularProgressIndicatorElement_indeterminateWavyAnimationLoop = new WeakMap();
_M3eCircularProgressIndicatorElement_instances = new WeakSet();
_M3eCircularProgressIndicatorElement_renderFlatIndicator = function _M3eCircularProgressIndicatorElement_renderFlatIndicator() {
  if (this.indeterminate) {
    const left = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
      startAngle: -45,
      endAngle: 90 + __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")
    });
    const right = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
      startAngle: -__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f"),
      endAngle: 135
    });
    return html`<div class="progress" aria-hidden="true"><div class="spinner"><div class="left"><svg width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" height="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" viewBox="${left.viewBox}" class="circle"><path class="active-track" d="${left.path}" stroke="currentColor" stroke-width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}" fill="none" stroke-linecap="round"/></svg></div><div class="right"><svg width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" height="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" viewBox="${right.viewBox}" class="circle"><path class="active-track" d="${right.path}" stroke="currentColor" stroke-width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}" fill="none" stroke-linecap="round"/></svg></div></div></div>${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderResizeObservedElements).call(this)}`;
  }
  const minDegrees = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") * 2, __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f"));
  let degrees = this.value / this.max * 360;
  if (degrees > 0) {
    degrees = Math.max(0, minDegrees, degrees);
  }
  const active = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
    gap: degrees < 360 ? __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") : 0,
    endAngle: degrees
  });
  const inactive = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
    gap: degrees > 0 ? __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") : 0,
    startAngle: degrees,
    endAngle: 360
  });
  return html`<div class="progress" aria-hidden="true"><svg width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" height="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" viewBox="${active.viewBox}">${degrees > 0 ? svg`<path
            class="active-track"
            d="${active.path}"
            stroke="currentColor"
            stroke-width=${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}
            fill="none"
            stroke-linecap="round"
          />` : nothing} ${360 - degrees >= minDegrees ? svg`<path
                class="track"
                d="${inactive.path}"
                stroke="currentColor"
                stroke-width=${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}
                fill="none"
                stroke-linecap="round"
              />` : nothing}</svg></div>${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderResizeObservedElements).call(this)}${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderContent).call(this)}`;
};
_M3eCircularProgressIndicatorElement_renderWavyIndicator = function _M3eCircularProgressIndicatorElement_renderWavyIndicator() {
  if (this.indeterminate) {
    return html`<div class="progress" aria-hidden="true"><svg width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" height="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")}" viewBox="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawWavyArc).call(this, {
      endAngle: 20
    }).viewBox}"><path class="spinner active-track" stroke="currentColor" stroke-width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}" stroke-linecap="round" stroke-linejoin="round" fill="none"/><path class="spinner track" stroke="currentColor" stroke-width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}" fill="none" stroke-linecap="round"/></svg></div>${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderResizeObservedElements).call(this)}`;
  }
  const minDegrees = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") * 2, __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f"));
  let degrees = this.value / this.max * 360;
  if (degrees > 0) {
    degrees = Math.max(0, minDegrees, degrees);
  }
  const amplitude = degrees <= minDegrees + minDegrees / 2 || degrees == 360 ? 0 : __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_amplitude, "f");
  const activeArc = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
    gap: degrees < 360 ? __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") : 0,
    endAngle: degrees
  });
  const active = amplitude == 0 ? activeArc : __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawWavyArc).call(this, {
    endAngle: 360,
    amplitude
  });
  const inactive = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
    gap: degrees > 0 ? __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") : 0,
    startAngle: degrees,
    endAngle: 360
  });
  const padding = amplitude > 0 ? amplitude + __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") / 2 : __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f");
  return html`<svg class="progress" viewBox="${inactive.viewBox}" aria-hidden="true">${degrees > 0 ? svg`${amplitude > 0 ? svg`<defs>
          <mask id="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_maskId, "f")}">
            <path
              d="${activeArc.path}"
              stroke="white"
              stroke-width="${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") + padding}"
              fill="none"
              stroke-linecap="round"
            />
          </mask>
        </defs>` : nothing}
        <g class="active-track" mask="${o(amplitude > 0 ? resolveFragmentUrl(__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_maskId, "f")) : undefined)}">
          <path
            class="${e({
    wave: amplitude > 0
  })}"
            d="${active.path}"
            stroke="currentColor"
            stroke-width=${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}
            stroke-linecap="round"
            stroke-linejoin="round"
            fill="none"
          />
        </g>` : nothing} ${360 - degrees >= minDegrees ? svg`<path
                class="track"
                d="${inactive.path}"
                stroke="currentColor"
                stroke-width=${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")}
                fill="none"
                stroke-linecap="round"
              />` : nothing}</svg> ${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderResizeObservedElements).call(this)}${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_renderContent).call(this)}`;
};
_M3eCircularProgressIndicatorElement_renderResizeObservedElements = function _M3eCircularProgressIndicatorElement_renderResizeObservedElements() {
  return html`<div class="diameter-and-stroke" aria-hidden="true"></div><div class="amplitude-and-wavelength" aria-hidden="true"></div>`;
};
_M3eCircularProgressIndicatorElement_renderContent = function _M3eCircularProgressIndicatorElement_renderContent() {
  return html`<div class="content" aria-hidden="true"><slot></slot></div>`;
};
_M3eCircularProgressIndicatorElement_updateDiameterAndStroke = function _M3eCircularProgressIndicatorElement_updateDiameterAndStroke() {
  const element = this.shadowRoot?.querySelector(".diameter-and-stroke");
  if (element) {
    __classPrivateFieldSet(this, _M3eCircularProgressIndicatorElement_diameter, element.clientWidth, "f");
    __classPrivateFieldSet(this, _M3eCircularProgressIndicatorElement_strokeWidth, element.clientHeight, "f");
  }
};
_M3eCircularProgressIndicatorElement_updateAmplitudeAndWavelength = function _M3eCircularProgressIndicatorElement_updateAmplitudeAndWavelength() {
  const element = this.shadowRoot?.querySelector(".amplitude-and-wavelength");
  if (element) {
    __classPrivateFieldSet(this, _M3eCircularProgressIndicatorElement_amplitude, element.clientWidth, "f");
    __classPrivateFieldSet(this, _M3eCircularProgressIndicatorElement_wavelength, element.clientHeight, "f");
  }
};
_M3eCircularProgressIndicatorElement_updateWavyIndeterminateSpinner = function _M3eCircularProgressIndicatorElement_updateWavyIndeterminateSpinner(t) {
  if (__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_amplitude, "f") === 0 || __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_wavelength, "f") === 0) return;
  const sweep = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_computeWavyIndeterminateSweep).call(this, t);
  __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_spinnerActiveTrack, "f")?.setAttribute("d", __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawWavyArc).call(this, {
    endAngle: sweep
  }).path);
  __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_spinnerTrack, "f")?.setAttribute("d", __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_drawArc).call(this, {
    gap: __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f"), __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f")),
    startAngle: sweep
  }).path);
};
_M3eCircularProgressIndicatorElement_computeWavyIndeterminateSweep = function _M3eCircularProgressIndicatorElement_computeWavyIndeterminateSweep(t) {
  const sweepPadding = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f")) * 2;
  const minSweep = 18 + sweepPadding;
  const maxSweep = 280 - sweepPadding;
  const duration = WAVY_INDETERMINATE_DURATION;
  const holdMin = duration;
  const growTime = duration;
  const holdMax = duration;
  const shrinkTime = duration;
  const cycle = holdMin + growTime + holdMax + shrinkTime;
  const u = t % cycle;
  if (u < holdMin) {
    return minSweep;
  }
  if (u < holdMin + growTime) {
    const p = (u - holdMin) / growTime;
    return minSweep + (maxSweep - minSweep) * (p * p * (3 - 2 * p));
  }
  if (u < holdMin + growTime + holdMax) {
    return maxSweep;
  }
  const p = (u - (holdMin + growTime + holdMax)) / shrinkTime;
  return maxSweep - (maxSweep - minSweep) * (p * p * (3 - 2 * p));
};
_M3eCircularProgressIndicatorElement_sizeToDegrees = function _M3eCircularProgressIndicatorElement_sizeToDegrees(size, padding = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_amplitude, "f")) {
  return size * (360 / (2 * Math.PI * __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_computeCircle).call(this, padding).r));
};
_M3eCircularProgressIndicatorElement_degreesToRadians = function _M3eCircularProgressIndicatorElement_degreesToRadians(degrees) {
  return (degrees - 90) * (Math.PI / 180);
};
_M3eCircularProgressIndicatorElement_polarToCartesian = function _M3eCircularProgressIndicatorElement_polarToCartesian(circle, degrees) {
  const rad = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_degreesToRadians).call(this, degrees);
  return {
    x: circle.cx + circle.r * Math.cos(rad),
    y: circle.cy + circle.r * Math.sin(rad)
  };
};
_M3eCircularProgressIndicatorElement_computeCircle = function _M3eCircularProgressIndicatorElement_computeCircle(padding) {
  padding = padding + __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") / 2;
  const r = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") / 2;
  const cx = r + padding;
  const cy = r + padding;
  return {
    cx,
    cy,
    r,
    padding
  };
};
_M3eCircularProgressIndicatorElement_drawArc = function _M3eCircularProgressIndicatorElement_drawArc({
  startAngle = 0,
  endAngle = 360,
  gap = 0,
  padding = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_amplitude, "f")
}) {
  if (__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") === 0 || __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") === 0) return {
    path: "",
    viewBox: "0 0 0 0"
  };
  const circle = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_computeCircle).call(this, padding);
  if (gap > 0) {
    startAngle += __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, gap, padding);
    endAngle -= __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, gap, padding);
  }
  if (endAngle - startAngle >= 360) {
    endAngle = startAngle + 359.999;
  }
  const start = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_polarToCartesian).call(this, circle, endAngle);
  const end = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_polarToCartesian).call(this, circle, startAngle);
  const path = `M ${start.x} ${start.y} A ${circle.r} ${circle.r} 0 ${endAngle - startAngle <= 180 ? "0" : "1"} 0 ${end.x} ${end.y}`;
  const viewBox = `0 0 ${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") + circle.padding * 2} ${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") + circle.padding * 2}`;
  return {
    path,
    viewBox
  };
};
_M3eCircularProgressIndicatorElement_drawWavyArc = function _M3eCircularProgressIndicatorElement_drawWavyArc({
  startAngle = 0,
  endAngle = 360,
  gap = 0,
  padding = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_amplitude, "f"),
  amplitude = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_amplitude, "f"),
  steps = 200
}) {
  if (__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") === 0 || __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_strokeWidth, "f") === 0) return {
    path: "",
    viewBox: "0 0 0 0"
  };
  const circle = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_computeCircle).call(this, padding);
  if (gap > 0) {
    startAngle += __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, gap, padding);
    endAngle -= __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_sizeToDegrees).call(this, gap, padding);
  }
  const startRad = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_degreesToRadians).call(this, startAngle);
  let endRad = __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_instances, "m", _M3eCircularProgressIndicatorElement_degreesToRadians).call(this, endAngle);
  if (startAngle === endAngle) {
    endRad = startRad;
  } else if (endRad < startRad) {
    endRad += Math.PI * 2;
  }
  const totalAngle = endRad - startRad;
  const waveCount = 2 * Math.PI * circle.r / __classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_wavelength, "f");
  const phase = Math.PI / 2 * (waveCount - 1);
  const points = [];
  for (let i = 0; i <= steps; i++) {
    const t = steps === 0 ? 0 : i / steps;
    const angle = startRad + t * totalAngle;
    const wave = Math.sin(angle * waveCount + phase);
    const radius = circle.r - amplitude * wave;
    const x = radius * Math.cos(angle) + circle.cx;
    const y = radius * Math.sin(angle) + circle.cy;
    points.push([x, y]);
  }
  const path = points.length === 1 ? `M ${points[0][0]},${points[0][1]}` : `M ${points[0][0]},${points[0][1]} ` + points.slice(1).map(([x, y]) => `L ${x},${y}`).join(" ");
  const viewBox = `0 0 ${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") + circle.padding * 2} ${__classPrivateFieldGet(this, _M3eCircularProgressIndicatorElement_diameter, "f") + circle.padding * 2}`;
  return {
    path,
    viewBox
  };
};
/** The styles of the element. */
M3eCircularProgressIndicatorElement.styles = [ProgressElementIndicatorBase.styles, css`:host { display: inline-flex; vertical-align: middle; aspect-ratio: 1; position: relative; align-items: center; justify-content: center; } .progress { --_arc-duration: 1333ms; --_cycle-duration: calc(4 * var(--_arc-duration)); --_linear-rotate-duration: calc(var(--_arc-duration) * 360 / 306); --_indeterminate-easing: cubic-bezier(0.4, 0, 0.2, 1); } .active-track { transition: stroke-dashoffset 500ms cubic-bezier(0, 0, 0.2, 1); } :host([variant="flat"]) .progress { flex: 1; align-self: stretch; pointer-events: none; } .progress, .spinner, .left, .right, .content, .circle { position: absolute; inset: 0; } .content { width: 100%; height: 100%; display: flex; align-items: center; justify-content: center; } .diameter-and-stroke, .amplitude-and-wavelength { visibility: hidden; position: absolute; } .diameter-and-stroke { width: inherit; height: var(--m3e-circular-progress-indicator-thickness, 0.25rem); } .amplitude-and-wavelength { width: var(--m3e-circular-wavy-progress-indicator-amplitude, 0.1rem); height: var(--m3e-circular-wavy-progress-indicator-wavelength, 0.9375rem); } :host([variant="flat"]) { width: var(--m3e-circular-flat-progress-indicator-diameter, 2.5rem); } :host([variant="wavy"]) { width: var(--m3e-circular-wavy-progress-indicator-diameter, 3rem); } :host([variant="flat"][indeterminate]) .progress { animation: linear infinite linear-rotate; animation-duration: var(--_linear-rotate-duration); } :host([variant="flat"][indeterminate]) .spinner { animation: infinite both rotate-arc; animation-duration: var(--_cycle-duration); animation-timing-function: var(--_indeterminate-easing); } :host([variant="wavy"][indeterminate]) .spinner { transform-origin: 50% 50%; animation: wavy-spin ${WAVY_INDETERMINATE_DURATION}s linear infinite; } .left { clip-path: inset(0); inset: 0 50% 0 0; } .right { clip-path: inset(0); inset: 0 0 0 50%; } .circle { animation: expand-arc; animation-iteration-count: infinite; animation-fill-mode: both; animation-duration: var(--_arc-duration), var(--_cycle-duration); animation-timing-function: var(--_indeterminate-easing); } .left .circle { rotate: 135deg; inset: 0 -100% 0 0; } .right .circle { rotate: 100deg; inset: 0 0 0 -100%; animation-delay: calc(-0.5 * var(--_arc-duration)), 0ms; } .track { color: var(--m3e-progress-indicator-track-color, ${DesignToken.color.secondaryContainer}); } .active-track { color: var(--m3e-progress-indicator-color, ${DesignToken.color.primary}); } .wave { animation: spin-reverse 8s linear infinite; transform-origin: 50% 50%; } @keyframes expand-arc { 0% { transform: rotate(265deg); } 50% { transform: rotate(130deg); } 100% { transform: rotate(265deg); } } @keyframes rotate-arc { 12.5% { transform: rotate(135deg); } 25% { transform: rotate(270deg); } 37.5% { transform: rotate(405deg); } 50% { transform: rotate(540deg); } 62.5% { transform: rotate(675deg); } 75% { transform: rotate(810deg); } 87.5% { transform: rotate(945deg); } 100% { transform: rotate(1080deg); } } @keyframes linear-rotate { to { transform: rotate(360deg); } } @keyframes spin-reverse { from { transform: rotate(360deg); } to { transform: rotate(0deg); } } @keyframes spin { from { transform: rotate(0deg); } to { transform: rotate(360deg); } } @keyframes wavy-spin { 0% { transform: rotate(0deg); } 10% { transform: rotate(90deg); } 100% { transform: rotate(360deg); } } @media (forced-colors: active) { :host([variant="flat"]) circle { fill: Canvas; } :host([variant="flat"]) .circle { border-color: var(--m3e-progress-indicator-color, ${DesignToken.color.primary}) var(--m3e-progress-indicator-color, ${DesignToken.color.primary}) Canvas Canvas; } } @media (forced-colors: active) { .progress { --m3e-progress-indicator-track-color: GrayText; --m3e-progress-indicator-color: CanvasText; } }`];
/** @private */
M3eCircularProgressIndicatorElement.__nextMaskId = 0;
__decorate([n$1({
  type: Boolean,
  reflect: true
})], M3eCircularProgressIndicatorElement.prototype, "indeterminate", void 0);
M3eCircularProgressIndicatorElement = M3eCircularProgressIndicatorElement_1 = __decorate([t$3("m3e-circular-progress-indicator")], M3eCircularProgressIndicatorElement);

var _M3eLinearProgressIndicatorElement_instances, _M3eLinearProgressIndicatorElement_maskId, _M3eLinearProgressIndicatorElement_strokeWidth, _M3eLinearProgressIndicatorElement_amplitude, _M3eLinearProgressIndicatorElement_wavelength, _M3eLinearProgressIndicatorElement_resizeController, _M3eLinearProgressIndicatorElement_renderWave, _M3eLinearProgressIndicatorElement_renderResizeObservedElements, _M3eLinearProgressIndicatorElement_updateStroke, _M3eLinearProgressIndicatorElement_updateAmplitudeAndWavelength, _M3eLinearProgressIndicatorElement_drawWavyPath;
var M3eLinearProgressIndicatorElement_1;
/**
 * A horizontal bar for indicating progress and activity.
 *
 * @description
 * The `m3e-linear-progress-indicator` component displays a horizontal progress bar for tracking
 * the completion of a task or process. It supports `determinate`, `indeterminate`, `buffer`,
 * and `query` modes, and can be customized with CSS custom properties for thickness, shape, and color.
 * The component is accessible, animates smoothly, and adapts to various use cases including loading,
 * buffering, and activity indication.
 *
 * @example
 * The following example illustrates a determinate linear progress indicator.
 * ```html
 * <m3e-linear-progress-indicator value="30"></m3e-linear-progress-indicator>
 * ```
 * @example
 * The next example illustrates an indeterminate linear progress indicator using the `mode` attribute.
 * ```html
 * <m3e-linear-progress-indicator mode="indeterminate"></m3e-linear-progress-indicator>
 * ```
 *
 * @tag m3e-linear-progress-indicator
 *
 * @attr buffer-value - A fractional value, between 0 and `max`, indicating buffer progress.
 * @attr max - The maximum progress value.
 * @attr mode - The mode of the progress bar.
 * @attr value - A fractional value, between 0 and `max`, indicating progress.
 * @attr variant - The appearance of the indicator.
 *
 * @cssprop --m3e-linear-progress-indicator-thickness - Thickness (height) of the progress bar.
 * @cssprop --m3e-linear-progress-indicator-shape - Border radius of the progress bar.
 * @cssprop --m3e-progress-indicator-track-color - Track color of the progress bar (background/buffer).
 * @cssprop --m3e-progress-indicator-color - Color of the progress indicator (foreground).
 * @cssprop --m3e-linear-wavy-progress-indicator-amplitude - Amplitude of the `wavy` variant.
 * @cssprop --m3e-linear-wavy-progress-indicator-wavelength - Wavelength of the `wavy` variant.
 * @cssprop --m3e-linear-wavy-indeterminate-progress-indicator-wavelength - Wavelength of the indeterminate/query `wavy` variant.
 */
let M3eLinearProgressIndicatorElement = M3eLinearProgressIndicatorElement_1 = class M3eLinearProgressIndicatorElement extends ProgressElementIndicatorBase {
  constructor() {
    super(...arguments);
    _M3eLinearProgressIndicatorElement_instances.add(this);
    /** @private */
    _M3eLinearProgressIndicatorElement_maskId.set(this, `m3e-linear-progress-mask-${M3eLinearProgressIndicatorElement_1.__nextMaskId++}`);
    /** @private */
    _M3eLinearProgressIndicatorElement_strokeWidth.set(this, 0);
    /** @private */
    _M3eLinearProgressIndicatorElement_amplitude.set(this, 0);
    /** @private */
    _M3eLinearProgressIndicatorElement_wavelength.set(this, 0);
    /** @private */
    _M3eLinearProgressIndicatorElement_resizeController.set(this, new ResizeController(this, {
      skipInitial: true,
      target: null,
      callback: () => {
        __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_updateStroke).call(this);
        __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_updateAmplitudeAndWavelength).call(this);
      }
    }));
    /**
     * The mode of the progress bar.
     * @default "determinate"
     */
    this.mode = "determinate";
    /**
     * A fractional value, between 0 and `max`, indicating buffer progress.
     * @default 0
     */
    this.bufferValue = 0;
  }
  /** @inheritdoc */
  firstUpdated(_changedProperties) {
    super.firstUpdated(_changedProperties);
    __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_updateStroke).call(this);
    __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_updateAmplitudeAndWavelength).call(this);
    const stroke = this.shadowRoot?.querySelector(".stroke");
    if (stroke) {
      __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_resizeController, "f").observe(stroke);
    }
    const amplitudeAndWavelength = this.shadowRoot?.querySelector(".amplitude-and-wavelength");
    if (amplitudeAndWavelength) {
      __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_resizeController, "f").observe(amplitudeAndWavelength);
    }
  }
  /** @inheritdoc */
  update(changedProperties) {
    super.update(changedProperties);
    if (changedProperties.has("mode")) {
      this.ariaValueNow = this.mode === "indeterminate" || this.mode === "query" ? null : `${this.value}`;
    }
  }
  /** @inheritdoc */
  updated(_changedProperties) {
    super.updated(_changedProperties);
    if (_changedProperties.has("value") || _changedProperties.has("bufferValue") || _changedProperties.has("max")) {
      const progress = this.shadowRoot?.querySelector(".progress");
      progress?.style.setProperty("--_value", `${this.value / this.max * 100}%`);
      progress?.style.setProperty("--_buffer-value", `${this.bufferValue / this.max * 100}%`);
    }
  }
  /** @inheritdoc */
  render() {
    const clampedValue = Math.max(0, Math.min(this.value, this.max));
    const activeWidth = clampedValue / this.max * this.clientWidth;
    const waveWidth = this.mode === "indeterminate" || this.mode === "query" ? this.clientWidth : activeWidth + __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_wavelength, "f") * 3;
    const wave = this.variant === "wavy" && __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_strokeWidth, "f") > 0 && __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_amplitude, "f") > 0 && __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_wavelength, "f") > 0 ? __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_drawWavyPath).call(this, waveWidth) : undefined;
    return html`<div class="progress" aria-hidden="true" style="${safeStyleMap({
      "--_translate-x": `-${activeWidth}px`
    })}">${!wave ? html`<div class="track"></div>` : nothing} ${(this.mode === "determinate" || this.mode === "buffer") && this.value <= 0 ? nothing : html`<div class="primary">${wave && (this.mode === "determinate" || this.mode === "buffer") ? __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_renderWave).call(this, waveWidth, wave.height, wave.viewBox, wave.path) : nothing}</div>${this.mode === "determinate" || this.mode === "buffer" ? html`<div class="gap"></div>` : nothing}`} ${this.mode !== "buffer" || this.bufferValue > 0 ? html`<div class="secondary"></div>` : nothing} ${this.mode === "buffer" && this.bufferValue > 0 ? html`<div class="gap"></div>` : nothing} ${this.mode === "buffer" ? html`<div class="buffer"></div>` : nothing} ${(this.mode === "determinate" || this.mode === "buffer") && this.value > 0 ? html`<div class="gap"></div><div class="stop"></div>` : nothing} ${wave && !(this.mode === "determinate" || this.mode === "buffer") ? __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_renderWave).call(this, waveWidth, wave.height, wave.viewBox, wave.path) : nothing}</div>${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_instances, "m", _M3eLinearProgressIndicatorElement_renderResizeObservedElements).call(this)}`;
  }
};
_M3eLinearProgressIndicatorElement_maskId = new WeakMap();
_M3eLinearProgressIndicatorElement_strokeWidth = new WeakMap();
_M3eLinearProgressIndicatorElement_amplitude = new WeakMap();
_M3eLinearProgressIndicatorElement_wavelength = new WeakMap();
_M3eLinearProgressIndicatorElement_resizeController = new WeakMap();
_M3eLinearProgressIndicatorElement_instances = new WeakSet();
_M3eLinearProgressIndicatorElement_renderWave = function _M3eLinearProgressIndicatorElement_renderWave(width, height, viewBox, path) {
  const masked = this.mode === "indeterminate" || this.mode === "query";
  if (!masked && this.value / this.max === 1) {
    return html`<div class="complete"></div>`;
  }
  return html`<svg class="wave" width="${width}" height="${height}" viewBox="${viewBox}" preserveAspectRatio="none">${masked ? svg`
        <mask id="${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_maskId, "f")}" maskUnits="userSpaceOnUse">
          <rect width="${width}" height="${height}" fill="black" />
          <rect class="primary" height="${height}" fill="white" />
          <rect class="secondary" height="${height}" fill="white" />
        </mask>
        <mask id="${`${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_maskId, "f")}-inverse`}" maskUnits="userSpaceOnUse">
          <rect width="${width}" height="${height}" fill="white" />
          <rect class="primary" height="${height}" fill="black" />
          <rect class="secondary" height="${height}" fill="black" />
        </mask>
        <g mask="${resolveFragmentUrl(__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_maskId, "f"))}">
          <path d="${path}" stroke="currentColor" stroke-width=${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_strokeWidth, "f")} fill="none" stroke-linecap="round" />
        </g>
        <g mask="${resolveFragmentUrl(`${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_maskId, "f")}-inverse`)}">
          <rect class="track" width="100%" height="${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_strokeWidth, "f")}" fill="currentColor" />
        </g>` : svg`<path d="${path}" stroke="currentColor" stroke-width=${__classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_strokeWidth, "f")} fill="none" stroke-linecap="round" />`}</svg>`;
};
_M3eLinearProgressIndicatorElement_renderResizeObservedElements = function _M3eLinearProgressIndicatorElement_renderResizeObservedElements() {
  return html`<div class="stroke" aria-hidden="true"></div><div class="amplitude-and-wavelength" aria-hidden="true"></div>`;
};
_M3eLinearProgressIndicatorElement_updateStroke = function _M3eLinearProgressIndicatorElement_updateStroke() {
  const element = this.shadowRoot?.querySelector(".stroke");
  if (element) {
    __classPrivateFieldSet(this, _M3eLinearProgressIndicatorElement_strokeWidth, element.clientHeight, "f");
  }
};
_M3eLinearProgressIndicatorElement_updateAmplitudeAndWavelength = function _M3eLinearProgressIndicatorElement_updateAmplitudeAndWavelength() {
  const element = this.shadowRoot?.querySelector(".amplitude-and-wavelength");
  if (element) {
    __classPrivateFieldSet(this, _M3eLinearProgressIndicatorElement_amplitude, element.clientHeight, "f");
    __classPrivateFieldSet(this, _M3eLinearProgressIndicatorElement_wavelength, element.clientWidth, "f");
  }
};
_M3eLinearProgressIndicatorElement_drawWavyPath = function _M3eLinearProgressIndicatorElement_drawWavyPath(width, phase = 0) {
  const amplitude = __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_amplitude, "f") + __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_strokeWidth, "f") / 2;
  const y = amplitude;
  const path = [];
  const step = __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_wavelength, "f") / 2;
  let x = 0;
  path.push(`M ${x},${y}`);
  while (x <= width) {
    const endX = x + step;
    const endY = y + amplitude * Math.sin(2 * Math.PI * endX / __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_wavelength, "f") + phase);
    const cpX = x + step / 2;
    const cpY = y + amplitude * Math.sin(2 * Math.PI * (x + step / 2) / __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_wavelength, "f") + phase);
    path.push(`Q ${cpX},${cpY} ${endX},${endY}`);
    x += step;
  }
  const padding = 1;
  const viewBox = `0 ${-padding} ${width} ${amplitude * 2 + padding * 2}`;
  return {
    path: path.join(" "),
    viewBox,
    height: __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_strokeWidth, "f") + __classPrivateFieldGet(this, _M3eLinearProgressIndicatorElement_amplitude, "f") * 2,
    padding
  };
};
/** The styles of the element. */
M3eLinearProgressIndicatorElement.styles = [ProgressElementIndicatorBase.styles, css`:host { display: block; position: relative; } .progress { width: 100%; height: 100%; position: relative; align-items: center; border-radius: var(--m3e-linear-progress-indicator-shape, ${DesignToken.shape.corner.extraSmall}); } .stroke, .amplitude-and-wavelength { visibility: hidden; position: absolute; } .stroke { width: 100%; height: var(--m3e-linear-progress-indicator-thickness, 0.25rem); } .amplitude-and-wavelength { height: var(--m3e-linear-wavy-progress-indicator-amplitude, 0.1875rem); width: var(--m3e-linear-wavy-progress-indicator-wavelength, 2.5rem); } :host([mode="indeterminate"]) .amplitude-and-wavelength, :host([mode="query"]) .amplitude-and-wavelength { width: var(--m3e-linear-wavy-indeterminate-progress-indicator-wavelength, 1.5rem); } .primary, .secondary, .stop { height: var(--m3e-linear-progress-indicator-thickness, 0.25rem); border-radius: inherit; } .stop { aspect-ratio: 1; flex: none; } :host([variant="flat"]) { height: var(--m3e-linear-progress-indicator-thickness, 0.25rem); } :host([variant="wavy"]) { height: calc( var(--m3e-linear-progress-indicator-thickness, 0.25rem) + calc(var(--m3e-linear-wavy-progress-indicator-amplitude, 0.1875rem) * 2) ); } :host([variant="wavy"]) .primary, :host([variant="wavy"]) .secondary { position: relative; height: 100%; overflow: hidden; } :host([variant="wavy"]) .complete { position: absolute; margin: auto; top: calc(50% - calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) / 2)); left: 0; right: 0; height: var(--m3e-linear-progress-indicator-thickness, 0.25rem); border-radius: inherit; } :host([variant="wavy"]) .secondary { height: var(--m3e-linear-progress-indicator-thickness, 0.25rem); } .wave { position: absolute; display: block; } .primary .wave, .secondary .wave { margin-inline-start: calc(0px - var(--m3e-linear-wavy-progress-indicator-wavelength, 2.5rem)); } :host([variant="wavy"][mode="determinate"]) .primary path, :host([variant="wavy"][mode="buffer"]) .primary path { animation: wave-slide 1.5s linear infinite; } @keyframes wave-slide { from { transform: translateX(0); } to { transform: translateX(calc(0px - var(--m3e-linear-wavy-progress-indicator-wavelength, 2.5rem))); } } :host([mode="determinate"]) .progress, :host([mode="buffer"]) .progress { display: flex; overflow: hidden; } :host([mode="determinate"]) .primary, :host([mode="buffer"]) .primary { width: var(--_value, 0px); flex: none; } :host([mode="determinate"]) .gap, :host([mode="buffer"]) .gap { flex-basis: var(--m3e-linear-progress-indicator-thickness, 0.25rem); flex-shrink: 1; } :host([mode="determinate"]) .secondary, :host([mode="buffer"]) .buffer { flex: 1 1 auto; } :host([mode="buffer"]) .buffer { flex-shrink: 5; height: 100%; width: 100%; background-color: var(--m3e-progress-indicator-track-color, ${DesignToken.color.secondaryContainer}); mask-image: radial-gradient( circle, black 0, black calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) / 2), transparent calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) / 2) ); mask-size: calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) * 2) 100%; mask-repeat: repeat; animation: buffer 250ms linear infinite; } :host(:dir(rtl)[mode="buffer"]) .buffer { transform: scaleX(-1); } @keyframes buffer { from { mask-position: 0 0; } to { mask-position: calc(-1 * calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) * 2)) 0; } } :host([mode="buffer"]) .secondary { width: var(--_buffer-value, 0px); flex: none; } :host([mode="indeterminate"]) .primary, :host([mode="query"]) .primary { position: absolute; top: 0; height: 100%; border-radius: inherit; animation: indeterminate-primary 2.1s infinite linear; } :host([variant="wavy"][mode="indeterminate"]) .primary, :host([variant="wavy"][mode="query"]) .primary { animation-name: wavy-indeterminate-primary; } :host([mode="indeterminate"]) .secondary, :host([mode="query"]) .secondary { position: absolute; top: 0; height: 100%; border-radius: inherit; animation: indeterminate-secondary 2.1s infinite linear; animation-delay: 1.15s; animation-fill-mode: backwards; } :host([variant="wavy"][mode="indeterminate"]) .secondary, :host([variant="wavy"][mode="query"]) .secondary { animation-name: wavy-indeterminate-secondary; } :host([mode="indeterminate"]) .progress, :host([mode="query"]) .progress { overflow: hidden; position: relative; } :host(:not(:dir(rtl))[mode="query"]) .progress, :host(:dir(rtl)[mode="indeterminate"]) .progress { transform: scaleX(-1); } :host([variant="flat"]) .primary, :host([variant="flat"][mode="indeterminate"]) .secondary, :host([variant="flat"][mode="query"]) .secondary, :host([variant="wavy"]) .complete, .stop { background-color: var(--m3e-progress-indicator-color, ${DesignToken.color.primary}); } :host([variant="wavy"]) .progress { color: var(--m3e-progress-indicator-color, ${DesignToken.color.primary}); } :host([mode="determinate"]) .secondary, :host([mode="buffer"]) .secondary, :host([variant="flat"][mode="indeterminate"]) .track, :host([variant="flat"][mode="query"]) .track { background-color: var(--m3e-progress-indicator-track-color, ${DesignToken.color.secondaryContainer}); } :host([variant="wavy"][mode="indeterminate"]) .track, :host([variant="wavy"][mode="query"]) .track { color: var(--m3e-progress-indicator-track-color, ${DesignToken.color.secondaryContainer}); } :host([variant="wavy"][mode="indeterminate"]) .track, :host([variant="wavy"][mode="query"]) .track { y: calc(50% - calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) / 2)); border-radius: inherit; } :host([variant="flat"][mode="indeterminate"]) .track, :host([variant="flat"][mode="query"]) .track { position: absolute; margin: auto; top: calc(50% - calc(var(--m3e-linear-progress-indicator-thickness, 0.25rem) / 2)); left: 0; right: 0; height: var(--m3e-linear-progress-indicator-thickness, 0.25rem); border-radius: inherit; } @keyframes indeterminate-primary { 0% { left: -145.167%; width: 8%; } 20% { left: -113.333%; width: 48%; } 60% { left: 56.333%; width: 78%; } 100% { left: 100%; width: 8%; } } @keyframes indeterminate-secondary { 0% { left: -54.888%; width: 8%; } 20% { left: -20%; width: 48%; } 60% { left: 60%; width: 78%; } 100% { left: 160%; width: 8%; } } @keyframes wavy-indeterminate-primary { 0% { transform: translateX(-145.167%); width: 8%; } 20% { transform: translateX(-113.333%); width: 48%; } 60% { transform: translateX(56.333%); width: 78%; } 100% { transform: translateX(100%); width: 8%; } } @keyframes wavy-indeterminate-secondary { 0% { transform: translateX(-54.888%); width: 8%; } 20% { transform: translateX(-20%); width: 48%; } 60% { transform: translateX(60%); width: 78%; } 100% { transform: translateX(160%); width: 8%; } } @media (forced-colors: active) { .progress { --m3e-progress-indicator-track-color: GrayText; --m3e-progress-indicator-color: CanvasText; } }`];
/** @private */
M3eLinearProgressIndicatorElement.__nextMaskId = 0;
__decorate([n$1({
  reflect: true
})], M3eLinearProgressIndicatorElement.prototype, "mode", void 0);
__decorate([n$1({
  attribute: "buffer-value",
  type: Number,
  reflect: true
})], M3eLinearProgressIndicatorElement.prototype, "bufferValue", void 0);
M3eLinearProgressIndicatorElement = M3eLinearProgressIndicatorElement_1 = __decorate([t$3("m3e-linear-progress-indicator")], M3eLinearProgressIndicatorElement);

export { M3eCircularProgressIndicatorElement, M3eLinearProgressIndicatorElement, ProgressElementIndicatorBase };
