// @m3e/web/dist/textarea-autosize.js
/**
 * @license MIT
 * Copyright (c) 2025 matraic
 * See LICENSE file in the project root for full license text.
 */
import { LitElement, css } from 'lit';
import { HtmlFor, Role, debounce } from '@m3e/web/core';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */

function __decorate(decorators, target, key, desc) {
  var c = arguments.length,
    r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
    d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function __classPrivateFieldGet(receiver, state, kind, f) {
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}
function __classPrivateFieldSet(receiver, state, value, kind, f) {
  if (kind === "m") throw new TypeError("Private method is not writable");
  if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
  if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value), value;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = t => (e, o) => {
  void 0 !== o ? o.addInitializer(() => {
    customElements.define(t, e);
  }) : customElements.define(t, e);
};

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t = globalThis,
  e$1 = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
  s = Symbol(),
  o$2 = new WeakMap();
let n$2 = class n {
  constructor(t, e, o) {
    if (this._$cssResult$ = true, o !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t, this.t = e;
  }
  get styleSheet() {
    let t = this.o;
    const s = this.t;
    if (e$1 && void 0 === t) {
      const e = void 0 !== s && 1 === s.length;
      e && (t = o$2.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), e && o$2.set(s, t));
    }
    return t;
  }
  toString() {
    return this.cssText;
  }
};
const r$2 = t => new n$2("string" == typeof t ? t : t + "", void 0, s),
  S = (s, o) => {
    if (e$1) s.adoptedStyleSheets = o.map(t => t instanceof CSSStyleSheet ? t : t.styleSheet);else for (const e of o) {
      const o = document.createElement("style"),
        n = t.litNonce;
      void 0 !== n && o.setAttribute("nonce", n), o.textContent = e.cssText, s.appendChild(o);
    }
  },
  c$1 = e$1 ? t => t : t => t instanceof CSSStyleSheet ? (t => {
    let e = "";
    for (const s of t.cssRules) e += s.cssText;
    return r$2(e);
  })(t) : t;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const {
    is: i,
    defineProperty: e,
    getOwnPropertyDescriptor: h,
    getOwnPropertyNames: r$1,
    getOwnPropertySymbols: o$1,
    getPrototypeOf: n$1
  } = Object,
  a = globalThis,
  c = a.trustedTypes,
  l = c ? c.emptyScript : "",
  p = a.reactiveElementPolyfillSupport,
  d = (t, s) => t,
  u = {
    toAttribute(t, s) {
      switch (s) {
        case Boolean:
          t = t ? l : null;
          break;
        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }
      return t;
    },
    fromAttribute(t, s) {
      let i = t;
      switch (s) {
        case Boolean:
          i = null !== t;
          break;
        case Number:
          i = null === t ? null : Number(t);
          break;
        case Object:
        case Array:
          try {
            i = JSON.parse(t);
          } catch (t) {
            i = null;
          }
      }
      return i;
    }
  },
  f = (t, s) => !i(t, s),
  b = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    useDefault: false,
    hasChanged: f
  };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= new WeakMap();
class y extends HTMLElement {
  static addInitializer(t) {
    this._$Ei(), (this.l ??= []).push(t);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t, s = b) {
    if (s.state && (s.attribute = false), this._$Ei(), this.prototype.hasOwnProperty(t) && ((s = Object.create(s)).wrapped = true), this.elementProperties.set(t, s), !s.noAccessor) {
      const i = Symbol(),
        h = this.getPropertyDescriptor(t, i, s);
      void 0 !== h && e(this.prototype, t, h);
    }
  }
  static getPropertyDescriptor(t, s, i) {
    const {
      get: e,
      set: r
    } = h(this.prototype, t) ?? {
      get() {
        return this[s];
      },
      set(t) {
        this[s] = t;
      }
    };
    return {
      get: e,
      set(s) {
        const h = e?.call(this);
        r?.call(this, s), this.requestUpdate(t, h, i);
      },
      configurable: true,
      enumerable: true
    };
  }
  static getPropertyOptions(t) {
    return this.elementProperties.get(t) ?? b;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties"))) return;
    const t = n$1(this);
    t.finalize(), void 0 !== t.l && (this.l = [...t.l]), this.elementProperties = new Map(t.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized"))) return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t = this.properties,
        s = [...r$1(t), ...o$1(t)];
      for (const i of s) this.createProperty(i, t[i]);
    }
    const t = this[Symbol.metadata];
    if (null !== t) {
      const s = litPropertyMetadata.get(t);
      if (void 0 !== s) for (const [t, i] of s) this.elementProperties.set(t, i);
    }
    this._$Eh = new Map();
    for (const [t, s] of this.elementProperties) {
      const i = this._$Eu(t, s);
      void 0 !== i && this._$Eh.set(i, t);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s) {
    const i = [];
    if (Array.isArray(s)) {
      const e = new Set(s.flat(1 / 0).reverse());
      for (const s of e) i.unshift(c$1(s));
    } else void 0 !== s && i.push(c$1(s));
    return i;
  }
  static _$Eu(t, s) {
    const i = s.attribute;
    return false === i ? void 0 : "string" == typeof i ? i : "string" == typeof t ? t.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise(t => this.enableUpdating = t), this._$AL = new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach(t => t(this));
  }
  addController(t) {
    (this._$EO ??= new Set()).add(t), void 0 !== this.renderRoot && this.isConnected && t.hostConnected?.();
  }
  removeController(t) {
    this._$EO?.delete(t);
  }
  _$E_() {
    const t = new Map(),
      s = this.constructor.elementProperties;
    for (const i of s.keys()) this.hasOwnProperty(i) && (t.set(i, this[i]), delete this[i]);
    t.size > 0 && (this._$Ep = t);
  }
  createRenderRoot() {
    const t = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t, this.constructor.elementStyles), t;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach(t => t.hostConnected?.());
  }
  enableUpdating(t) {}
  disconnectedCallback() {
    this._$EO?.forEach(t => t.hostDisconnected?.());
  }
  attributeChangedCallback(t, s, i) {
    this._$AK(t, i);
  }
  _$ET(t, s) {
    const i = this.constructor.elementProperties.get(t),
      e = this.constructor._$Eu(t, i);
    if (void 0 !== e && true === i.reflect) {
      const h = (void 0 !== i.converter?.toAttribute ? i.converter : u).toAttribute(s, i.type);
      this._$Em = t, null == h ? this.removeAttribute(e) : this.setAttribute(e, h), this._$Em = null;
    }
  }
  _$AK(t, s) {
    const i = this.constructor,
      e = i._$Eh.get(t);
    if (void 0 !== e && this._$Em !== e) {
      const t = i.getPropertyOptions(e),
        h = "function" == typeof t.converter ? {
          fromAttribute: t.converter
        } : void 0 !== t.converter?.fromAttribute ? t.converter : u;
      this._$Em = e;
      const r = h.fromAttribute(s, t.type);
      this[e] = r ?? this._$Ej?.get(e) ?? r, this._$Em = null;
    }
  }
  requestUpdate(t, s, i) {
    if (void 0 !== t) {
      const e = this.constructor,
        h = this[t];
      if (i ??= e.getPropertyOptions(t), !((i.hasChanged ?? f)(h, s) || i.useDefault && i.reflect && h === this._$Ej?.get(t) && !this.hasAttribute(e._$Eu(t, i)))) return;
      this.C(t, s, i);
    }
    false === this.isUpdatePending && (this._$ES = this._$EP());
  }
  C(t, s, {
    useDefault: i,
    reflect: e,
    wrapped: h
  }, r) {
    i && !(this._$Ej ??= new Map()).has(t) && (this._$Ej.set(t, r ?? s ?? this[t]), true !== h || void 0 !== r) || (this._$AL.has(t) || (this.hasUpdated || i || (s = void 0), this._$AL.set(t, s)), true === e && this._$Em !== t && (this._$Eq ??= new Set()).add(t));
  }
  async _$EP() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t) {
      Promise.reject(t);
    }
    const t = this.scheduleUpdate();
    return null != t && (await t), !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending) return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t, s] of this._$Ep) this[t] = s;
        this._$Ep = void 0;
      }
      const t = this.constructor.elementProperties;
      if (t.size > 0) for (const [s, i] of t) {
        const {
            wrapped: t
          } = i,
          e = this[s];
        true !== t || this._$AL.has(s) || void 0 === e || this.C(s, void 0, i, e);
      }
    }
    let t = false;
    const s = this._$AL;
    try {
      t = this.shouldUpdate(s), t ? (this.willUpdate(s), this._$EO?.forEach(t => t.hostUpdate?.()), this.update(s)) : this._$EM();
    } catch (s) {
      throw t = false, this._$EM(), s;
    }
    t && this._$AE(s);
  }
  willUpdate(t) {}
  _$AE(t) {
    this._$EO?.forEach(t => t.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t)), this.updated(t);
  }
  _$EM() {
    this._$AL = new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t) {
    return true;
  }
  update(t) {
    this._$Eq &&= this._$Eq.forEach(t => this._$ET(t, this[t])), this._$EM();
  }
  updated(t) {}
  firstUpdated(t) {}
}
y.elementStyles = [], y.shadowRootOptions = {
  mode: "open"
}, y[d("elementProperties")] = new Map(), y[d("finalized")] = new Map(), p?.({
  ReactiveElement: y
}), (a.reactiveElementVersions ??= []).push("2.1.1");

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const o = {
    attribute: true,
    type: String,
    converter: u,
    reflect: false,
    hasChanged: f
  },
  r = (t = o, e, r) => {
    const {
      kind: n,
      metadata: i
    } = r;
    let s = globalThis.litPropertyMetadata.get(i);
    if (void 0 === s && globalThis.litPropertyMetadata.set(i, s = new Map()), "setter" === n && ((t = Object.create(t)).wrapped = true), s.set(r.name, t), "accessor" === n) {
      const {
        name: o
      } = r;
      return {
        set(r) {
          const n = e.get.call(this);
          e.set.call(this, r), this.requestUpdate(o, n, t);
        },
        init(e) {
          return void 0 !== e && this.C(o, void 0, t, e), e;
        }
      };
    }
    if ("setter" === n) {
      const {
        name: o
      } = r;
      return function (r) {
        const n = this[o];
        e.call(this, r), this.requestUpdate(o, n, t);
      };
    }
    throw Error("Unsupported decorator location: " + n);
  };
function n(t) {
  return (e, o) => "object" == typeof o ? r(t, e, o) : ((t, e, o) => {
    const r = e.hasOwnProperty(o);
    return e.constructor.createProperty(o, t), r ? Object.getOwnPropertyDescriptor(e, o) : void 0;
  })(t, e, o);
}

/**
 * Adapted from Angular Material CDK Text Field
 * Source: https://github.com/angular/components/blob/main/src/cdk/text-field/autosize.ts
 *
 * @license MIT
 * Copyright (c) 2025 Google LLC
 * See LICENSE file in the project root for full license text.
 */
var _M3eTextareaAutosizeElement_instances, _M3eTextareaAutosizeElement_initialHeight, _M3eTextareaAutosizeElement_cachedLineHeight, _M3eTextareaAutosizeElement_cachedPlaceholderHeight, _M3eTextareaAutosizeElement_previousMinRows, _M3eTextareaAutosizeElement_previousValue, _M3eTextareaAutosizeElement_hasFocus, _M3eTextareaAutosizeElement_windowResizeHandler, _M3eTextareaAutosizeElement_focusHandler, _M3eTextareaAutosizeElement_inputHandler, _M3eTextareaAutosizeElement_cacheTextareaLineHeight, _M3eTextareaAutosizeElement_cacheTextareaPlaceholderHeight, _M3eTextareaAutosizeElement_setMinHeight, _M3eTextareaAutosizeElement_setMaxHeight, _M3eTextareaAutosizeElement_measureScrollHeight, _M3eTextareaAutosizeElement_scrollToCaretPosition;
/**
 * A non-visual element used to automatically resize a `textarea` to fit its content.
 *
 * @description
 * The `m3e-textarea-autosize` component automatically adjusts the height of a linked `textarea` to fit its content,
 * preserving layout integrity and user experience. This non-visual element listens to input changes and applies
 * dynamic resizing, constrained by optional row limits. It supports declarative configuration via attributes and
 * can be disabled when manual control is preferred.
 *
 * @example
 * The following example illustrates the `m3e-textarea-autosize` in conjunction with the `m3e-form-field` to
 * automatically resize a field's `textarea` with a 5 row limit.
 * ```html
 * <m3e-form-field>
 *  <label slot="label" for="fld">Textarea Autosize</label>
 *  <textarea id="fld"></textarea>
 *  <m3e-textarea-autosize for="fld" max-rows="5"></m3e-textarea-autosize>
 * </m3e-form-field>
 * ```
 *
 * @tag m3e-textarea-autosize
 *
 * @attr disabled - Whether auto-sizing is disabled.
 * @attr for - The identifier of the interactive control to which this element is attached.
 * @attr max-rows - The maximum amount of rows in the `textarea`.
 * @attr min-rows - The minimum amount of rows in the `textarea`.
 */
let M3eTextareaAutosizeElement = class M3eTextareaAutosizeElement extends HtmlFor(Role(LitElement, "none")) {
  constructor() {
    super(...arguments);
    _M3eTextareaAutosizeElement_instances.add(this);
    /** @private */
    _M3eTextareaAutosizeElement_initialHeight.set(this, void 0);
    /** @private */
    _M3eTextareaAutosizeElement_cachedLineHeight.set(this, void 0);
    /** @private */
    _M3eTextareaAutosizeElement_cachedPlaceholderHeight.set(this, void 0);
    /** @private */
    _M3eTextareaAutosizeElement_previousMinRows.set(this, void 0);
    /** @private */
    _M3eTextareaAutosizeElement_previousValue.set(this, void 0);
    /** @private */
    _M3eTextareaAutosizeElement_hasFocus.set(this, false);
    /** @private */
    _M3eTextareaAutosizeElement_windowResizeHandler.set(this, () => this._handleWindowResize());
    /** @private */
    _M3eTextareaAutosizeElement_focusHandler.set(this, e => __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_hasFocus, e.type === "focus", "f"));
    /** @private */
    _M3eTextareaAutosizeElement_inputHandler.set(this, () => this.resizeToFitContent());
    /**
     * The maximum amount of rows in the `textarea`.
     * @default 0
     */
    this.maxRows = 0;
    /**
     * The minimum amount of rows in the `textarea`.
     * @default 0
     */
    this.minRows = 0;
    /**
     * Whether auto-sizing is disabled.
     * @default false
     */
    this.disabled = false;
  }
  /** @inheritdoc */
  attach(control) {
    super.attach(control);
    if (control instanceof HTMLTextAreaElement) {
      control.style.resize = "none";
      __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_initialHeight, control.style.height, "f");
      control.addEventListener("focus", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_focusHandler, "f"));
      control.addEventListener("blur", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_focusHandler, "f"));
      control.addEventListener("input", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_inputHandler, "f"));
      window.addEventListener("resize", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_windowResizeHandler, "f"));
      this.resizeToFitContent(true);
    }
  }
  /** @inheritdoc */
  detach() {
    if (this.control instanceof HTMLTextAreaElement) {
      window.removeEventListener("resize", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_windowResizeHandler, "f"));
      this.control.removeEventListener("focus", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_focusHandler, "f"));
      this.control.removeEventListener("blur", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_focusHandler, "f"));
      this.control.removeEventListener("input", __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_inputHandler, "f"));
    }
    super.detach();
  }
  /** @inheritdoc */
  connectedCallback() {
    this.ariaHidden = "true";
    super.connectedCallback();
  }
  /** @inheritdoc */
  updated(_changedProperties) {
    super.updated(_changedProperties);
    if (_changedProperties.has("disabled")) {
      if (this.disabled) {
        this.reset();
      } else {
        this.resizeToFitContent(true);
      }
    }
  }
  /**
   * Resize the `textarea` to fit its content.
   * @param {boolean} [force=false] - Whether to force a height recalculation.
   */
  resizeToFitContent(force = false) {
    if (this.disabled || !(this.control instanceof HTMLTextAreaElement)) {
      return;
    }
    __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_cacheTextareaLineHeight).call(this);
    __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_cacheTextareaPlaceholderHeight).call(this);
    if (!__classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedLineHeight, "f")) {
      return;
    }
    const value = this.control.value;
    if (!force && this.minRows === __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_previousMinRows, "f") && value === __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_previousValue, "f")) {
      return;
    }
    const scrollHeight = __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_measureScrollHeight).call(this);
    const height = Math.max(scrollHeight, __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedPlaceholderHeight, "f") || 0);
    this.control.style.height = `${height}px`;
    setTimeout(() => __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_scrollToCaretPosition).call(this));
    __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_previousValue, value, "f");
    __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_previousMinRows, this.minRows, "f");
  }
  /** Resets the `textarea` to its original size. */
  reset() {
    if (__classPrivateFieldGet(this, _M3eTextareaAutosizeElement_initialHeight, "f") !== undefined && this.control instanceof HTMLTextAreaElement) {
      this.control.style.height = __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_initialHeight, "f");
    }
  }
  /** @private */
  _handleWindowResize() {
    __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_cachedLineHeight, __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_cachedPlaceholderHeight, undefined, "f"), "f");
    this.resizeToFitContent(true);
  }
};
_M3eTextareaAutosizeElement_initialHeight = new WeakMap();
_M3eTextareaAutosizeElement_cachedLineHeight = new WeakMap();
_M3eTextareaAutosizeElement_cachedPlaceholderHeight = new WeakMap();
_M3eTextareaAutosizeElement_previousMinRows = new WeakMap();
_M3eTextareaAutosizeElement_previousValue = new WeakMap();
_M3eTextareaAutosizeElement_hasFocus = new WeakMap();
_M3eTextareaAutosizeElement_windowResizeHandler = new WeakMap();
_M3eTextareaAutosizeElement_focusHandler = new WeakMap();
_M3eTextareaAutosizeElement_inputHandler = new WeakMap();
_M3eTextareaAutosizeElement_instances = new WeakSet();
_M3eTextareaAutosizeElement_cacheTextareaLineHeight = function _M3eTextareaAutosizeElement_cacheTextareaLineHeight() {
  if (__classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedLineHeight, "f") || !(this.control instanceof HTMLTextAreaElement)) {
    return;
  }
  const clone = this.control.cloneNode(false);
  clone.rows = 1;
  clone.style.position = "absolute";
  clone.style.visibility = "hidden";
  clone.style.border = "none";
  clone.style.padding = "0";
  clone.style.height = "";
  clone.style.minHeight = "";
  clone.style.maxHeight = "";
  clone.style.top = clone.style.bottom = clone.style.left = clone.style.right = "auto";
  clone.style.overflow = "hidden";
  this.control.parentElement?.appendChild(clone);
  __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_cachedLineHeight, clone.clientHeight, "f");
  clone.remove();
  __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_setMinHeight).call(this);
  __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_setMaxHeight).call(this);
};
_M3eTextareaAutosizeElement_cacheTextareaPlaceholderHeight = function _M3eTextareaAutosizeElement_cacheTextareaPlaceholderHeight() {
  if (!(this.control instanceof HTMLTextAreaElement) || __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedPlaceholderHeight, "f") != undefined) {
    return;
  }
  if (!this.control.placeholder) {
    __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_cachedPlaceholderHeight, 0, "f");
    return;
  }
  const value = this.control.value;
  this.control.value = this.control.placeholder;
  __classPrivateFieldSet(this, _M3eTextareaAutosizeElement_cachedPlaceholderHeight, __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_instances, "m", _M3eTextareaAutosizeElement_measureScrollHeight).call(this), "f");
  this.control.value = value;
};
_M3eTextareaAutosizeElement_setMinHeight = function _M3eTextareaAutosizeElement_setMinHeight() {
  const minHeight = this.minRows && __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedLineHeight, "f") ? `${this.minRows * __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedLineHeight, "f")}px` : null;
  if (minHeight && this.control) {
    this.control.style.minHeight = minHeight;
  }
};
_M3eTextareaAutosizeElement_setMaxHeight = function _M3eTextareaAutosizeElement_setMaxHeight() {
  const maxHeight = this.maxRows && __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedLineHeight, "f") ? `${this.maxRows * __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_cachedLineHeight, "f")}px` : null;
  if (maxHeight && this.control) {
    this.control.style.maxHeight = maxHeight;
  }
};
_M3eTextareaAutosizeElement_measureScrollHeight = function _M3eTextareaAutosizeElement_measureScrollHeight() {
  if (!this.control) {
    return 0;
  }
  const element = this.control;
  const previousMargin = element.style.marginBottom || "";
  const isFirefox = navigator.userAgent.includes("Firefox");
  const needsMarginFiller = isFirefox && __classPrivateFieldGet(this, _M3eTextareaAutosizeElement_hasFocus, "f");
  if (needsMarginFiller) {
    element.style.marginBottom = `${element.clientHeight}px`;
  }
  const initialStyle = {
    padding: element.style.padding,
    boxSizing: element.style.boxSizing,
    height: element.style.height,
    overflow: element.style.overflow
  };
  element.style.padding = "2px 0";
  element.style.boxSizing = "content-box";
  if (!isFirefox) {
    element.style.height = "auto";
    element.style.overflow = "hidden";
  } else {
    element.style.height = "0";
  }
  const scrollHeight = element.scrollHeight - 4;
  element.style.padding = initialStyle.padding;
  element.style.boxSizing = initialStyle.boxSizing;
  element.style.height = initialStyle.height;
  element.style.overflow = initialStyle.overflow;
  if (needsMarginFiller) {
    element.style.marginBottom = previousMargin;
  }
  return scrollHeight;
};
_M3eTextareaAutosizeElement_scrollToCaretPosition = function _M3eTextareaAutosizeElement_scrollToCaretPosition() {
  if (!(this.control instanceof HTMLTextAreaElement) || !__classPrivateFieldGet(this, _M3eTextareaAutosizeElement_hasFocus, "f")) {
    return;
  }
  const {
    selectionStart,
    selectionEnd
  } = this.control;
  this.control.setSelectionRange(selectionStart, selectionEnd);
};
/** The styles of the element. */
M3eTextareaAutosizeElement.styles = css`:host { display: none; }`;
__decorate([n({
  attribute: "max-rows",
  type: Number
})], M3eTextareaAutosizeElement.prototype, "maxRows", void 0);
__decorate([n({
  attribute: "min-rows",
  type: Number
})], M3eTextareaAutosizeElement.prototype, "minRows", void 0);
__decorate([n({
  type: Boolean,
  reflect: true
})], M3eTextareaAutosizeElement.prototype, "disabled", void 0);
__decorate([debounce(16)], M3eTextareaAutosizeElement.prototype, "_handleWindowResize", null);
M3eTextareaAutosizeElement = __decorate([t$1("m3e-textarea-autosize")], M3eTextareaAutosizeElement);

export { M3eTextareaAutosizeElement };
